// @babylonjs/gui@9.16.1/3D/materials/mrdl/mrdlInnerquadMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
class MRDLInnerquadMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        this._needNormals = true;
        this._needUVs = true;
        this.rebuild();
    }
}
let MRDLInnerquadMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _color_decorators;
    let _color_initializers = [];
    let _color_extraInitializers = [];
    let _radius_decorators;
    let _radius_initializers = [];
    let _radius_extraInitializers = [];
    let _fixedRadius_decorators;
    let _fixedRadius_initializers = [];
    let _fixedRadius_extraInitializers = [];
    let _glowFraction_decorators;
    let _glowFraction_initializers = [];
    let _glowFraction_extraInitializers = [];
    let _glowMax_decorators;
    let _glowMax_initializers = [];
    let _glowMax_extraInitializers = [];
    let _glowFalloff_decorators;
    let _glowFalloff_initializers = [];
    let _glowFalloff_extraInitializers = [];
    return _a = class MRDLInnerquadMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                /**
                 * Gets or sets the color of the innerquad.
                 */
                this.color = __runInitializers(this, _color_initializers, new Color4(1, 1, 1, 0.05));
                /**
                 * Gets or sets the corner radius on the innerquad. If this value is changed, update the lineWidth to match.
                 */
                this.radius = (__runInitializers(this, _color_extraInitializers), __runInitializers(this, _radius_initializers, 0.12));
                /**
                 * Gets or sets whether the radius of the innerquad should be fixed.
                 */
                this.fixedRadius = (__runInitializers(this, _radius_extraInitializers), __runInitializers(this, _fixedRadius_initializers, true));
                /** @hidden */
                this._filterWidth = (__runInitializers(this, _fixedRadius_extraInitializers), 1.0);
                /**
                 * Gets or sets the glow fraction of the innerquad.
                 */
                this.glowFraction = __runInitializers(this, _glowFraction_initializers, 0.0);
                /**
                 * Gets or sets the maximum glow intensity of the innerquad.
                 */
                this.glowMax = (__runInitializers(this, _glowFraction_extraInitializers), __runInitializers(this, _glowMax_initializers, 0.5));
                /**
                 * Gets or sets the glow falloff effect of the innerquad.
                 */
                this.glowFalloff = (__runInitializers(this, _glowMax_extraInitializers), __runInitializers(this, _glowFalloff_initializers, 2.0));
                __runInitializers(this, _glowFalloff_extraInitializers);
                this.alphaMode = Constants.ALPHA_COMBINE;
                this.backFaceCulling = false;
            }
            needAlphaBlending() {
                return true;
            }
            needAlphaTesting() {
                return false;
            }
            getAlphaTestTexture() {
                return null;
            }
            // Methods
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new MRDLInnerquadMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, true, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "mrdlInnerquad";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "worldView",
                        "worldViewProjection",
                        "view",
                        "projection",
                        "viewProjection",
                        "cameraPosition",
                        "_Color_",
                        "_Radius_",
                        "_Fixed_Radius_",
                        "_Filter_Width_",
                        "_Glow_Fraction_",
                        "_Glow_Max_",
                        "_Glow_Falloff_",
                    ];
                    const samplers = [];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        extraInitializationsAsync: async () => {
                            await Promise.all([import("./shaders/mrdlInnerquad.vertex.js"), import("./shaders/mrdlInnerquad.fragment.js")]);
                        },
                    }, engine), defines);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const scene = this.getScene();
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", scene.getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", scene.activeCamera.position);
                // "Color"
                this._activeEffect.setDirectColor4("_Color_", this.color);
                // "Shape"
                this._activeEffect.setFloat("_Radius_", this.radius);
                this._activeEffect.setFloat("_Fixed_Radius_", this.fixedRadius ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Filter_Width_", this._filterWidth);
                // "Glow"
                this._activeEffect.setFloat("_Glow_Fraction_", this.glowFraction);
                this._activeEffect.setFloat("_Glow_Max_", this.glowMax);
                this._activeEffect.setFloat("_Glow_Falloff_", this.glowFalloff);
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = SerializationHelper.Serialize(this);
                serializationObject.customType = "BABYLON.MRDLInnerquadMaterial";
                return serializationObject;
            }
            getClassName() {
                return "MRDLInnerquadMaterial";
            }
            // Statics
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _color_decorators = [serialize()];
            _radius_decorators = [serialize()];
            _fixedRadius_decorators = [serialize()];
            _glowFraction_decorators = [serialize()];
            _glowMax_decorators = [serialize()];
            _glowFalloff_decorators = [serialize()];
            __esDecorate(null, null, _color_decorators, { kind: "field", name: "color", static: false, private: false, access: { has: obj => "color" in obj, get: obj => obj.color, set: (obj, value) => { obj.color = value; } }, metadata: _metadata }, _color_initializers, _color_extraInitializers);
            __esDecorate(null, null, _radius_decorators, { kind: "field", name: "radius", static: false, private: false, access: { has: obj => "radius" in obj, get: obj => obj.radius, set: (obj, value) => { obj.radius = value; } }, metadata: _metadata }, _radius_initializers, _radius_extraInitializers);
            __esDecorate(null, null, _fixedRadius_decorators, { kind: "field", name: "fixedRadius", static: false, private: false, access: { has: obj => "fixedRadius" in obj, get: obj => obj.fixedRadius, set: (obj, value) => { obj.fixedRadius = value; } }, metadata: _metadata }, _fixedRadius_initializers, _fixedRadius_extraInitializers);
            __esDecorate(null, null, _glowFraction_decorators, { kind: "field", name: "glowFraction", static: false, private: false, access: { has: obj => "glowFraction" in obj, get: obj => obj.glowFraction, set: (obj, value) => { obj.glowFraction = value; } }, metadata: _metadata }, _glowFraction_initializers, _glowFraction_extraInitializers);
            __esDecorate(null, null, _glowMax_decorators, { kind: "field", name: "glowMax", static: false, private: false, access: { has: obj => "glowMax" in obj, get: obj => obj.glowMax, set: (obj, value) => { obj.glowMax = value; } }, metadata: _metadata }, _glowMax_initializers, _glowMax_extraInitializers);
            __esDecorate(null, null, _glowFalloff_decorators, { kind: "field", name: "glowFalloff", static: false, private: false, access: { has: obj => "glowFalloff" in obj, get: obj => obj.glowFalloff, set: (obj, value) => { obj.glowFalloff = value; } }, metadata: _metadata }, _glowFalloff_initializers, _glowFalloff_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
export { MRDLInnerquadMaterial };
let _Registered = false;
/**
 * Registers the MRDLInnerquadMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterMRDLInnerquadMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.MRDLInnerquadMaterial", MRDLInnerquadMaterial);
}
