// @babylonjs/gui@9.16.1/3D/materials/mrdl/mrdlFrontplateMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { Vector3 } from "@babylonjs/core/Maths/math.vector.pure.js";
import { Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { Texture } from "@babylonjs/core/Materials/Textures/texture.pure.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
/** @hidden */
class MRDLFrontplateMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        /**
         * Sets default value for "SMOOTH_EDGES"
         */
        this.SMOOTH_EDGES = true;
        this._needNormals = true;
        this._needUVs = true;
        this.rebuild();
    }
}
let MRDLFrontplateMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _radius_decorators;
    let _radius_initializers = [];
    let _radius_extraInitializers = [];
    let _lineWidth_decorators;
    let _lineWidth_initializers = [];
    let _lineWidth_extraInitializers = [];
    let _relativeToHeight_decorators;
    let _relativeToHeight_initializers = [];
    let _relativeToHeight_extraInitializers = [];
    let _edgeColor_decorators;
    let _edgeColor_initializers = [];
    let _edgeColor_extraInitializers = [];
    let _blobEnable_decorators;
    let _blobEnable_initializers = [];
    let _blobEnable_extraInitializers = [];
    let _blobPosition_decorators;
    let _blobPosition_initializers = [];
    let _blobPosition_extraInitializers = [];
    let _blobIntensity_decorators;
    let _blobIntensity_initializers = [];
    let _blobIntensity_extraInitializers = [];
    let _blobNearSize_decorators;
    let _blobNearSize_initializers = [];
    let _blobNearSize_extraInitializers = [];
    let _blobFarSize_decorators;
    let _blobFarSize_initializers = [];
    let _blobFarSize_extraInitializers = [];
    let _blobNearDistance_decorators;
    let _blobNearDistance_initializers = [];
    let _blobNearDistance_extraInitializers = [];
    let _blobFarDistance_decorators;
    let _blobFarDistance_initializers = [];
    let _blobFarDistance_extraInitializers = [];
    let _blobFadeLength_decorators;
    let _blobFadeLength_initializers = [];
    let _blobFadeLength_extraInitializers = [];
    let _blobInnerFade_decorators;
    let _blobInnerFade_initializers = [];
    let _blobInnerFade_extraInitializers = [];
    let _blobPulse_decorators;
    let _blobPulse_initializers = [];
    let _blobPulse_extraInitializers = [];
    let _blobFade_decorators;
    let _blobFade_initializers = [];
    let _blobFade_extraInitializers = [];
    let _blobPulseMaxSize_decorators;
    let _blobPulseMaxSize_initializers = [];
    let _blobPulseMaxSize_extraInitializers = [];
    let _blobEnable2_decorators;
    let _blobEnable2_initializers = [];
    let _blobEnable2_extraInitializers = [];
    let _blobPosition2_decorators;
    let _blobPosition2_initializers = [];
    let _blobPosition2_extraInitializers = [];
    let _blobNearSize2_decorators;
    let _blobNearSize2_initializers = [];
    let _blobNearSize2_extraInitializers = [];
    let _blobInnerFade2_decorators;
    let _blobInnerFade2_initializers = [];
    let _blobInnerFade2_extraInitializers = [];
    let _blobPulse2_decorators;
    let _blobPulse2_initializers = [];
    let _blobPulse2_extraInitializers = [];
    let _blobFade2_decorators;
    let _blobFade2_initializers = [];
    let _blobFade2_extraInitializers = [];
    let _gazeIntensity_decorators;
    let _gazeIntensity_initializers = [];
    let _gazeIntensity_extraInitializers = [];
    let _gazeFocus_decorators;
    let _gazeFocus_initializers = [];
    let _gazeFocus_extraInitializers = [];
    let _selectionFuzz_decorators;
    let _selectionFuzz_initializers = [];
    let _selectionFuzz_extraInitializers = [];
    let _selected_decorators;
    let _selected_initializers = [];
    let _selected_extraInitializers = [];
    let _selectionFade_decorators;
    let _selectionFade_initializers = [];
    let _selectionFade_extraInitializers = [];
    let _selectionFadeSize_decorators;
    let _selectionFadeSize_initializers = [];
    let _selectionFadeSize_extraInitializers = [];
    let _selectedDistance_decorators;
    let _selectedDistance_initializers = [];
    let _selectedDistance_extraInitializers = [];
    let _selectedFadeLength_decorators;
    let _selectedFadeLength_initializers = [];
    let _selectedFadeLength_extraInitializers = [];
    let _proximityMaxIntensity_decorators;
    let _proximityMaxIntensity_initializers = [];
    let _proximityMaxIntensity_extraInitializers = [];
    let _proximityFarDistance_decorators;
    let _proximityFarDistance_initializers = [];
    let _proximityFarDistance_extraInitializers = [];
    let _proximityNearRadius_decorators;
    let _proximityNearRadius_initializers = [];
    let _proximityNearRadius_extraInitializers = [];
    let _proximityAnisotropy_decorators;
    let _proximityAnisotropy_initializers = [];
    let _proximityAnisotropy_extraInitializers = [];
    let _useGlobalLeftIndex_decorators;
    let _useGlobalLeftIndex_initializers = [];
    let _useGlobalLeftIndex_extraInitializers = [];
    let _useGlobalRightIndex_decorators;
    let _useGlobalRightIndex_initializers = [];
    let _useGlobalRightIndex_extraInitializers = [];
    return _a = class MRDLFrontplateMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                /**
                 * Gets or sets the corner radius on the frontplate. If this value is changed, update the lineWidth to match.
                 */
                this.radius = __runInitializers(this, _radius_initializers, 0.12);
                /**
                 * Gets or sets the line width of the frontplate.
                 */
                this.lineWidth = (__runInitializers(this, _radius_extraInitializers), __runInitializers(this, _lineWidth_initializers, 0.01));
                /**
                 * Gets or sets whether the scale is relative to the frontplate height.
                 */
                this.relativeToHeight = (__runInitializers(this, _lineWidth_extraInitializers), __runInitializers(this, _relativeToHeight_initializers, false));
                /** @hidden */
                this._filterWidth = (__runInitializers(this, _relativeToHeight_extraInitializers), 1.0);
                /**
                 * Gets or sets the edge color of the frontplate.
                 */
                this.edgeColor = __runInitializers(this, _edgeColor_initializers, new Color4(0.53, 0.53, 0.53, 1));
                /**
                 * Gets or sets whether to enable blob effects on the frontplate.
                 */
                this.blobEnable = (__runInitializers(this, _edgeColor_extraInitializers), __runInitializers(this, _blobEnable_initializers, true));
                /**
                 * Gets or sets the blob position on the frontplate.
                 */
                this.blobPosition = (__runInitializers(this, _blobEnable_extraInitializers), __runInitializers(this, _blobPosition_initializers, new Vector3(100, 100, 100)));
                /**
                 * Gets or sets the blob intensity of the frontplate.
                 */
                this.blobIntensity = (__runInitializers(this, _blobPosition_extraInitializers), __runInitializers(this, _blobIntensity_initializers, 0.5));
                /**
                 * Gets or sets the blob near size of the frontplate.
                 */
                this.blobNearSize = (__runInitializers(this, _blobIntensity_extraInitializers), __runInitializers(this, _blobNearSize_initializers, 0.032));
                /**
                 * Gets or sets the blob far size of the frontplate.
                 */
                this.blobFarSize = (__runInitializers(this, _blobNearSize_extraInitializers), __runInitializers(this, _blobFarSize_initializers, 0.048));
                /**
                 * Gets or sets the blob near distance of the frontplate.
                 */
                this.blobNearDistance = (__runInitializers(this, _blobFarSize_extraInitializers), __runInitializers(this, _blobNearDistance_initializers, 0.008));
                /**
                 * Gets or sets the blob far distance of the frontplate.
                 */
                this.blobFarDistance = (__runInitializers(this, _blobNearDistance_extraInitializers), __runInitializers(this, _blobFarDistance_initializers, 0.064));
                /**
                 * Gets or sets the blob fade length of the frontplate.
                 */
                this.blobFadeLength = (__runInitializers(this, _blobFarDistance_extraInitializers), __runInitializers(this, _blobFadeLength_initializers, 0.04));
                /**
                 * Gets or sets the blob inner fade of the frontplate.
                 */
                this.blobInnerFade = (__runInitializers(this, _blobFadeLength_extraInitializers), __runInitializers(this, _blobInnerFade_initializers, 0.01));
                /**
                 * Gets or sets the blob pulse of the frontplate.
                 */
                this.blobPulse = (__runInitializers(this, _blobInnerFade_extraInitializers), __runInitializers(this, _blobPulse_initializers, 0.0));
                /**
                 * Gets or sets the blob fade effect on the frontplate.
                 */
                this.blobFade = (__runInitializers(this, _blobPulse_extraInitializers), __runInitializers(this, _blobFade_initializers, 1.0));
                /**
                 * Gets or sets the maximum size of the blob pulse on the frontplate.
                 */
                this.blobPulseMaxSize = (__runInitializers(this, _blobFade_extraInitializers), __runInitializers(this, _blobPulseMaxSize_initializers, 0.05));
                /**
                 * Gets or sets whether to enable extra blob effects of the frontplate.
                 */
                this.blobEnable2 = (__runInitializers(this, _blobPulseMaxSize_extraInitializers), __runInitializers(this, _blobEnable2_initializers, true));
                /**
                 * Gets or sets blob2 position of the frontplate.
                 */
                this.blobPosition2 = (__runInitializers(this, _blobEnable2_extraInitializers), __runInitializers(this, _blobPosition2_initializers, new Vector3(10, 10.1, -0.6)));
                /**
                 * Gets or sets the blob2 near size of the frontplate.
                 */
                this.blobNearSize2 = (__runInitializers(this, _blobPosition2_extraInitializers), __runInitializers(this, _blobNearSize2_initializers, 0.008));
                /**
                 * Gets or sets the blob2 inner fade of the frontplate.
                 */
                this.blobInnerFade2 = (__runInitializers(this, _blobNearSize2_extraInitializers), __runInitializers(this, _blobInnerFade2_initializers, 0.1));
                /**
                 * Gets or sets the blob2 pulse of the frontplate.
                 */
                this.blobPulse2 = (__runInitializers(this, _blobInnerFade2_extraInitializers), __runInitializers(this, _blobPulse2_initializers, 0.0));
                /**
                 * Gets or sets the blob2 fade effect on the frontplate.
                 */
                this.blobFade2 = (__runInitializers(this, _blobPulse2_extraInitializers), __runInitializers(this, _blobFade2_initializers, 1.0));
                /**
                 * Gets or sets the gaze intensity of the frontplate.
                 */
                this.gazeIntensity = (__runInitializers(this, _blobFade2_extraInitializers), __runInitializers(this, _gazeIntensity_initializers, 0.8));
                /**
                 * Gets or sets the gaze focus of the frontplate.
                 */
                this.gazeFocus = (__runInitializers(this, _gazeIntensity_extraInitializers), __runInitializers(this, _gazeFocus_initializers, 0.0));
                /**
                 * Gets or sets the selection fuzz of the frontplate.
                 */
                this.selectionFuzz = (__runInitializers(this, _gazeFocus_extraInitializers), __runInitializers(this, _selectionFuzz_initializers, 0.5));
                /**
                 * Gets or sets the fade intensity of the frontplate.
                 */
                this.selected = (__runInitializers(this, _selectionFuzz_extraInitializers), __runInitializers(this, _selected_initializers, 1.0));
                /**
                 * Gets or sets the selection fade intensity of the frontplate.
                 */
                this.selectionFade = (__runInitializers(this, _selected_extraInitializers), __runInitializers(this, _selectionFade_initializers, 0.2));
                /**
                 * Gets or sets the selection fade size of the frontplate.
                 */
                this.selectionFadeSize = (__runInitializers(this, _selectionFade_extraInitializers), __runInitializers(this, _selectionFadeSize_initializers, 0.0));
                /**
                 * Gets or sets the selected distance of the frontplate.
                 */
                this.selectedDistance = (__runInitializers(this, _selectionFadeSize_extraInitializers), __runInitializers(this, _selectedDistance_initializers, 0.08));
                /**
                 * Gets or sets the selected fade length of the frontplate.
                 */
                this.selectedFadeLength = (__runInitializers(this, _selectedDistance_extraInitializers), __runInitializers(this, _selectedFadeLength_initializers, 0.08));
                /**
                 * Gets or sets the proximity maximum intensity of the frontplate.
                 */
                this.proximityMaxIntensity = (__runInitializers(this, _selectedFadeLength_extraInitializers), __runInitializers(this, _proximityMaxIntensity_initializers, 0.45));
                /**
                 * Gets or sets the proximity far distance of the frontplate.
                 */
                this.proximityFarDistance = (__runInitializers(this, _proximityMaxIntensity_extraInitializers), __runInitializers(this, _proximityFarDistance_initializers, 0.16));
                /**
                 * Gets or sets the proximity near radius of the frontplate.
                 */
                this.proximityNearRadius = (__runInitializers(this, _proximityFarDistance_extraInitializers), __runInitializers(this, _proximityNearRadius_initializers, 0.016));
                /**
                 * Gets or sets the proximity anisotropy of the frontplate.
                 */
                this.proximityAnisotropy = (__runInitializers(this, _proximityNearRadius_extraInitializers), __runInitializers(this, _proximityAnisotropy_initializers, 1.0));
                /**
                 * Gets or sets whether to use global left index on the frontplate.
                 */
                this.useGlobalLeftIndex = (__runInitializers(this, _proximityAnisotropy_extraInitializers), __runInitializers(this, _useGlobalLeftIndex_initializers, true));
                /**
                 * Gets or sets  whether to use global right index of the frontplate.
                 */
                this.useGlobalRightIndex = (__runInitializers(this, _useGlobalLeftIndex_extraInitializers), __runInitializers(this, _useGlobalRightIndex_initializers, true));
                /**
                 * Gets or sets the opacity of the frontplate (0.0 - 1.0).
                 */
                this.fadeOut = (__runInitializers(this, _useGlobalRightIndex_extraInitializers), 1.0);
                this.alphaMode = Constants.ALPHA_ADD;
                this.disableDepthWrite = true;
                this.backFaceCulling = false;
                this._blobTexture = new Texture(_a.BLOB_TEXTURE_URL, scene, true, false, Texture.NEAREST_SAMPLINGMODE);
            }
            needAlphaBlending() {
                return true;
            }
            needAlphaTesting() {
                return false;
            }
            getAlphaTestTexture() {
                return null;
            }
            // Methods
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new MRDLFrontplateMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                if (_a.BLOB_TEXTURE_URL !== "" && !this._blobTexture.isReady()) {
                    return false;
                }
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, false, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "mrdlFrontplate";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "worldView",
                        "worldViewProjection",
                        "view",
                        "projection",
                        "viewProjection",
                        "cameraPosition",
                        "_Radius_",
                        "_Line_Width_",
                        "_Relative_To_Height_",
                        "_Filter_Width_",
                        "_Edge_Color_",
                        "_Fade_Out_",
                        "_Smooth_Edges_",
                        "_Blob_Enable_",
                        "_Blob_Position_",
                        "_Blob_Intensity_",
                        "_Blob_Near_Size_",
                        "_Blob_Far_Size_",
                        "_Blob_Near_Distance_",
                        "_Blob_Far_Distance_",
                        "_Blob_Fade_Length_",
                        "_Blob_Inner_Fade_",
                        "_Blob_Pulse_",
                        "_Blob_Fade_",
                        "_Blob_Pulse_Max_Size_",
                        "_Blob_Enable_2_",
                        "_Blob_Position_2_",
                        "_Blob_Near_Size_2_",
                        "_Blob_Inner_Fade_2_",
                        "_Blob_Pulse_2_",
                        "_Blob_Fade_2_",
                        "_Gaze_Intensity_",
                        "_Gaze_Focus_",
                        "_Blob_Texture_",
                        "_Selection_Fuzz_",
                        "_Selected_",
                        "_Selection_Fade_",
                        "_Selection_Fade_Size_",
                        "_Selected_Distance_",
                        "_Selected_Fade_Length_",
                        "_Proximity_Max_Intensity_",
                        "_Proximity_Far_Distance_",
                        "_Proximity_Near_Radius_",
                        "_Proximity_Anisotropy_",
                        "Global_Left_Index_Tip_Position",
                        "Global_Right_Index_Tip_Position",
                        "_Use_Global_Left_Index_",
                        "_Use_Global_Right_Index_",
                    ];
                    const samplers = [];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        extraInitializationsAsync: async () => {
                            await Promise.all([import("./shaders/mrdlFrontplate.vertex.js"), import("./shaders/mrdlFrontplate.fragment.js")]);
                        },
                    }, engine), defines);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const scene = this.getScene();
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", scene.getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", scene.activeCamera.position);
                // "Round Rect"
                this._activeEffect.setFloat("_Radius_", this.radius);
                this._activeEffect.setFloat("_Line_Width_", this.lineWidth);
                this._activeEffect.setFloat("_Relative_To_Height_", this.relativeToHeight ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Filter_Width_", this._filterWidth);
                this._activeEffect.setDirectColor4("_Edge_Color_", this.edgeColor);
                // "Fade"
                this._activeEffect.setFloat("_Fade_Out_", this.fadeOut);
                // "Blob"
                this._activeEffect.setFloat("_Blob_Enable_", this.blobEnable ? 1.0 : 0.0);
                this._activeEffect.setVector3("_Blob_Position_", this.blobPosition);
                this._activeEffect.setFloat("_Blob_Intensity_", this.blobIntensity);
                this._activeEffect.setFloat("_Blob_Near_Size_", this.blobNearSize);
                this._activeEffect.setFloat("_Blob_Far_Size_", this.blobFarSize);
                this._activeEffect.setFloat("_Blob_Near_Distance_", this.blobNearDistance);
                this._activeEffect.setFloat("_Blob_Far_Distance_", this.blobFarDistance);
                this._activeEffect.setFloat("_Blob_Fade_Length_", this.blobFadeLength);
                this._activeEffect.setFloat("_Blob_Inner_Fade_", this.blobInnerFade);
                this._activeEffect.setFloat("_Blob_Pulse_", this.blobPulse);
                this._activeEffect.setFloat("_Blob_Fade_", this.blobFade);
                this._activeEffect.setFloat("_Blob_Pulse_Max_Size_", this.blobPulseMaxSize);
                // "Blob 2"
                this._activeEffect.setFloat("_Blob_Enable_2_", this.blobEnable2 ? 1.0 : 0.0);
                this._activeEffect.setVector3("_Blob_Position_2_", this.blobPosition2);
                this._activeEffect.setFloat("_Blob_Near_Size_2_", this.blobNearSize2);
                this._activeEffect.setFloat("_Blob_Inner_Fade_2_", this.blobInnerFade2);
                this._activeEffect.setFloat("_Blob_Pulse_2_", this.blobPulse2);
                this._activeEffect.setFloat("_Blob_Fade_2_", this.blobFade2);
                // "Gaze"
                this._activeEffect.setFloat("_Gaze_Intensity_", this.gazeIntensity);
                this._activeEffect.setFloat("_Gaze_Focus_", this.gazeFocus);
                // "Blob Texture"
                this._activeEffect.setTexture("_Blob_Texture_", this._blobTexture);
                // "Selection"
                this._activeEffect.setFloat("_Selection_Fuzz_", this.selectionFuzz);
                this._activeEffect.setFloat("_Selected_", this.selected);
                this._activeEffect.setFloat("_Selection_Fade_", this.selectionFade);
                this._activeEffect.setFloat("_Selection_Fade_Size_", this.selectionFadeSize);
                this._activeEffect.setFloat("_Selected_Distance_", this.selectedDistance);
                this._activeEffect.setFloat("_Selected_Fade_Length_", this.selectedFadeLength);
                // "Proximity"
                this._activeEffect.setFloat("_Proximity_Max_Intensity_", this.proximityMaxIntensity);
                this._activeEffect.setFloat("_Proximity_Far_Distance_", this.proximityFarDistance);
                this._activeEffect.setFloat("_Proximity_Near_Radius_", this.proximityNearRadius);
                this._activeEffect.setFloat("_Proximity_Anisotropy_", this.proximityAnisotropy);
                // "Global"
                this._activeEffect.setFloat("_Use_Global_Left_Index_", this.useGlobalLeftIndex ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Use_Global_Right_Index_", this.useGlobalRightIndex ? 1.0 : 0.0);
                // "Antialiasing"
                //define SMOOTH_EDGES true;
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = SerializationHelper.Serialize(this);
                serializationObject.customType = "BABYLON.MRDLFrontplateMaterial";
                return serializationObject;
            }
            getClassName() {
                return "MRDLFrontplateMaterial";
            }
            // Statics
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _radius_decorators = [serialize()];
            _lineWidth_decorators = [serialize()];
            _relativeToHeight_decorators = [serialize()];
            _edgeColor_decorators = [serialize()];
            _blobEnable_decorators = [serialize()];
            _blobPosition_decorators = [serialize()];
            _blobIntensity_decorators = [serialize()];
            _blobNearSize_decorators = [serialize()];
            _blobFarSize_decorators = [serialize()];
            _blobNearDistance_decorators = [serialize()];
            _blobFarDistance_decorators = [serialize()];
            _blobFadeLength_decorators = [serialize()];
            _blobInnerFade_decorators = [serialize()];
            _blobPulse_decorators = [serialize()];
            _blobFade_decorators = [serialize()];
            _blobPulseMaxSize_decorators = [serialize()];
            _blobEnable2_decorators = [serialize()];
            _blobPosition2_decorators = [serialize()];
            _blobNearSize2_decorators = [serialize()];
            _blobInnerFade2_decorators = [serialize()];
            _blobPulse2_decorators = [serialize()];
            _blobFade2_decorators = [serialize()];
            _gazeIntensity_decorators = [serialize()];
            _gazeFocus_decorators = [serialize()];
            _selectionFuzz_decorators = [serialize()];
            _selected_decorators = [serialize()];
            _selectionFade_decorators = [serialize()];
            _selectionFadeSize_decorators = [serialize()];
            _selectedDistance_decorators = [serialize()];
            _selectedFadeLength_decorators = [serialize()];
            _proximityMaxIntensity_decorators = [serialize()];
            _proximityFarDistance_decorators = [serialize()];
            _proximityNearRadius_decorators = [serialize()];
            _proximityAnisotropy_decorators = [serialize()];
            _useGlobalLeftIndex_decorators = [serialize()];
            _useGlobalRightIndex_decorators = [serialize()];
            __esDecorate(null, null, _radius_decorators, { kind: "field", name: "radius", static: false, private: false, access: { has: obj => "radius" in obj, get: obj => obj.radius, set: (obj, value) => { obj.radius = value; } }, metadata: _metadata }, _radius_initializers, _radius_extraInitializers);
            __esDecorate(null, null, _lineWidth_decorators, { kind: "field", name: "lineWidth", static: false, private: false, access: { has: obj => "lineWidth" in obj, get: obj => obj.lineWidth, set: (obj, value) => { obj.lineWidth = value; } }, metadata: _metadata }, _lineWidth_initializers, _lineWidth_extraInitializers);
            __esDecorate(null, null, _relativeToHeight_decorators, { kind: "field", name: "relativeToHeight", static: false, private: false, access: { has: obj => "relativeToHeight" in obj, get: obj => obj.relativeToHeight, set: (obj, value) => { obj.relativeToHeight = value; } }, metadata: _metadata }, _relativeToHeight_initializers, _relativeToHeight_extraInitializers);
            __esDecorate(null, null, _edgeColor_decorators, { kind: "field", name: "edgeColor", static: false, private: false, access: { has: obj => "edgeColor" in obj, get: obj => obj.edgeColor, set: (obj, value) => { obj.edgeColor = value; } }, metadata: _metadata }, _edgeColor_initializers, _edgeColor_extraInitializers);
            __esDecorate(null, null, _blobEnable_decorators, { kind: "field", name: "blobEnable", static: false, private: false, access: { has: obj => "blobEnable" in obj, get: obj => obj.blobEnable, set: (obj, value) => { obj.blobEnable = value; } }, metadata: _metadata }, _blobEnable_initializers, _blobEnable_extraInitializers);
            __esDecorate(null, null, _blobPosition_decorators, { kind: "field", name: "blobPosition", static: false, private: false, access: { has: obj => "blobPosition" in obj, get: obj => obj.blobPosition, set: (obj, value) => { obj.blobPosition = value; } }, metadata: _metadata }, _blobPosition_initializers, _blobPosition_extraInitializers);
            __esDecorate(null, null, _blobIntensity_decorators, { kind: "field", name: "blobIntensity", static: false, private: false, access: { has: obj => "blobIntensity" in obj, get: obj => obj.blobIntensity, set: (obj, value) => { obj.blobIntensity = value; } }, metadata: _metadata }, _blobIntensity_initializers, _blobIntensity_extraInitializers);
            __esDecorate(null, null, _blobNearSize_decorators, { kind: "field", name: "blobNearSize", static: false, private: false, access: { has: obj => "blobNearSize" in obj, get: obj => obj.blobNearSize, set: (obj, value) => { obj.blobNearSize = value; } }, metadata: _metadata }, _blobNearSize_initializers, _blobNearSize_extraInitializers);
            __esDecorate(null, null, _blobFarSize_decorators, { kind: "field", name: "blobFarSize", static: false, private: false, access: { has: obj => "blobFarSize" in obj, get: obj => obj.blobFarSize, set: (obj, value) => { obj.blobFarSize = value; } }, metadata: _metadata }, _blobFarSize_initializers, _blobFarSize_extraInitializers);
            __esDecorate(null, null, _blobNearDistance_decorators, { kind: "field", name: "blobNearDistance", static: false, private: false, access: { has: obj => "blobNearDistance" in obj, get: obj => obj.blobNearDistance, set: (obj, value) => { obj.blobNearDistance = value; } }, metadata: _metadata }, _blobNearDistance_initializers, _blobNearDistance_extraInitializers);
            __esDecorate(null, null, _blobFarDistance_decorators, { kind: "field", name: "blobFarDistance", static: false, private: false, access: { has: obj => "blobFarDistance" in obj, get: obj => obj.blobFarDistance, set: (obj, value) => { obj.blobFarDistance = value; } }, metadata: _metadata }, _blobFarDistance_initializers, _blobFarDistance_extraInitializers);
            __esDecorate(null, null, _blobFadeLength_decorators, { kind: "field", name: "blobFadeLength", static: false, private: false, access: { has: obj => "blobFadeLength" in obj, get: obj => obj.blobFadeLength, set: (obj, value) => { obj.blobFadeLength = value; } }, metadata: _metadata }, _blobFadeLength_initializers, _blobFadeLength_extraInitializers);
            __esDecorate(null, null, _blobInnerFade_decorators, { kind: "field", name: "blobInnerFade", static: false, private: false, access: { has: obj => "blobInnerFade" in obj, get: obj => obj.blobInnerFade, set: (obj, value) => { obj.blobInnerFade = value; } }, metadata: _metadata }, _blobInnerFade_initializers, _blobInnerFade_extraInitializers);
            __esDecorate(null, null, _blobPulse_decorators, { kind: "field", name: "blobPulse", static: false, private: false, access: { has: obj => "blobPulse" in obj, get: obj => obj.blobPulse, set: (obj, value) => { obj.blobPulse = value; } }, metadata: _metadata }, _blobPulse_initializers, _blobPulse_extraInitializers);
            __esDecorate(null, null, _blobFade_decorators, { kind: "field", name: "blobFade", static: false, private: false, access: { has: obj => "blobFade" in obj, get: obj => obj.blobFade, set: (obj, value) => { obj.blobFade = value; } }, metadata: _metadata }, _blobFade_initializers, _blobFade_extraInitializers);
            __esDecorate(null, null, _blobPulseMaxSize_decorators, { kind: "field", name: "blobPulseMaxSize", static: false, private: false, access: { has: obj => "blobPulseMaxSize" in obj, get: obj => obj.blobPulseMaxSize, set: (obj, value) => { obj.blobPulseMaxSize = value; } }, metadata: _metadata }, _blobPulseMaxSize_initializers, _blobPulseMaxSize_extraInitializers);
            __esDecorate(null, null, _blobEnable2_decorators, { kind: "field", name: "blobEnable2", static: false, private: false, access: { has: obj => "blobEnable2" in obj, get: obj => obj.blobEnable2, set: (obj, value) => { obj.blobEnable2 = value; } }, metadata: _metadata }, _blobEnable2_initializers, _blobEnable2_extraInitializers);
            __esDecorate(null, null, _blobPosition2_decorators, { kind: "field", name: "blobPosition2", static: false, private: false, access: { has: obj => "blobPosition2" in obj, get: obj => obj.blobPosition2, set: (obj, value) => { obj.blobPosition2 = value; } }, metadata: _metadata }, _blobPosition2_initializers, _blobPosition2_extraInitializers);
            __esDecorate(null, null, _blobNearSize2_decorators, { kind: "field", name: "blobNearSize2", static: false, private: false, access: { has: obj => "blobNearSize2" in obj, get: obj => obj.blobNearSize2, set: (obj, value) => { obj.blobNearSize2 = value; } }, metadata: _metadata }, _blobNearSize2_initializers, _blobNearSize2_extraInitializers);
            __esDecorate(null, null, _blobInnerFade2_decorators, { kind: "field", name: "blobInnerFade2", static: false, private: false, access: { has: obj => "blobInnerFade2" in obj, get: obj => obj.blobInnerFade2, set: (obj, value) => { obj.blobInnerFade2 = value; } }, metadata: _metadata }, _blobInnerFade2_initializers, _blobInnerFade2_extraInitializers);
            __esDecorate(null, null, _blobPulse2_decorators, { kind: "field", name: "blobPulse2", static: false, private: false, access: { has: obj => "blobPulse2" in obj, get: obj => obj.blobPulse2, set: (obj, value) => { obj.blobPulse2 = value; } }, metadata: _metadata }, _blobPulse2_initializers, _blobPulse2_extraInitializers);
            __esDecorate(null, null, _blobFade2_decorators, { kind: "field", name: "blobFade2", static: false, private: false, access: { has: obj => "blobFade2" in obj, get: obj => obj.blobFade2, set: (obj, value) => { obj.blobFade2 = value; } }, metadata: _metadata }, _blobFade2_initializers, _blobFade2_extraInitializers);
            __esDecorate(null, null, _gazeIntensity_decorators, { kind: "field", name: "gazeIntensity", static: false, private: false, access: { has: obj => "gazeIntensity" in obj, get: obj => obj.gazeIntensity, set: (obj, value) => { obj.gazeIntensity = value; } }, metadata: _metadata }, _gazeIntensity_initializers, _gazeIntensity_extraInitializers);
            __esDecorate(null, null, _gazeFocus_decorators, { kind: "field", name: "gazeFocus", static: false, private: false, access: { has: obj => "gazeFocus" in obj, get: obj => obj.gazeFocus, set: (obj, value) => { obj.gazeFocus = value; } }, metadata: _metadata }, _gazeFocus_initializers, _gazeFocus_extraInitializers);
            __esDecorate(null, null, _selectionFuzz_decorators, { kind: "field", name: "selectionFuzz", static: false, private: false, access: { has: obj => "selectionFuzz" in obj, get: obj => obj.selectionFuzz, set: (obj, value) => { obj.selectionFuzz = value; } }, metadata: _metadata }, _selectionFuzz_initializers, _selectionFuzz_extraInitializers);
            __esDecorate(null, null, _selected_decorators, { kind: "field", name: "selected", static: false, private: false, access: { has: obj => "selected" in obj, get: obj => obj.selected, set: (obj, value) => { obj.selected = value; } }, metadata: _metadata }, _selected_initializers, _selected_extraInitializers);
            __esDecorate(null, null, _selectionFade_decorators, { kind: "field", name: "selectionFade", static: false, private: false, access: { has: obj => "selectionFade" in obj, get: obj => obj.selectionFade, set: (obj, value) => { obj.selectionFade = value; } }, metadata: _metadata }, _selectionFade_initializers, _selectionFade_extraInitializers);
            __esDecorate(null, null, _selectionFadeSize_decorators, { kind: "field", name: "selectionFadeSize", static: false, private: false, access: { has: obj => "selectionFadeSize" in obj, get: obj => obj.selectionFadeSize, set: (obj, value) => { obj.selectionFadeSize = value; } }, metadata: _metadata }, _selectionFadeSize_initializers, _selectionFadeSize_extraInitializers);
            __esDecorate(null, null, _selectedDistance_decorators, { kind: "field", name: "selectedDistance", static: false, private: false, access: { has: obj => "selectedDistance" in obj, get: obj => obj.selectedDistance, set: (obj, value) => { obj.selectedDistance = value; } }, metadata: _metadata }, _selectedDistance_initializers, _selectedDistance_extraInitializers);
            __esDecorate(null, null, _selectedFadeLength_decorators, { kind: "field", name: "selectedFadeLength", static: false, private: false, access: { has: obj => "selectedFadeLength" in obj, get: obj => obj.selectedFadeLength, set: (obj, value) => { obj.selectedFadeLength = value; } }, metadata: _metadata }, _selectedFadeLength_initializers, _selectedFadeLength_extraInitializers);
            __esDecorate(null, null, _proximityMaxIntensity_decorators, { kind: "field", name: "proximityMaxIntensity", static: false, private: false, access: { has: obj => "proximityMaxIntensity" in obj, get: obj => obj.proximityMaxIntensity, set: (obj, value) => { obj.proximityMaxIntensity = value; } }, metadata: _metadata }, _proximityMaxIntensity_initializers, _proximityMaxIntensity_extraInitializers);
            __esDecorate(null, null, _proximityFarDistance_decorators, { kind: "field", name: "proximityFarDistance", static: false, private: false, access: { has: obj => "proximityFarDistance" in obj, get: obj => obj.proximityFarDistance, set: (obj, value) => { obj.proximityFarDistance = value; } }, metadata: _metadata }, _proximityFarDistance_initializers, _proximityFarDistance_extraInitializers);
            __esDecorate(null, null, _proximityNearRadius_decorators, { kind: "field", name: "proximityNearRadius", static: false, private: false, access: { has: obj => "proximityNearRadius" in obj, get: obj => obj.proximityNearRadius, set: (obj, value) => { obj.proximityNearRadius = value; } }, metadata: _metadata }, _proximityNearRadius_initializers, _proximityNearRadius_extraInitializers);
            __esDecorate(null, null, _proximityAnisotropy_decorators, { kind: "field", name: "proximityAnisotropy", static: false, private: false, access: { has: obj => "proximityAnisotropy" in obj, get: obj => obj.proximityAnisotropy, set: (obj, value) => { obj.proximityAnisotropy = value; } }, metadata: _metadata }, _proximityAnisotropy_initializers, _proximityAnisotropy_extraInitializers);
            __esDecorate(null, null, _useGlobalLeftIndex_decorators, { kind: "field", name: "useGlobalLeftIndex", static: false, private: false, access: { has: obj => "useGlobalLeftIndex" in obj, get: obj => obj.useGlobalLeftIndex, set: (obj, value) => { obj.useGlobalLeftIndex = value; } }, metadata: _metadata }, _useGlobalLeftIndex_initializers, _useGlobalLeftIndex_extraInitializers);
            __esDecorate(null, null, _useGlobalRightIndex_decorators, { kind: "field", name: "useGlobalRightIndex", static: false, private: false, access: { has: obj => "useGlobalRightIndex" in obj, get: obj => obj.useGlobalRightIndex, set: (obj, value) => { obj.useGlobalRightIndex = value; } }, metadata: _metadata }, _useGlobalRightIndex_initializers, _useGlobalRightIndex_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        /**
         * URL pointing to the texture used to define the coloring for the BLOB.
         */
        _a.BLOB_TEXTURE_URL = "",
        _a;
})();
export { MRDLFrontplateMaterial };
let _Registered = false;
/**
 * Registers the MRDLFrontplateMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterMRDLFrontplateMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.MRDLFrontplateMaterial", MRDLFrontplateMaterial);
}
