// @babylonjs/gui@9.16.1/3D/materials/mrdl/mrdlBackglowMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
/** @hidden */
class MRDLBackglowMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        this._needNormals = true;
        this._needUVs = true;
        this.rebuild();
    }
}
let MRDLBackglowMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _bevelRadius_decorators;
    let _bevelRadius_initializers = [];
    let _bevelRadius_extraInitializers = [];
    let _lineWidth_decorators;
    let _lineWidth_initializers = [];
    let _lineWidth_extraInitializers = [];
    let _absoluteSizes_decorators;
    let _absoluteSizes_initializers = [];
    let _absoluteSizes_extraInitializers = [];
    let _tuningMotion_decorators;
    let _tuningMotion_initializers = [];
    let _tuningMotion_extraInitializers = [];
    let _motion_decorators;
    let _motion_initializers = [];
    let _motion_extraInitializers = [];
    let _maxIntensity_decorators;
    let _maxIntensity_initializers = [];
    let _maxIntensity_extraInitializers = [];
    let _intensityFadeInExponent_decorators;
    let _intensityFadeInExponent_initializers = [];
    let _intensityFadeInExponent_extraInitializers = [];
    let _outerFuzzStart_decorators;
    let _outerFuzzStart_initializers = [];
    let _outerFuzzStart_extraInitializers = [];
    let _outerFuzzEnd_decorators;
    let _outerFuzzEnd_initializers = [];
    let _outerFuzzEnd_extraInitializers = [];
    let _color_decorators;
    let _color_initializers = [];
    let _color_extraInitializers = [];
    let _innerColor_decorators;
    let _innerColor_initializers = [];
    let _innerColor_extraInitializers = [];
    let _blendExponent_decorators;
    let _blendExponent_initializers = [];
    let _blendExponent_extraInitializers = [];
    let _falloff_decorators;
    let _falloff_initializers = [];
    let _falloff_extraInitializers = [];
    let _bias_decorators;
    let _bias_initializers = [];
    let _bias_extraInitializers = [];
    return _a = class MRDLBackglowMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                /**
                 * Gets or sets the bevel radius on the backglow. If this value is changed, update the lineWidth to match.
                 */
                this.bevelRadius = __runInitializers(this, _bevelRadius_initializers, 0.16);
                /**
                 * Gets or sets the line width of the backglow.
                 */
                this.lineWidth = (__runInitializers(this, _bevelRadius_extraInitializers), __runInitializers(this, _lineWidth_initializers, 0.16));
                /**
                 * Gets or sets whether to use absolute sizes when calculating effects on the backglow.
                 * Since desktop and VR/AR have different relative sizes, it's usually best to keep this false.
                 */
                this.absoluteSizes = (__runInitializers(this, _lineWidth_extraInitializers), __runInitializers(this, _absoluteSizes_initializers, false));
                /**
                 * Gets or sets the tuning motion of the backglow.
                 */
                this.tuningMotion = (__runInitializers(this, _absoluteSizes_extraInitializers), __runInitializers(this, _tuningMotion_initializers, 0.0));
                /**
                 * Gets or sets the motion of the backglow.
                 */
                this.motion = (__runInitializers(this, _tuningMotion_extraInitializers), __runInitializers(this, _motion_initializers, 1.0));
                /**
                 * Gets or sets the maximum intensity of the backglow.
                 */
                this.maxIntensity = (__runInitializers(this, _motion_extraInitializers), __runInitializers(this, _maxIntensity_initializers, 0.7));
                /**
                 * Gets or sets the fade-in exponent of the intensity of the backglow.
                 */
                this.intensityFadeInExponent = (__runInitializers(this, _maxIntensity_extraInitializers), __runInitializers(this, _intensityFadeInExponent_initializers, 2.0));
                /**
                 * Gets or sets the start of the outer fuzz effect on the backglow.
                 */
                this.outerFuzzStart = (__runInitializers(this, _intensityFadeInExponent_extraInitializers), __runInitializers(this, _outerFuzzStart_initializers, 0.04));
                /**
                 * Gets or sets the end of the outer fuzz effect on the backglow.
                 */
                this.outerFuzzEnd = (__runInitializers(this, _outerFuzzStart_extraInitializers), __runInitializers(this, _outerFuzzEnd_initializers, 0.04));
                /**
                 * Gets or sets the color of the backglow.
                 */
                this.color = (__runInitializers(this, _outerFuzzEnd_extraInitializers), __runInitializers(this, _color_initializers, new Color4(0.682353, 0.698039, 1, 1)));
                /**
                 * Gets or sets the inner color of the backglow.
                 */
                this.innerColor = (__runInitializers(this, _color_extraInitializers), __runInitializers(this, _innerColor_initializers, new Color4(0.356863, 0.392157, 0.796078, 1)));
                /**
                 * Gets or sets the blend exponent of the backglow.
                 */
                this.blendExponent = (__runInitializers(this, _innerColor_extraInitializers), __runInitializers(this, _blendExponent_initializers, 1.5));
                /**
                 * Gets or sets the falloff of the backglow.
                 */
                this.falloff = (__runInitializers(this, _blendExponent_extraInitializers), __runInitializers(this, _falloff_initializers, 2.0));
                /**
                 * Gets or sets the bias of the backglow.
                 */
                this.bias = (__runInitializers(this, _falloff_extraInitializers), __runInitializers(this, _bias_initializers, 0.5));
                __runInitializers(this, _bias_extraInitializers);
                this.alphaMode = Constants.ALPHA_ADD;
                this.disableDepthWrite = true;
                this.backFaceCulling = false;
            }
            needAlphaBlending() {
                return true;
            }
            needAlphaTesting() {
                return false;
            }
            getAlphaTestTexture() {
                return null;
            }
            // Methods
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new MRDLBackglowMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, false, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "mrdlBackglow";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "worldView",
                        "worldViewProjection",
                        "view",
                        "projection",
                        "viewProjection",
                        "cameraPosition",
                        "_Bevel_Radius_",
                        "_Line_Width_",
                        "_Absolute_Sizes_",
                        "_Tuning_Motion_",
                        "_Motion_",
                        "_Max_Intensity_",
                        "_Intensity_Fade_In_Exponent_",
                        "_Outer_Fuzz_Start_",
                        "_Outer_Fuzz_End_",
                        "_Color_",
                        "_Inner_Color_",
                        "_Blend_Exponent_",
                        "_Falloff_",
                        "_Bias_",
                    ];
                    const samplers = [];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        extraInitializationsAsync: async () => {
                            await Promise.all([import("./shaders/mrdlBackglow.vertex.js"), import("./shaders/mrdlBackglow.fragment.js")]);
                        },
                    }, engine), defines);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const scene = this.getScene();
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", scene.getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", scene.activeCamera.position);
                // "Rounded Rectangle"
                this._activeEffect.setFloat("_Bevel_Radius_", this.bevelRadius);
                this._activeEffect.setFloat("_Line_Width_", this.lineWidth);
                this._activeEffect.setFloat("_Absolute_Sizes_", this.absoluteSizes ? 1.0 : 0.0);
                // "Animation"
                this._activeEffect.setFloat("_Tuning_Motion_", this.tuningMotion);
                this._activeEffect.setFloat("_Motion_", this.motion);
                this._activeEffect.setFloat("_Max_Intensity_", this.maxIntensity);
                this._activeEffect.setFloat("_Intensity_Fade_In_Exponent_", this.intensityFadeInExponent);
                this._activeEffect.setFloat("_Outer_Fuzz_Start_", this.outerFuzzStart);
                this._activeEffect.setFloat("_Outer_Fuzz_End_", this.outerFuzzEnd);
                // "Color"
                this._activeEffect.setDirectColor4("_Color_", this.color);
                this._activeEffect.setDirectColor4("_Inner_Color_", this.innerColor);
                this._activeEffect.setFloat("_Blend_Exponent_", this.blendExponent);
                // "Inner Transition"
                this._activeEffect.setFloat("_Falloff_", this.falloff);
                this._activeEffect.setFloat("_Bias_", this.bias);
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = SerializationHelper.Serialize(this);
                serializationObject.customType = "BABYLON.MRDLBackglowMaterial";
                return serializationObject;
            }
            getClassName() {
                return "MRDLBackglowMaterial";
            }
            // Statics
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _bevelRadius_decorators = [serialize()];
            _lineWidth_decorators = [serialize()];
            _absoluteSizes_decorators = [serialize()];
            _tuningMotion_decorators = [serialize()];
            _motion_decorators = [serialize()];
            _maxIntensity_decorators = [serialize()];
            _intensityFadeInExponent_decorators = [serialize()];
            _outerFuzzStart_decorators = [serialize()];
            _outerFuzzEnd_decorators = [serialize()];
            _color_decorators = [serialize()];
            _innerColor_decorators = [serialize()];
            _blendExponent_decorators = [serialize()];
            _falloff_decorators = [serialize()];
            _bias_decorators = [serialize()];
            __esDecorate(null, null, _bevelRadius_decorators, { kind: "field", name: "bevelRadius", static: false, private: false, access: { has: obj => "bevelRadius" in obj, get: obj => obj.bevelRadius, set: (obj, value) => { obj.bevelRadius = value; } }, metadata: _metadata }, _bevelRadius_initializers, _bevelRadius_extraInitializers);
            __esDecorate(null, null, _lineWidth_decorators, { kind: "field", name: "lineWidth", static: false, private: false, access: { has: obj => "lineWidth" in obj, get: obj => obj.lineWidth, set: (obj, value) => { obj.lineWidth = value; } }, metadata: _metadata }, _lineWidth_initializers, _lineWidth_extraInitializers);
            __esDecorate(null, null, _absoluteSizes_decorators, { kind: "field", name: "absoluteSizes", static: false, private: false, access: { has: obj => "absoluteSizes" in obj, get: obj => obj.absoluteSizes, set: (obj, value) => { obj.absoluteSizes = value; } }, metadata: _metadata }, _absoluteSizes_initializers, _absoluteSizes_extraInitializers);
            __esDecorate(null, null, _tuningMotion_decorators, { kind: "field", name: "tuningMotion", static: false, private: false, access: { has: obj => "tuningMotion" in obj, get: obj => obj.tuningMotion, set: (obj, value) => { obj.tuningMotion = value; } }, metadata: _metadata }, _tuningMotion_initializers, _tuningMotion_extraInitializers);
            __esDecorate(null, null, _motion_decorators, { kind: "field", name: "motion", static: false, private: false, access: { has: obj => "motion" in obj, get: obj => obj.motion, set: (obj, value) => { obj.motion = value; } }, metadata: _metadata }, _motion_initializers, _motion_extraInitializers);
            __esDecorate(null, null, _maxIntensity_decorators, { kind: "field", name: "maxIntensity", static: false, private: false, access: { has: obj => "maxIntensity" in obj, get: obj => obj.maxIntensity, set: (obj, value) => { obj.maxIntensity = value; } }, metadata: _metadata }, _maxIntensity_initializers, _maxIntensity_extraInitializers);
            __esDecorate(null, null, _intensityFadeInExponent_decorators, { kind: "field", name: "intensityFadeInExponent", static: false, private: false, access: { has: obj => "intensityFadeInExponent" in obj, get: obj => obj.intensityFadeInExponent, set: (obj, value) => { obj.intensityFadeInExponent = value; } }, metadata: _metadata }, _intensityFadeInExponent_initializers, _intensityFadeInExponent_extraInitializers);
            __esDecorate(null, null, _outerFuzzStart_decorators, { kind: "field", name: "outerFuzzStart", static: false, private: false, access: { has: obj => "outerFuzzStart" in obj, get: obj => obj.outerFuzzStart, set: (obj, value) => { obj.outerFuzzStart = value; } }, metadata: _metadata }, _outerFuzzStart_initializers, _outerFuzzStart_extraInitializers);
            __esDecorate(null, null, _outerFuzzEnd_decorators, { kind: "field", name: "outerFuzzEnd", static: false, private: false, access: { has: obj => "outerFuzzEnd" in obj, get: obj => obj.outerFuzzEnd, set: (obj, value) => { obj.outerFuzzEnd = value; } }, metadata: _metadata }, _outerFuzzEnd_initializers, _outerFuzzEnd_extraInitializers);
            __esDecorate(null, null, _color_decorators, { kind: "field", name: "color", static: false, private: false, access: { has: obj => "color" in obj, get: obj => obj.color, set: (obj, value) => { obj.color = value; } }, metadata: _metadata }, _color_initializers, _color_extraInitializers);
            __esDecorate(null, null, _innerColor_decorators, { kind: "field", name: "innerColor", static: false, private: false, access: { has: obj => "innerColor" in obj, get: obj => obj.innerColor, set: (obj, value) => { obj.innerColor = value; } }, metadata: _metadata }, _innerColor_initializers, _innerColor_extraInitializers);
            __esDecorate(null, null, _blendExponent_decorators, { kind: "field", name: "blendExponent", static: false, private: false, access: { has: obj => "blendExponent" in obj, get: obj => obj.blendExponent, set: (obj, value) => { obj.blendExponent = value; } }, metadata: _metadata }, _blendExponent_initializers, _blendExponent_extraInitializers);
            __esDecorate(null, null, _falloff_decorators, { kind: "field", name: "falloff", static: false, private: false, access: { has: obj => "falloff" in obj, get: obj => obj.falloff, set: (obj, value) => { obj.falloff = value; } }, metadata: _metadata }, _falloff_initializers, _falloff_extraInitializers);
            __esDecorate(null, null, _bias_decorators, { kind: "field", name: "bias", static: false, private: false, access: { has: obj => "bias" in obj, get: obj => obj.bias, set: (obj, value) => { obj.bias = value; } }, metadata: _metadata }, _bias_initializers, _bias_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
export { MRDLBackglowMaterial };
let _Registered = false;
/**
 * Registers the MRDLBackglowMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterMRDLBackglowMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.MRDLBackglowMaterial", MRDLBackglowMaterial);
}
