// @babylonjs/gui@9.16.1/3D/materials/mrdl/mrdlBackplateMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { Vector4 } from "@babylonjs/core/Maths/math.vector.pure.js";
import { Texture } from "@babylonjs/core/Materials/Textures/texture.pure.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
import { Tools } from "@babylonjs/core/Misc/tools.pure.js";
/** @internal */
class MRDLBackplateMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        /*
            "IRIDESCENCE_ENABLE", "SMOOTH_EDGES"
        */
        this.IRIDESCENCE_ENABLE = true;
        this.SMOOTH_EDGES = true;
        this._needNormals = true;
        this.rebuild();
    }
}
/**
 * Class used to render backplate material with MRDL
 */
let MRDLBackplateMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _radius_decorators;
    let _radius_initializers = [];
    let _radius_extraInitializers = [];
    let _lineWidth_decorators;
    let _lineWidth_initializers = [];
    let _lineWidth_extraInitializers = [];
    let _absoluteSizes_decorators;
    let _absoluteSizes_initializers = [];
    let _absoluteSizes_extraInitializers = [];
    let _baseColor_decorators;
    let _baseColor_initializers = [];
    let _baseColor_extraInitializers = [];
    let _lineColor_decorators;
    let _lineColor_initializers = [];
    let _lineColor_extraInitializers = [];
    let _radiusTopLeft_decorators;
    let _radiusTopLeft_initializers = [];
    let _radiusTopLeft_extraInitializers = [];
    let _radiusTopRight_decorators;
    let _radiusTopRight_initializers = [];
    let _radiusTopRight_extraInitializers = [];
    let _radiusBottomLeft_decorators;
    let _radiusBottomLeft_initializers = [];
    let _radiusBottomLeft_extraInitializers = [];
    let _radiusBottomRight_decorators;
    let _radiusBottomRight_initializers = [];
    let _radiusBottomRight_extraInitializers = [];
    let _highlightColor_decorators;
    let _highlightColor_initializers = [];
    let _highlightColor_extraInitializers = [];
    let _highlightWidth_decorators;
    let _highlightWidth_initializers = [];
    let _highlightWidth_extraInitializers = [];
    let _iridescenceIntensity_decorators;
    let _iridescenceIntensity_initializers = [];
    let _iridescenceIntensity_extraInitializers = [];
    let _iridescenceEdgeIntensity_decorators;
    let _iridescenceEdgeIntensity_initializers = [];
    let _iridescenceEdgeIntensity_extraInitializers = [];
    let _iridescenceTint_decorators;
    let _iridescenceTint_initializers = [];
    let _iridescenceTint_extraInitializers = [];
    let _fadeOut_decorators;
    let _fadeOut_initializers = [];
    let _fadeOut_extraInitializers = [];
    let _gradientColor_decorators;
    let _gradientColor_initializers = [];
    let _gradientColor_extraInitializers = [];
    let _topLeftGradientColor_decorators;
    let _topLeftGradientColor_initializers = [];
    let _topLeftGradientColor_extraInitializers = [];
    let _topRightGradientColor_decorators;
    let _topRightGradientColor_initializers = [];
    let _topRightGradientColor_extraInitializers = [];
    let _bottomLeftGradientColor_decorators;
    let _bottomLeftGradientColor_initializers = [];
    let _bottomLeftGradientColor_extraInitializers = [];
    let _bottomRightGradientColor_decorators;
    let _bottomRightGradientColor_initializers = [];
    let _bottomRightGradientColor_extraInitializers = [];
    let _edgeWidth_decorators;
    let _edgeWidth_initializers = [];
    let _edgeWidth_extraInitializers = [];
    let _edgePower_decorators;
    let _edgePower_initializers = [];
    let _edgePower_extraInitializers = [];
    let _edgeLineGradientBlend_decorators;
    let _edgeLineGradientBlend_initializers = [];
    let _edgeLineGradientBlend_extraInitializers = [];
    return _a = class MRDLBackplateMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                /**
                 * Gets or sets the corner radius on the backplate. If this value is changed, update the lineWidth to match.
                 */
                this.radius = __runInitializers(this, _radius_initializers, 0.3);
                /**
                 * Gets or sets the line width of the backplate.
                 */
                this.lineWidth = (__runInitializers(this, _radius_extraInitializers), __runInitializers(this, _lineWidth_initializers, 0.003));
                /**
                 * Gets or sets whether to use absolute sizes when calculating effects on the backplate.
                 * Since desktop and VR/AR have different relative sizes, it's usually best to keep this false.
                 */
                this.absoluteSizes = (__runInitializers(this, _lineWidth_extraInitializers), __runInitializers(this, _absoluteSizes_initializers, false));
                /** @internal */
                this._filterWidth = (__runInitializers(this, _absoluteSizes_extraInitializers), 1);
                /**
                 * Gets or sets the base color of the backplate.
                 */
                this.baseColor = __runInitializers(this, _baseColor_initializers, new Color4(0, 0, 0, 1));
                /**
                 * Gets or sets the line color of the backplate.
                 */
                this.lineColor = (__runInitializers(this, _baseColor_extraInitializers), __runInitializers(this, _lineColor_initializers, new Color4(0.2, 0.262745, 0.4, 1)));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusTopLeft = (__runInitializers(this, _lineColor_extraInitializers), __runInitializers(this, _radiusTopLeft_initializers, 1.0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusTopRight = (__runInitializers(this, _radiusTopLeft_extraInitializers), __runInitializers(this, _radiusTopRight_initializers, 1.0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusBottomLeft = (__runInitializers(this, _radiusTopRight_extraInitializers), __runInitializers(this, _radiusBottomLeft_initializers, 1.0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusBottomRight = (__runInitializers(this, _radiusBottomLeft_extraInitializers), __runInitializers(this, _radiusBottomRight_initializers, 1.0));
                /** @internal */
                this._rate = (__runInitializers(this, _radiusBottomRight_extraInitializers), 0);
                /**
                 * Gets or sets the color of the highlights on the backplate line.
                 */
                this.highlightColor = __runInitializers(this, _highlightColor_initializers, new Color4(0.239216, 0.435294, 0.827451, 1));
                /**
                 * Gets or sets the width of the highlights on the backplate line.
                 */
                this.highlightWidth = (__runInitializers(this, _highlightColor_extraInitializers), __runInitializers(this, _highlightWidth_initializers, 0));
                /** @internal */
                this._highlightTransform = (__runInitializers(this, _highlightWidth_extraInitializers), new Vector4(1, 1, 0, 0));
                /** @internal */
                this._highlight = 1;
                /**
                 * Gets or sets the intensity of the iridescence effect.
                 */
                this.iridescenceIntensity = __runInitializers(this, _iridescenceIntensity_initializers, 0.45);
                /**
                 * Gets or sets the intensity of the iridescence effect on the backplate edges.
                 */
                this.iridescenceEdgeIntensity = (__runInitializers(this, _iridescenceIntensity_extraInitializers), __runInitializers(this, _iridescenceEdgeIntensity_initializers, 1));
                /**
                 * Gets or sets the Tint of the iridescence effect on the backplate.
                 */
                this.iridescenceTint = (__runInitializers(this, _iridescenceEdgeIntensity_extraInitializers), __runInitializers(this, _iridescenceTint_initializers, new Color4(1, 1, 1, 1)));
                /** @internal */
                this._angle = (__runInitializers(this, _iridescenceTint_extraInitializers), -45);
                /**
                 * Gets or sets the opacity of the backplate (0.0 - 1.0).
                 */
                this.fadeOut = __runInitializers(this, _fadeOut_initializers, 1);
                /** @internal */
                this._reflected = (__runInitializers(this, _fadeOut_extraInitializers), true);
                /** @internal */
                this._frequency = 1;
                /** @internal */
                this._verticalOffset = 0;
                /**
                 * Gets or sets the gradient color effect on the backplate.
                 */
                this.gradientColor = __runInitializers(this, _gradientColor_initializers, new Color4(0.74902, 0.74902, 0.74902, 1));
                /**
                 * Gets or sets the top left gradient color effect on the backplate.
                 */
                this.topLeftGradientColor = (__runInitializers(this, _gradientColor_extraInitializers), __runInitializers(this, _topLeftGradientColor_initializers, new Color4(0.00784314, 0.294118, 0.580392, 1)));
                /**
                 * Gets or sets the top right gradient color effect on the backplate.
                 */
                this.topRightGradientColor = (__runInitializers(this, _topLeftGradientColor_extraInitializers), __runInitializers(this, _topRightGradientColor_initializers, new Color4(0.305882, 0, 1, 1)));
                /**
                 * Gets or sets the bottom left gradient color effect on the backplate.
                 */
                this.bottomLeftGradientColor = (__runInitializers(this, _topRightGradientColor_extraInitializers), __runInitializers(this, _bottomLeftGradientColor_initializers, new Color4(0.133333, 0.258824, 0.992157, 1)));
                /**
                 * Gets or sets the bottom right gradient color effect on the backplate.
                 */
                this.bottomRightGradientColor = (__runInitializers(this, _bottomLeftGradientColor_extraInitializers), __runInitializers(this, _bottomRightGradientColor_initializers, new Color4(0.176471, 0.176471, 0.619608, 1)));
                /**
                 * Gets or sets the edge width of the backplate.
                 */
                this.edgeWidth = (__runInitializers(this, _bottomRightGradientColor_extraInitializers), __runInitializers(this, _edgeWidth_initializers, 0.5));
                /**
                 * Gets or sets the edge width of the backplate.
                 */
                this.edgePower = (__runInitializers(this, _edgeWidth_extraInitializers), __runInitializers(this, _edgePower_initializers, 1));
                /**
                 * Gets or sets the edge width of the backplate.
                 */
                this.edgeLineGradientBlend = (__runInitializers(this, _edgePower_extraInitializers), __runInitializers(this, _edgeLineGradientBlend_initializers, 0.5));
                __runInitializers(this, _edgeLineGradientBlend_extraInitializers);
                this.alphaMode = Constants.ALPHA_DISABLE;
                this.backFaceCulling = false;
                const textureUrl = Tools.GetAssetUrl(_a.IRIDESCENT_MAP_TEXTURE_URL);
                this._iridescentMapTexture = new Texture(textureUrl, this.getScene(), true, false, Texture.NEAREST_SAMPLINGMODE);
            }
            needAlphaBlending() {
                return false;
            }
            needAlphaTesting() {
                return false;
            }
            getAlphaTestTexture() {
                return null;
            }
            // Methods
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new MRDLBackplateMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                if (!this._iridescentMapTexture.isReady()) {
                    return false;
                }
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, false, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "mrdlBackplate";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "viewProjection",
                        "cameraPosition",
                        "_Radius_",
                        "_Line_Width_",
                        "_Absolute_Sizes_",
                        "_Filter_Width_",
                        "_Base_Color_",
                        "_Line_Color_",
                        "_Radius_Top_Left_",
                        "_Radius_Top_Right_",
                        "_Radius_Bottom_Left_",
                        "_Radius_Bottom_Right_",
                        "_Rate_",
                        "_Highlight_Color_",
                        "_Highlight_Width_",
                        "_Highlight_Transform_",
                        "_Highlight_",
                        "_Iridescence_Intensity_",
                        "_Iridescence_Edge_Intensity_",
                        "_Iridescence_Tint_",
                        "_Iridescent_Map_",
                        "_Angle_",
                        "_Reflected_",
                        "_Frequency_",
                        "_Vertical_Offset_",
                        "_Gradient_Color_",
                        "_Top_Left_",
                        "_Top_Right_",
                        "_Bottom_Left_",
                        "_Bottom_Right_",
                        "_Edge_Width_",
                        "_Edge_Power_",
                        "_Line_Gradient_Blend_",
                        "_Fade_Out_",
                    ];
                    const samplers = ["_Iridescent_Map_"];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        extraInitializationsAsync: async () => {
                            await Promise.all([import("./shaders/mrdlBackplate.vertex.js"), import("./shaders/mrdlBackplate.fragment.js")]);
                        },
                    }, engine), defines);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", this.getScene().getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", this.getScene().activeCamera.position);
                // "Round Rect"
                this._activeEffect.setFloat("_Radius_", this.radius);
                this._activeEffect.setFloat("_Line_Width_", this.lineWidth);
                this._activeEffect.setFloat("_Absolute_Sizes_", this.absoluteSizes ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Filter_Width_", this._filterWidth);
                this._activeEffect.setDirectColor4("_Base_Color_", this.baseColor);
                this._activeEffect.setDirectColor4("_Line_Color_", this.lineColor);
                // "Radii Multipliers"
                this._activeEffect.setFloat("_Radius_Top_Left_", this.radiusTopLeft);
                this._activeEffect.setFloat("_Radius_Top_Right_", this.radiusTopRight);
                this._activeEffect.setFloat("_Radius_Bottom_Left_", this.radiusBottomLeft);
                this._activeEffect.setFloat("_Radius_Bottom_Right_", this.radiusBottomRight);
                // "Line Highlight"
                this._activeEffect.setFloat("_Rate_", this._rate);
                this._activeEffect.setDirectColor4("_Highlight_Color_", this.highlightColor);
                this._activeEffect.setFloat("_Highlight_Width_", this.highlightWidth);
                this._activeEffect.setVector4("_Highlight_Transform_", this._highlightTransform);
                this._activeEffect.setFloat("_Highlight_", this._highlight);
                // "Iridescence"
                //define IRIDESCENCE_ENABLE true;
                this._activeEffect.setFloat("_Iridescence_Intensity_", this.iridescenceIntensity);
                this._activeEffect.setFloat("_Iridescence_Edge_Intensity_", this.iridescenceEdgeIntensity);
                this._activeEffect.setDirectColor4("_Iridescence_Tint_", this.iridescenceTint);
                this._activeEffect.setTexture("_Iridescent_Map_", this._iridescentMapTexture);
                this._activeEffect.setFloat("_Angle_", this._angle);
                this._activeEffect.setFloat("_Reflected_", this._reflected ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Frequency_", this._frequency);
                this._activeEffect.setFloat("_Vertical_Offset_", this._verticalOffset);
                // "Gradient"
                this._activeEffect.setDirectColor4("_Gradient_Color_", this.gradientColor);
                this._activeEffect.setDirectColor4("_Top_Left_", this.topLeftGradientColor);
                this._activeEffect.setDirectColor4("_Top_Right_", this.topRightGradientColor);
                this._activeEffect.setDirectColor4("_Bottom_Left_", this.bottomLeftGradientColor);
                this._activeEffect.setDirectColor4("_Bottom_Right_", this.bottomRightGradientColor);
                //define EDGE_ONLY false;
                this._activeEffect.setFloat("_Edge_Width_", this.edgeWidth);
                this._activeEffect.setFloat("_Edge_Power_", this.edgePower);
                this._activeEffect.setFloat("_Line_Gradient_Blend_", this.edgeLineGradientBlend);
                // "Fade"
                this._activeEffect.setFloat("_Fade_Out_", this.fadeOut);
                // "Antialiasing"
                //define SMOOTH_EDGES true;
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = super.serialize();
                serializationObject.customType = "BABYLON.MRDLBackplateMaterial";
                return serializationObject;
            }
            getClassName() {
                return "MRDLBackplateMaterial";
            }
            // Statics
            /**
             * Parses a serialized MRDLBackplateMaterial and returns a new instance.
             * @param source - the serialized object
             * @param scene - defines the hosting scene
             * @param rootUrl - the root URL used to load any associated textures
             * @returns a new MRDLBackplateMaterial
             */
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _radius_decorators = [serialize()];
            _lineWidth_decorators = [serialize()];
            _absoluteSizes_decorators = [serialize()];
            _baseColor_decorators = [serialize()];
            _lineColor_decorators = [serialize()];
            _radiusTopLeft_decorators = [serialize()];
            _radiusTopRight_decorators = [serialize()];
            _radiusBottomLeft_decorators = [serialize()];
            _radiusBottomRight_decorators = [serialize()];
            _highlightColor_decorators = [serialize()];
            _highlightWidth_decorators = [serialize()];
            _iridescenceIntensity_decorators = [serialize()];
            _iridescenceEdgeIntensity_decorators = [serialize()];
            _iridescenceTint_decorators = [serialize()];
            _fadeOut_decorators = [serialize()];
            _gradientColor_decorators = [serialize()];
            _topLeftGradientColor_decorators = [serialize()];
            _topRightGradientColor_decorators = [serialize()];
            _bottomLeftGradientColor_decorators = [serialize()];
            _bottomRightGradientColor_decorators = [serialize()];
            _edgeWidth_decorators = [serialize()];
            _edgePower_decorators = [serialize()];
            _edgeLineGradientBlend_decorators = [serialize()];
            __esDecorate(null, null, _radius_decorators, { kind: "field", name: "radius", static: false, private: false, access: { has: obj => "radius" in obj, get: obj => obj.radius, set: (obj, value) => { obj.radius = value; } }, metadata: _metadata }, _radius_initializers, _radius_extraInitializers);
            __esDecorate(null, null, _lineWidth_decorators, { kind: "field", name: "lineWidth", static: false, private: false, access: { has: obj => "lineWidth" in obj, get: obj => obj.lineWidth, set: (obj, value) => { obj.lineWidth = value; } }, metadata: _metadata }, _lineWidth_initializers, _lineWidth_extraInitializers);
            __esDecorate(null, null, _absoluteSizes_decorators, { kind: "field", name: "absoluteSizes", static: false, private: false, access: { has: obj => "absoluteSizes" in obj, get: obj => obj.absoluteSizes, set: (obj, value) => { obj.absoluteSizes = value; } }, metadata: _metadata }, _absoluteSizes_initializers, _absoluteSizes_extraInitializers);
            __esDecorate(null, null, _baseColor_decorators, { kind: "field", name: "baseColor", static: false, private: false, access: { has: obj => "baseColor" in obj, get: obj => obj.baseColor, set: (obj, value) => { obj.baseColor = value; } }, metadata: _metadata }, _baseColor_initializers, _baseColor_extraInitializers);
            __esDecorate(null, null, _lineColor_decorators, { kind: "field", name: "lineColor", static: false, private: false, access: { has: obj => "lineColor" in obj, get: obj => obj.lineColor, set: (obj, value) => { obj.lineColor = value; } }, metadata: _metadata }, _lineColor_initializers, _lineColor_extraInitializers);
            __esDecorate(null, null, _radiusTopLeft_decorators, { kind: "field", name: "radiusTopLeft", static: false, private: false, access: { has: obj => "radiusTopLeft" in obj, get: obj => obj.radiusTopLeft, set: (obj, value) => { obj.radiusTopLeft = value; } }, metadata: _metadata }, _radiusTopLeft_initializers, _radiusTopLeft_extraInitializers);
            __esDecorate(null, null, _radiusTopRight_decorators, { kind: "field", name: "radiusTopRight", static: false, private: false, access: { has: obj => "radiusTopRight" in obj, get: obj => obj.radiusTopRight, set: (obj, value) => { obj.radiusTopRight = value; } }, metadata: _metadata }, _radiusTopRight_initializers, _radiusTopRight_extraInitializers);
            __esDecorate(null, null, _radiusBottomLeft_decorators, { kind: "field", name: "radiusBottomLeft", static: false, private: false, access: { has: obj => "radiusBottomLeft" in obj, get: obj => obj.radiusBottomLeft, set: (obj, value) => { obj.radiusBottomLeft = value; } }, metadata: _metadata }, _radiusBottomLeft_initializers, _radiusBottomLeft_extraInitializers);
            __esDecorate(null, null, _radiusBottomRight_decorators, { kind: "field", name: "radiusBottomRight", static: false, private: false, access: { has: obj => "radiusBottomRight" in obj, get: obj => obj.radiusBottomRight, set: (obj, value) => { obj.radiusBottomRight = value; } }, metadata: _metadata }, _radiusBottomRight_initializers, _radiusBottomRight_extraInitializers);
            __esDecorate(null, null, _highlightColor_decorators, { kind: "field", name: "highlightColor", static: false, private: false, access: { has: obj => "highlightColor" in obj, get: obj => obj.highlightColor, set: (obj, value) => { obj.highlightColor = value; } }, metadata: _metadata }, _highlightColor_initializers, _highlightColor_extraInitializers);
            __esDecorate(null, null, _highlightWidth_decorators, { kind: "field", name: "highlightWidth", static: false, private: false, access: { has: obj => "highlightWidth" in obj, get: obj => obj.highlightWidth, set: (obj, value) => { obj.highlightWidth = value; } }, metadata: _metadata }, _highlightWidth_initializers, _highlightWidth_extraInitializers);
            __esDecorate(null, null, _iridescenceIntensity_decorators, { kind: "field", name: "iridescenceIntensity", static: false, private: false, access: { has: obj => "iridescenceIntensity" in obj, get: obj => obj.iridescenceIntensity, set: (obj, value) => { obj.iridescenceIntensity = value; } }, metadata: _metadata }, _iridescenceIntensity_initializers, _iridescenceIntensity_extraInitializers);
            __esDecorate(null, null, _iridescenceEdgeIntensity_decorators, { kind: "field", name: "iridescenceEdgeIntensity", static: false, private: false, access: { has: obj => "iridescenceEdgeIntensity" in obj, get: obj => obj.iridescenceEdgeIntensity, set: (obj, value) => { obj.iridescenceEdgeIntensity = value; } }, metadata: _metadata }, _iridescenceEdgeIntensity_initializers, _iridescenceEdgeIntensity_extraInitializers);
            __esDecorate(null, null, _iridescenceTint_decorators, { kind: "field", name: "iridescenceTint", static: false, private: false, access: { has: obj => "iridescenceTint" in obj, get: obj => obj.iridescenceTint, set: (obj, value) => { obj.iridescenceTint = value; } }, metadata: _metadata }, _iridescenceTint_initializers, _iridescenceTint_extraInitializers);
            __esDecorate(null, null, _fadeOut_decorators, { kind: "field", name: "fadeOut", static: false, private: false, access: { has: obj => "fadeOut" in obj, get: obj => obj.fadeOut, set: (obj, value) => { obj.fadeOut = value; } }, metadata: _metadata }, _fadeOut_initializers, _fadeOut_extraInitializers);
            __esDecorate(null, null, _gradientColor_decorators, { kind: "field", name: "gradientColor", static: false, private: false, access: { has: obj => "gradientColor" in obj, get: obj => obj.gradientColor, set: (obj, value) => { obj.gradientColor = value; } }, metadata: _metadata }, _gradientColor_initializers, _gradientColor_extraInitializers);
            __esDecorate(null, null, _topLeftGradientColor_decorators, { kind: "field", name: "topLeftGradientColor", static: false, private: false, access: { has: obj => "topLeftGradientColor" in obj, get: obj => obj.topLeftGradientColor, set: (obj, value) => { obj.topLeftGradientColor = value; } }, metadata: _metadata }, _topLeftGradientColor_initializers, _topLeftGradientColor_extraInitializers);
            __esDecorate(null, null, _topRightGradientColor_decorators, { kind: "field", name: "topRightGradientColor", static: false, private: false, access: { has: obj => "topRightGradientColor" in obj, get: obj => obj.topRightGradientColor, set: (obj, value) => { obj.topRightGradientColor = value; } }, metadata: _metadata }, _topRightGradientColor_initializers, _topRightGradientColor_extraInitializers);
            __esDecorate(null, null, _bottomLeftGradientColor_decorators, { kind: "field", name: "bottomLeftGradientColor", static: false, private: false, access: { has: obj => "bottomLeftGradientColor" in obj, get: obj => obj.bottomLeftGradientColor, set: (obj, value) => { obj.bottomLeftGradientColor = value; } }, metadata: _metadata }, _bottomLeftGradientColor_initializers, _bottomLeftGradientColor_extraInitializers);
            __esDecorate(null, null, _bottomRightGradientColor_decorators, { kind: "field", name: "bottomRightGradientColor", static: false, private: false, access: { has: obj => "bottomRightGradientColor" in obj, get: obj => obj.bottomRightGradientColor, set: (obj, value) => { obj.bottomRightGradientColor = value; } }, metadata: _metadata }, _bottomRightGradientColor_initializers, _bottomRightGradientColor_extraInitializers);
            __esDecorate(null, null, _edgeWidth_decorators, { kind: "field", name: "edgeWidth", static: false, private: false, access: { has: obj => "edgeWidth" in obj, get: obj => obj.edgeWidth, set: (obj, value) => { obj.edgeWidth = value; } }, metadata: _metadata }, _edgeWidth_initializers, _edgeWidth_extraInitializers);
            __esDecorate(null, null, _edgePower_decorators, { kind: "field", name: "edgePower", static: false, private: false, access: { has: obj => "edgePower" in obj, get: obj => obj.edgePower, set: (obj, value) => { obj.edgePower = value; } }, metadata: _metadata }, _edgePower_initializers, _edgePower_extraInitializers);
            __esDecorate(null, null, _edgeLineGradientBlend_decorators, { kind: "field", name: "edgeLineGradientBlend", static: false, private: false, access: { has: obj => "edgeLineGradientBlend" in obj, get: obj => obj.edgeLineGradientBlend, set: (obj, value) => { obj.edgeLineGradientBlend = value; } }, metadata: _metadata }, _edgeLineGradientBlend_initializers, _edgeLineGradientBlend_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        /**
         * URL pointing to the texture used to define the coloring for the Iridescent Map effect.
         */
        _a.IRIDESCENT_MAP_TEXTURE_URL = "https://assets.babylonjs.com/core/MRTK/MRDL/mrtk-mrdl-backplate-iridescence.png",
        _a;
})();
export { MRDLBackplateMaterial };
let _Registered = false;
/**
 * Registers the MRDLBackplateMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterMRDLBackplateMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.MRDLBackplateMaterial", MRDLBackplateMaterial);
}
