// @babylonjs/gui@9.16.1/3D/materials/mrdl/mrdlSliderThumbMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { Vector2, Vector3, Vector4 } from "@babylonjs/core/Maths/math.vector.pure.js";
import { Texture } from "@babylonjs/core/Materials/Textures/texture.pure.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
import { Tools } from "@babylonjs/core/Misc/tools.pure.js";
/** @internal */
class MRDLSliderThumbMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        /*
            "SKY_ENABLED", "BLOB_ENABLE_2", "IRIDESCENCE_ENABLED"
        */
        this.SKY_ENABLED = true;
        this.BLOB_ENABLE_2 = true;
        this.IRIDESCENCE_ENABLED = true;
        this._needNormals = true;
        this._needUVs = true;
        this.rebuild();
    }
}
/**
 * Class used to render Slider Thumb material with MRDL
 */
let MRDLSliderThumbMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _radius_decorators;
    let _radius_initializers = [];
    let _radius_extraInitializers = [];
    let _bevelFront_decorators;
    let _bevelFront_initializers = [];
    let _bevelFront_extraInitializers = [];
    let _bevelFrontStretch_decorators;
    let _bevelFrontStretch_initializers = [];
    let _bevelFrontStretch_extraInitializers = [];
    let _bevelBack_decorators;
    let _bevelBack_initializers = [];
    let _bevelBack_extraInitializers = [];
    let _bevelBackStretch_decorators;
    let _bevelBackStretch_initializers = [];
    let _bevelBackStretch_extraInitializers = [];
    let _radiusTopLeft_decorators;
    let _radiusTopLeft_initializers = [];
    let _radiusTopLeft_extraInitializers = [];
    let _radiusTopRight_decorators;
    let _radiusTopRight_initializers = [];
    let _radiusTopRight_extraInitializers = [];
    let _radiusBottomLeft_decorators;
    let _radiusBottomLeft_initializers = [];
    let _radiusBottomLeft_extraInitializers = [];
    let _radiusBottomRight_decorators;
    let _radiusBottomRight_initializers = [];
    let _radiusBottomRight_extraInitializers = [];
    let _bulgeEnabled_decorators;
    let _bulgeEnabled_initializers = [];
    let _bulgeEnabled_extraInitializers = [];
    let _bulgeHeight_decorators;
    let _bulgeHeight_initializers = [];
    let _bulgeHeight_extraInitializers = [];
    let _bulgeRadius_decorators;
    let _bulgeRadius_initializers = [];
    let _bulgeRadius_extraInitializers = [];
    let _sunIntensity_decorators;
    let _sunIntensity_initializers = [];
    let _sunIntensity_extraInitializers = [];
    let _sunTheta_decorators;
    let _sunTheta_initializers = [];
    let _sunTheta_extraInitializers = [];
    let _sunPhi_decorators;
    let _sunPhi_initializers = [];
    let _sunPhi_extraInitializers = [];
    let _indirectDiffuse_decorators;
    let _indirectDiffuse_initializers = [];
    let _indirectDiffuse_extraInitializers = [];
    let _albedo_decorators;
    let _albedo_initializers = [];
    let _albedo_extraInitializers = [];
    let _specular_decorators;
    let _specular_initializers = [];
    let _specular_extraInitializers = [];
    let _shininess_decorators;
    let _shininess_initializers = [];
    let _shininess_extraInitializers = [];
    let _sharpness_decorators;
    let _sharpness_initializers = [];
    let _sharpness_extraInitializers = [];
    let _subsurface_decorators;
    let _subsurface_initializers = [];
    let _subsurface_extraInitializers = [];
    let _leftGradientColor_decorators;
    let _leftGradientColor_initializers = [];
    let _leftGradientColor_extraInitializers = [];
    let _rightGradientColor_decorators;
    let _rightGradientColor_initializers = [];
    let _rightGradientColor_extraInitializers = [];
    let _reflection_decorators;
    let _reflection_initializers = [];
    let _reflection_extraInitializers = [];
    let _frontReflect_decorators;
    let _frontReflect_initializers = [];
    let _frontReflect_extraInitializers = [];
    let _edgeReflect_decorators;
    let _edgeReflect_initializers = [];
    let _edgeReflect_extraInitializers = [];
    let _power_decorators;
    let _power_initializers = [];
    let _power_extraInitializers = [];
    let _skyColor_decorators;
    let _skyColor_initializers = [];
    let _skyColor_extraInitializers = [];
    let _horizonColor_decorators;
    let _horizonColor_initializers = [];
    let _horizonColor_extraInitializers = [];
    let _groundColor_decorators;
    let _groundColor_initializers = [];
    let _groundColor_extraInitializers = [];
    let _horizonPower_decorators;
    let _horizonPower_initializers = [];
    let _horizonPower_extraInitializers = [];
    let _width_decorators;
    let _width_initializers = [];
    let _width_extraInitializers = [];
    let _fuzz_decorators;
    let _fuzz_initializers = [];
    let _fuzz_extraInitializers = [];
    let _minFuzz_decorators;
    let _minFuzz_initializers = [];
    let _minFuzz_extraInitializers = [];
    let _clipFade_decorators;
    let _clipFade_initializers = [];
    let _clipFade_extraInitializers = [];
    let _hueShift_decorators;
    let _hueShift_initializers = [];
    let _hueShift_extraInitializers = [];
    let _saturationShift_decorators;
    let _saturationShift_initializers = [];
    let _saturationShift_extraInitializers = [];
    let _valueShift_decorators;
    let _valueShift_initializers = [];
    let _valueShift_extraInitializers = [];
    let _blobPosition_decorators;
    let _blobPosition_initializers = [];
    let _blobPosition_extraInitializers = [];
    let _blobIntensity_decorators;
    let _blobIntensity_initializers = [];
    let _blobIntensity_extraInitializers = [];
    let _blobNearSize_decorators;
    let _blobNearSize_initializers = [];
    let _blobNearSize_extraInitializers = [];
    let _blobFarSize_decorators;
    let _blobFarSize_initializers = [];
    let _blobFarSize_extraInitializers = [];
    let _blobNearDistance_decorators;
    let _blobNearDistance_initializers = [];
    let _blobNearDistance_extraInitializers = [];
    let _blobFarDistance_decorators;
    let _blobFarDistance_initializers = [];
    let _blobFarDistance_extraInitializers = [];
    let _blobFadeLength_decorators;
    let _blobFadeLength_initializers = [];
    let _blobFadeLength_extraInitializers = [];
    let _blobPulse_decorators;
    let _blobPulse_initializers = [];
    let _blobPulse_extraInitializers = [];
    let _blobFade_decorators;
    let _blobFade_initializers = [];
    let _blobFade_extraInitializers = [];
    let _blobPosition2_decorators;
    let _blobPosition2_initializers = [];
    let _blobPosition2_extraInitializers = [];
    let _blobNearSize2_decorators;
    let _blobNearSize2_initializers = [];
    let _blobNearSize2_extraInitializers = [];
    let _blobPulse2_decorators;
    let _blobPulse2_initializers = [];
    let _blobPulse2_extraInitializers = [];
    let _blobFade2_decorators;
    let _blobFade2_initializers = [];
    let _blobFade2_extraInitializers = [];
    let _blobTexture_decorators;
    let _blobTexture_initializers = [];
    let _blobTexture_extraInitializers = [];
    let _leftIndexPosition_decorators;
    let _leftIndexPosition_initializers = [];
    let _leftIndexPosition_extraInitializers = [];
    let _rightIndexPosition_decorators;
    let _rightIndexPosition_initializers = [];
    let _rightIndexPosition_extraInitializers = [];
    let _leftIndexMiddlePosition_decorators;
    let _leftIndexMiddlePosition_initializers = [];
    let _leftIndexMiddlePosition_extraInitializers = [];
    let _rightIndexMiddlePosition_decorators;
    let _rightIndexMiddlePosition_initializers = [];
    let _rightIndexMiddlePosition_extraInitializers = [];
    let _decalScaleXY_decorators;
    let _decalScaleXY_initializers = [];
    let _decalScaleXY_extraInitializers = [];
    let _decalFrontOnly_decorators;
    let _decalFrontOnly_initializers = [];
    let _decalFrontOnly_extraInitializers = [];
    let _rimIntensity_decorators;
    let _rimIntensity_initializers = [];
    let _rimIntensity_extraInitializers = [];
    let _rimHueShift_decorators;
    let _rimHueShift_initializers = [];
    let _rimHueShift_extraInitializers = [];
    let _rimSaturationShift_decorators;
    let _rimSaturationShift_initializers = [];
    let _rimSaturationShift_extraInitializers = [];
    let _rimValueShift_decorators;
    let _rimValueShift_initializers = [];
    let _rimValueShift_extraInitializers = [];
    let _iridescenceIntensity_decorators;
    let _iridescenceIntensity_initializers = [];
    let _iridescenceIntensity_extraInitializers = [];
    return _a = class MRDLSliderThumbMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                /**
                 * Gets or sets the corner Radius on the slider thumb.
                 */
                this.radius = __runInitializers(this, _radius_initializers, 0.157);
                /**
                 * Gets or sets the Bevel Front on the slider thumb.
                 */
                this.bevelFront = (__runInitializers(this, _radius_extraInitializers), __runInitializers(this, _bevelFront_initializers, 0.065));
                /**
                 * Gets or sets the Bevel Front Stretch on the slider thumb.
                 */
                this.bevelFrontStretch = (__runInitializers(this, _bevelFront_extraInitializers), __runInitializers(this, _bevelFrontStretch_initializers, 0.077));
                /**
                 * Gets or sets the Bevel Back on the slider thumb.
                 */
                this.bevelBack = (__runInitializers(this, _bevelFrontStretch_extraInitializers), __runInitializers(this, _bevelBack_initializers, 0.031));
                /**
                 * Gets or sets the Bevel Back Stretch on the slider thumb.
                 */
                this.bevelBackStretch = (__runInitializers(this, _bevelBack_extraInitializers), __runInitializers(this, _bevelBackStretch_initializers, 0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusTopLeft = (__runInitializers(this, _bevelBackStretch_extraInitializers), __runInitializers(this, _radiusTopLeft_initializers, 1.0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusTopRight = (__runInitializers(this, _radiusTopLeft_extraInitializers), __runInitializers(this, _radiusTopRight_initializers, 1.0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusBottomLeft = (__runInitializers(this, _radiusTopRight_extraInitializers), __runInitializers(this, _radiusBottomLeft_initializers, 1.0));
                /**
                 * Gets or sets the top left Radii Multiplier.
                 */
                this.radiusBottomRight = (__runInitializers(this, _radiusBottomLeft_extraInitializers), __runInitializers(this, _radiusBottomRight_initializers, 1.0));
                /**
                 * Gets or sets whether Bulge is enabled.
                 * Default is false.
                 */
                this.bulgeEnabled = (__runInitializers(this, _radiusBottomRight_extraInitializers), __runInitializers(this, _bulgeEnabled_initializers, false));
                /**
                 * Gets or sets the Bulge Height.
                 */
                this.bulgeHeight = (__runInitializers(this, _bulgeEnabled_extraInitializers), __runInitializers(this, _bulgeHeight_initializers, -0.323));
                /**
                 * Gets or sets the Bulge Radius.
                 */
                this.bulgeRadius = (__runInitializers(this, _bulgeHeight_extraInitializers), __runInitializers(this, _bulgeRadius_initializers, 0.73));
                /**
                 * Gets or sets the Sun Intensity.
                 */
                this.sunIntensity = (__runInitializers(this, _bulgeRadius_extraInitializers), __runInitializers(this, _sunIntensity_initializers, 2));
                /**
                 * Gets or sets the Sun Theta.
                 */
                this.sunTheta = (__runInitializers(this, _sunIntensity_extraInitializers), __runInitializers(this, _sunTheta_initializers, 0.937));
                /**
                 * Gets or sets the Sun Phi.
                 */
                this.sunPhi = (__runInitializers(this, _sunTheta_extraInitializers), __runInitializers(this, _sunPhi_initializers, 0.555));
                /**
                 * Gets or sets the Indirect Diffuse.
                 */
                this.indirectDiffuse = (__runInitializers(this, _sunPhi_extraInitializers), __runInitializers(this, _indirectDiffuse_initializers, 1));
                /**
                 * Gets or sets the base albedo.
                 */
                this.albedo = (__runInitializers(this, _indirectDiffuse_extraInitializers), __runInitializers(this, _albedo_initializers, new Color4(0.0117647, 0.505882, 0.996078, 1)));
                /**
                 * Gets or sets the Specular value.
                 */
                this.specular = (__runInitializers(this, _albedo_extraInitializers), __runInitializers(this, _specular_initializers, 0));
                /**
                 * Gets or sets the Shininess value.
                 */
                this.shininess = (__runInitializers(this, _specular_extraInitializers), __runInitializers(this, _shininess_initializers, 10));
                /**
                 * Gets or sets the Sharpness value.
                 */
                this.sharpness = (__runInitializers(this, _shininess_extraInitializers), __runInitializers(this, _sharpness_initializers, 0));
                /**
                 * Gets or sets the Subsurface value.
                 */
                this.subsurface = (__runInitializers(this, _sharpness_extraInitializers), __runInitializers(this, _subsurface_initializers, 0.31));
                /**
                 * Gets or sets the left gradient color.
                 */
                this.leftGradientColor = (__runInitializers(this, _subsurface_extraInitializers), __runInitializers(this, _leftGradientColor_initializers, new Color4(0.0117647, 0.505882, 0.996078, 1)));
                /**
                 * Gets or sets the right gradient color.
                 */
                this.rightGradientColor = (__runInitializers(this, _leftGradientColor_extraInitializers), __runInitializers(this, _rightGradientColor_initializers, new Color4(0.0117647, 0.505882, 0.996078, 1)));
                /**
                 * Gets or sets the reflection value.
                 */
                this.reflection = (__runInitializers(this, _rightGradientColor_extraInitializers), __runInitializers(this, _reflection_initializers, 0.749));
                /**
                 * Gets or sets the front reflect value.
                 */
                this.frontReflect = (__runInitializers(this, _reflection_extraInitializers), __runInitializers(this, _frontReflect_initializers, 0));
                /**
                 * Gets or sets the edge reflect value.
                 */
                this.edgeReflect = (__runInitializers(this, _frontReflect_extraInitializers), __runInitializers(this, _edgeReflect_initializers, 0.09));
                /**
                 * Gets or sets the power value.
                 */
                this.power = (__runInitializers(this, _edgeReflect_extraInitializers), __runInitializers(this, _power_initializers, 8.1));
                /**
                 * Gets or sets the sky color.
                 */
                this.skyColor = (__runInitializers(this, _power_extraInitializers), __runInitializers(this, _skyColor_initializers, new Color4(0.0117647, 0.960784, 0.996078, 1)));
                /**
                 * Gets or sets the horizon color.
                 */
                this.horizonColor = (__runInitializers(this, _skyColor_extraInitializers), __runInitializers(this, _horizonColor_initializers, new Color4(0.0117647, 0.333333, 0.996078, 1)));
                /**
                 * Gets or sets the ground color.
                 */
                this.groundColor = (__runInitializers(this, _horizonColor_extraInitializers), __runInitializers(this, _groundColor_initializers, new Color4(0, 0.254902, 0.996078, 1)));
                /**
                 * Gets or sets the horizon power value.
                 */
                this.horizonPower = (__runInitializers(this, _groundColor_extraInitializers), __runInitializers(this, _horizonPower_initializers, 1));
                /**
                 * Gets or sets the finger occlusion width value.
                 */
                this.width = (__runInitializers(this, _horizonPower_extraInitializers), __runInitializers(this, _width_initializers, 0.02));
                /**
                 * Gets or sets the finger occlusion fuzz value.
                 */
                this.fuzz = (__runInitializers(this, _width_extraInitializers), __runInitializers(this, _fuzz_initializers, 0.5));
                /**
                 * Gets or sets the minimum finger occlusion fuzz value.
                 */
                this.minFuzz = (__runInitializers(this, _fuzz_extraInitializers), __runInitializers(this, _minFuzz_initializers, 0.001));
                /**
                 * Gets or sets the finger occlusion clip fade value.
                 */
                this.clipFade = (__runInitializers(this, _minFuzz_extraInitializers), __runInitializers(this, _clipFade_initializers, 0.01));
                /**
                 * Gets or sets the hue shift value.
                 */
                this.hueShift = (__runInitializers(this, _clipFade_extraInitializers), __runInitializers(this, _hueShift_initializers, 0));
                /**
                 * Gets or sets the saturation shift value.
                 */
                this.saturationShift = (__runInitializers(this, _hueShift_extraInitializers), __runInitializers(this, _saturationShift_initializers, 0));
                /**
                 * Gets or sets the value shift.
                 */
                this.valueShift = (__runInitializers(this, _saturationShift_extraInitializers), __runInitializers(this, _valueShift_initializers, 0));
                /**
                 * Gets or sets the position of the hover glow effect.
                 */
                this.blobPosition = (__runInitializers(this, _valueShift_extraInitializers), __runInitializers(this, _blobPosition_initializers, new Vector3(0, 0, 0.1)));
                /**
                 * Gets or sets the intensity of the hover glow effect.
                 */
                this.blobIntensity = (__runInitializers(this, _blobPosition_extraInitializers), __runInitializers(this, _blobIntensity_initializers, 0.5));
                /**
                 * Gets or sets the near size of the hover glow effect.
                 */
                this.blobNearSize = (__runInitializers(this, _blobIntensity_extraInitializers), __runInitializers(this, _blobNearSize_initializers, 0.01));
                /**
                 * Gets or sets the far size of the hover glow effect.
                 */
                this.blobFarSize = (__runInitializers(this, _blobNearSize_extraInitializers), __runInitializers(this, _blobFarSize_initializers, 0.03));
                /**
                 * Gets or sets the distance considered "near" to the mesh, which controls the size of the hover glow effect (see blobNearSize).
                 */
                this.blobNearDistance = (__runInitializers(this, _blobFarSize_extraInitializers), __runInitializers(this, _blobNearDistance_initializers, 0));
                /**
                 * Gets or sets the distance considered "far" from the mesh, which controls the size of the hover glow effect (see blobFarSize).
                 */
                this.blobFarDistance = (__runInitializers(this, _blobNearDistance_extraInitializers), __runInitializers(this, _blobFarDistance_initializers, 0.08));
                /**
                 * Gets or sets the length of the hover glow effect fade.
                 */
                this.blobFadeLength = (__runInitializers(this, _blobFarDistance_extraInitializers), __runInitializers(this, _blobFadeLength_initializers, 0.576));
                /**
                 * Gets or sets the progress of the hover glow effect selection animation corresponding to the left pointer (0.0 - 1.0).
                 */
                this.blobPulse = (__runInitializers(this, _blobFadeLength_extraInitializers), __runInitializers(this, _blobPulse_initializers, 0));
                /**
                 * Gets or sets the opacity of the hover glow effect corresponding to the left pointer (0.0 - 1.0). Default is 0.
                 */
                this.blobFade = (__runInitializers(this, _blobPulse_extraInitializers), __runInitializers(this, _blobFade_initializers, 1));
                /**
                 * Gets or sets the position of the hover glow effect.
                 */
                this.blobPosition2 = (__runInitializers(this, _blobFade_extraInitializers), __runInitializers(this, _blobPosition2_initializers, new Vector3(0.2, 0, 0.1)));
                /**
                 * Gets or sets the size of the hover glow effect when the right pointer is considered "near" to the mesh (see blobNearDistance).
                 */
                this.blobNearSize2 = (__runInitializers(this, _blobPosition2_extraInitializers), __runInitializers(this, _blobNearSize2_initializers, 0.01));
                /**
                 * Gets or sets the progress of the hover glow effect selection animation corresponding to the right pointer (0.0 - 1.0).
                 */
                this.blobPulse2 = (__runInitializers(this, _blobNearSize2_extraInitializers), __runInitializers(this, _blobPulse2_initializers, 0));
                /**
                 * Gets or sets the opacity of the hover glow effect corresponding to the right pointer (0.0 - 1.0). Default is 1.
                 */
                this.blobFade2 = (__runInitializers(this, _blobPulse2_extraInitializers), __runInitializers(this, _blobFade2_initializers, 1));
                /**
                 * Gets or sets the texture of the hover glow effect.
                 */
                this.blobTexture = (__runInitializers(this, _blobFade2_extraInitializers), __runInitializers(this, _blobTexture_initializers, new Texture("", this.getScene())));
                /**
                 * Gets or sets the finger position for left index.
                 */
                this.leftIndexPosition = (__runInitializers(this, _blobTexture_extraInitializers), __runInitializers(this, _leftIndexPosition_initializers, new Vector3(0, 0, 1)));
                /**
                 * Gets or sets the finger position for right index.
                 */
                this.rightIndexPosition = (__runInitializers(this, _leftIndexPosition_extraInitializers), __runInitializers(this, _rightIndexPosition_initializers, new Vector3(-1, -1, -1)));
                /**
                 * Gets or sets the finger position for left index middle position.
                 */
                this.leftIndexMiddlePosition = (__runInitializers(this, _rightIndexPosition_extraInitializers), __runInitializers(this, _leftIndexMiddlePosition_initializers, new Vector3(0, 0, 0)));
                /**
                 * Gets or sets the finger position for right index middle position.
                 */
                this.rightIndexMiddlePosition = (__runInitializers(this, _leftIndexMiddlePosition_extraInitializers), __runInitializers(this, _rightIndexMiddlePosition_initializers, new Vector3(0, 0, 0)));
                /**
                 * Gets or sets the Decal Scale for XY.
                 */
                this.decalScaleXY = (__runInitializers(this, _rightIndexMiddlePosition_extraInitializers), __runInitializers(this, _decalScaleXY_initializers, new Vector2(1.5, 1.5)));
                /**
                 * Gets or sets decalFrontOnly
                 * Default is true
                 */
                this.decalFrontOnly = (__runInitializers(this, _decalScaleXY_extraInitializers), __runInitializers(this, _decalFrontOnly_initializers, true));
                /**
                 * Gets or sets the Rim Light intensity.
                 */
                this.rimIntensity = (__runInitializers(this, _decalFrontOnly_extraInitializers), __runInitializers(this, _rimIntensity_initializers, 0.287));
                /**
                 * Gets or sets the Rim Light hue shift value.
                 */
                this.rimHueShift = (__runInitializers(this, _rimIntensity_extraInitializers), __runInitializers(this, _rimHueShift_initializers, 0));
                /**
                 * Gets or sets the Rim Light saturation shift value.
                 */
                this.rimSaturationShift = (__runInitializers(this, _rimHueShift_extraInitializers), __runInitializers(this, _rimSaturationShift_initializers, 0));
                /**
                 * Gets or sets the Rim Light value shift.
                 */
                this.rimValueShift = (__runInitializers(this, _rimSaturationShift_extraInitializers), __runInitializers(this, _rimValueShift_initializers, -1));
                /**
                 * Gets or sets the intensity of the iridescence effect.
                 */
                this.iridescenceIntensity = (__runInitializers(this, _rimValueShift_extraInitializers), __runInitializers(this, _iridescenceIntensity_initializers, 0));
                /**
                 * @internal
                 */
                this.useGlobalLeftIndex = (__runInitializers(this, _iridescenceIntensity_extraInitializers), 1.0);
                /**
                 * @internal
                 */
                this.useGlobalRightIndex = 1.0;
                /**
                 * @internal
                 */
                this.globalLeftIndexTipProximity = 0.0;
                /**
                 * @internal
                 */
                this.globalRightIndexTipProximity = 0.0;
                /**
                 * @internal
                 */
                this.globalLeftIndexTipPosition = new Vector4(0.5, 0.0, -0.55, 1.0);
                /**
                 * @internal
                 */
                this.globaRightIndexTipPosition = new Vector4(0.0, 0.0, 0.0, 1.0);
                /**
                 * @internal
                 */
                this.globalLeftThumbTipPosition = new Vector4(0.5, 0.0, -0.55, 1.0);
                /**
                 * @internal
                 */
                this.globalRightThumbTipPosition = new Vector4(0.0, 0.0, 0.0, 1.0);
                /**
                 * @internal
                 */
                this.globalLeftIndexMiddlePosition = new Vector4(0.5, 0.0, -0.55, 1.0);
                /**
                 * @internal
                 */
                this.globalRightIndexMiddlePosition = new Vector4(0.0, 0.0, 0.0, 1.0);
                this.alphaMode = Constants.ALPHA_DISABLE;
                this.backFaceCulling = false;
                const textureUrl = Tools.GetAssetUrl(_a.BLUE_GRADIENT_TEXTURE_URL);
                this._blueGradientTexture = new Texture(textureUrl, scene, true, false, Texture.NEAREST_SAMPLINGMODE);
                this._decalTexture = new Texture("", this.getScene());
                this._reflectionMapTexture = new Texture("", this.getScene());
                this._indirectEnvTexture = new Texture("", this.getScene());
            }
            needAlphaBlending() {
                return false;
            }
            needAlphaTesting() {
                return false;
            }
            getAlphaTestTexture() {
                return null;
            }
            // Methods
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new MRDLSliderThumbMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                if (!this._blueGradientTexture.isReady()) {
                    return false;
                }
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, false, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "mrdlSliderThumb";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "viewProjection",
                        "cameraPosition",
                        "_Radius_",
                        "_Bevel_Front_",
                        "_Bevel_Front_Stretch_",
                        "_Bevel_Back_",
                        "_Bevel_Back_Stretch_",
                        "_Radius_Top_Left_",
                        "_Radius_Top_Right_",
                        "_Radius_Bottom_Left_",
                        "_Radius_Bottom_Right_",
                        "_Bulge_Enabled_",
                        "_Bulge_Height_",
                        "_Bulge_Radius_",
                        "_Sun_Intensity_",
                        "_Sun_Theta_",
                        "_Sun_Phi_",
                        "_Indirect_Diffuse_",
                        "_Albedo_",
                        "_Specular_",
                        "_Shininess_",
                        "_Sharpness_",
                        "_Subsurface_",
                        "_Left_Color_",
                        "_Right_Color_",
                        "_Reflection_",
                        "_Front_Reflect_",
                        "_Edge_Reflect_",
                        "_Power_",
                        "_Sky_Color_",
                        "_Horizon_Color_",
                        "_Ground_Color_",
                        "_Horizon_Power_",
                        "_Reflection_Map_",
                        "_Indirect_Environment_",
                        "_Width_",
                        "_Fuzz_",
                        "_Min_Fuzz_",
                        "_Clip_Fade_",
                        "_Hue_Shift_",
                        "_Saturation_Shift_",
                        "_Value_Shift_",
                        "_Blob_Position_",
                        "_Blob_Intensity_",
                        "_Blob_Near_Size_",
                        "_Blob_Far_Size_",
                        "_Blob_Near_Distance_",
                        "_Blob_Far_Distance_",
                        "_Blob_Fade_Length_",
                        "_Blob_Pulse_",
                        "_Blob_Fade_",
                        "_Blob_Texture_",
                        "_Blob_Position_2_",
                        "_Blob_Near_Size_2_",
                        "_Blob_Pulse_2_",
                        "_Blob_Fade_2_",
                        "_Left_Index_Pos_",
                        "_Right_Index_Pos_",
                        "_Left_Index_Middle_Pos_",
                        "_Right_Index_Middle_Pos_",
                        "_Decal_",
                        "_Decal_Scale_XY_",
                        "_Decal_Front_Only_",
                        "_Rim_Intensity_",
                        "_Rim_Texture_",
                        "_Rim_Hue_Shift_",
                        "_Rim_Saturation_Shift_",
                        "_Rim_Value_Shift_",
                        "_Iridescence_Intensity_",
                        "_Iridescence_Texture_",
                        "Use_Global_Left_Index",
                        "Use_Global_Right_Index",
                        "Global_Left_Index_Tip_Position",
                        "Global_Right_Index_Tip_Position",
                        "Global_Left_Thumb_Tip_Position",
                        "Global_Right_Thumb_Tip_Position",
                        "Global_Left_Index_Middle_Position;",
                        "Global_Right_Index_Middle_Position",
                        "Global_Left_Index_Tip_Proximity",
                        "Global_Right_Index_Tip_Proximity",
                    ];
                    const samplers = ["_Rim_Texture_", "_Iridescence_Texture_"];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        extraInitializationsAsync: async () => {
                            await Promise.all([import("./shaders/mrdlSliderThumb.vertex.js"), import("./shaders/mrdlSliderThumb.fragment.js")]);
                        },
                    }, engine), defines);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", this.getScene().getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", this.getScene().activeCamera.position);
                // "Round Rect"
                this._activeEffect.setFloat("_Radius_", this.radius);
                this._activeEffect.setFloat("_Bevel_Front_", this.bevelFront);
                this._activeEffect.setFloat("_Bevel_Front_Stretch_", this.bevelFrontStretch);
                this._activeEffect.setFloat("_Bevel_Back_", this.bevelBack);
                this._activeEffect.setFloat("_Bevel_Back_Stretch_", this.bevelBackStretch);
                // "Radii Multipliers"
                this._activeEffect.setFloat("_Radius_Top_Left_", this.radiusTopLeft);
                this._activeEffect.setFloat("_Radius_Top_Right_", this.radiusTopRight);
                this._activeEffect.setFloat("_Radius_Bottom_Left_", this.radiusBottomLeft);
                this._activeEffect.setFloat("_Radius_Bottom_Right_", this.radiusBottomRight);
                // "Bulge"
                this._activeEffect.setFloat("_Bulge_Enabled_", this.bulgeEnabled ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Bulge_Height_", this.bulgeHeight);
                this._activeEffect.setFloat("_Bulge_Radius_", this.bulgeRadius);
                // "Sun"
                this._activeEffect.setFloat("_Sun_Intensity_", this.sunIntensity);
                this._activeEffect.setFloat("_Sun_Theta_", this.sunTheta);
                this._activeEffect.setFloat("_Sun_Phi_", this.sunPhi);
                this._activeEffect.setFloat("_Indirect_Diffuse_", this.indirectDiffuse);
                // "Diffuse And Specular"
                this._activeEffect.setDirectColor4("_Albedo_", this.albedo);
                this._activeEffect.setFloat("_Specular_", this.specular);
                this._activeEffect.setFloat("_Shininess_", this.shininess);
                this._activeEffect.setFloat("_Sharpness_", this.sharpness);
                this._activeEffect.setFloat("_Subsurface_", this.subsurface);
                // "Gradient"
                this._activeEffect.setDirectColor4("_Left_Color_", this.leftGradientColor);
                this._activeEffect.setDirectColor4("_Right_Color_", this.rightGradientColor);
                // "Reflection"
                this._activeEffect.setFloat("_Reflection_", this.reflection);
                this._activeEffect.setFloat("_Front_Reflect_", this.frontReflect);
                this._activeEffect.setFloat("_Edge_Reflect_", this.edgeReflect);
                this._activeEffect.setFloat("_Power_", this.power);
                // "Sky Environment"
                //define SKY_ENABLED true;
                this._activeEffect.setDirectColor4("_Sky_Color_", this.skyColor);
                this._activeEffect.setDirectColor4("_Horizon_Color_", this.horizonColor);
                this._activeEffect.setDirectColor4("_Ground_Color_", this.groundColor);
                this._activeEffect.setFloat("_Horizon_Power_", this.horizonPower);
                // "Mapped Environment"
                //define ENV_ENABLE false;
                this._activeEffect.setTexture("_Reflection_Map_", this._reflectionMapTexture);
                this._activeEffect.setTexture("_Indirect_Environment_", this._indirectEnvTexture);
                // "FingerOcclusion"
                //define OCCLUSION_ENABLED false;
                this._activeEffect.setFloat("_Width_", this.width);
                this._activeEffect.setFloat("_Fuzz_", this.fuzz);
                this._activeEffect.setFloat("_Min_Fuzz_", this.minFuzz);
                this._activeEffect.setFloat("_Clip_Fade_", this.clipFade);
                // "View Based Color Shift"
                this._activeEffect.setFloat("_Hue_Shift_", this.hueShift);
                this._activeEffect.setFloat("_Saturation_Shift_", this.saturationShift);
                this._activeEffect.setFloat("_Value_Shift_", this.valueShift);
                // "Blob"
                //define BLOB_ENABLE false;
                this._activeEffect.setVector3("_Blob_Position_", this.blobPosition);
                this._activeEffect.setFloat("_Blob_Intensity_", this.blobIntensity);
                this._activeEffect.setFloat("_Blob_Near_Size_", this.blobNearSize);
                this._activeEffect.setFloat("_Blob_Far_Size_", this.blobFarSize);
                this._activeEffect.setFloat("_Blob_Near_Distance_", this.blobNearDistance);
                this._activeEffect.setFloat("_Blob_Far_Distance_", this.blobFarDistance);
                this._activeEffect.setFloat("_Blob_Fade_Length_", this.blobFadeLength);
                this._activeEffect.setFloat("_Blob_Pulse_", this.blobPulse);
                this._activeEffect.setFloat("_Blob_Fade_", this.blobFade);
                // "Blob Texture"
                this._activeEffect.setTexture("_Blob_Texture_", this.blobTexture);
                // "Blob 2"
                //define BLOB_ENABLE_2 true;
                this._activeEffect.setVector3("_Blob_Position_2_", this.blobPosition2);
                this._activeEffect.setFloat("_Blob_Near_Size_2_", this.blobNearSize2);
                this._activeEffect.setFloat("_Blob_Pulse_2_", this.blobPulse2);
                this._activeEffect.setFloat("_Blob_Fade_2_", this.blobFade2);
                // "Finger Positions"
                this._activeEffect.setVector3("_Left_Index_Pos_", this.leftIndexPosition);
                this._activeEffect.setVector3("_Right_Index_Pos_", this.rightIndexPosition);
                this._activeEffect.setVector3("_Left_Index_Middle_Pos_", this.leftIndexMiddlePosition);
                this._activeEffect.setVector3("_Right_Index_Middle_Pos_", this.rightIndexMiddlePosition);
                // "Decal Texture"
                //define DECAL_ENABLE false;
                this._activeEffect.setTexture("_Decal_", this._decalTexture);
                this._activeEffect.setVector2("_Decal_Scale_XY_", this.decalScaleXY);
                this._activeEffect.setFloat("_Decal_Front_Only_", this.decalFrontOnly ? 1.0 : 0.0);
                // "Rim Light"
                this._activeEffect.setFloat("_Rim_Intensity_", this.rimIntensity);
                this._activeEffect.setTexture("_Rim_Texture_", this._blueGradientTexture);
                this._activeEffect.setFloat("_Rim_Hue_Shift_", this.rimHueShift);
                this._activeEffect.setFloat("_Rim_Saturation_Shift_", this.rimSaturationShift);
                this._activeEffect.setFloat("_Rim_Value_Shift_", this.rimValueShift);
                // "Iridescence"
                //define IRIDESCENCE_ENABLED true;
                this._activeEffect.setFloat("_Iridescence_Intensity_", this.iridescenceIntensity);
                this._activeEffect.setTexture("_Iridescence_Texture_", this._blueGradientTexture);
                // Global inputs
                this._activeEffect.setFloat("Use_Global_Left_Index", this.useGlobalLeftIndex);
                this._activeEffect.setFloat("Use_Global_Right_Index", this.useGlobalRightIndex);
                this._activeEffect.setVector4("Global_Left_Index_Tip_Position", this.globalLeftIndexTipPosition);
                this._activeEffect.setVector4("Global_Right_Index_Tip_Position", this.globaRightIndexTipPosition);
                this._activeEffect.setVector4("Global_Left_Thumb_Tip_Position", this.globalLeftThumbTipPosition);
                this._activeEffect.setVector4("Global_Right_Thumb_Tip_Position", this.globalRightThumbTipPosition);
                this._activeEffect.setVector4("Global_Left_Index_Middle_Position", this.globalLeftIndexMiddlePosition);
                this._activeEffect.setVector4("Global_Right_Index_Middle_Position", this.globalRightIndexMiddlePosition);
                this._activeEffect.setFloat("Global_Left_Index_Tip_Proximity", this.globalLeftIndexTipProximity);
                this._activeEffect.setFloat("Global_Right_Index_Tip_Proximity", this.globalRightIndexTipProximity);
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
                this._reflectionMapTexture.dispose();
                this._indirectEnvTexture.dispose();
                this._blueGradientTexture.dispose();
                this._decalTexture.dispose();
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = super.serialize();
                serializationObject.customType = "BABYLON.MRDLSliderThumbMaterial";
                return serializationObject;
            }
            getClassName() {
                return "MRDLSliderThumbMaterial";
            }
            // Statics
            /**
             * Parses a serialized MRDLSliderThumbMaterial and returns a new instance.
             * @param source - the serialized object
             * @param scene - defines the hosting scene
             * @param rootUrl - the root URL used to load any associated textures
             * @returns a new MRDLSliderThumbMaterial
             */
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _radius_decorators = [serialize()];
            _bevelFront_decorators = [serialize()];
            _bevelFrontStretch_decorators = [serialize()];
            _bevelBack_decorators = [serialize()];
            _bevelBackStretch_decorators = [serialize()];
            _radiusTopLeft_decorators = [serialize()];
            _radiusTopRight_decorators = [serialize()];
            _radiusBottomLeft_decorators = [serialize()];
            _radiusBottomRight_decorators = [serialize()];
            _bulgeEnabled_decorators = [serialize()];
            _bulgeHeight_decorators = [serialize()];
            _bulgeRadius_decorators = [serialize()];
            _sunIntensity_decorators = [serialize()];
            _sunTheta_decorators = [serialize()];
            _sunPhi_decorators = [serialize()];
            _indirectDiffuse_decorators = [serialize()];
            _albedo_decorators = [serialize()];
            _specular_decorators = [serialize()];
            _shininess_decorators = [serialize()];
            _sharpness_decorators = [serialize()];
            _subsurface_decorators = [serialize()];
            _leftGradientColor_decorators = [serialize()];
            _rightGradientColor_decorators = [serialize()];
            _reflection_decorators = [serialize()];
            _frontReflect_decorators = [serialize()];
            _edgeReflect_decorators = [serialize()];
            _power_decorators = [serialize()];
            _skyColor_decorators = [serialize()];
            _horizonColor_decorators = [serialize()];
            _groundColor_decorators = [serialize()];
            _horizonPower_decorators = [serialize()];
            _width_decorators = [serialize()];
            _fuzz_decorators = [serialize()];
            _minFuzz_decorators = [serialize()];
            _clipFade_decorators = [serialize()];
            _hueShift_decorators = [serialize()];
            _saturationShift_decorators = [serialize()];
            _valueShift_decorators = [serialize()];
            _blobPosition_decorators = [serialize()];
            _blobIntensity_decorators = [serialize()];
            _blobNearSize_decorators = [serialize()];
            _blobFarSize_decorators = [serialize()];
            _blobNearDistance_decorators = [serialize()];
            _blobFarDistance_decorators = [serialize()];
            _blobFadeLength_decorators = [serialize()];
            _blobPulse_decorators = [serialize()];
            _blobFade_decorators = [serialize()];
            _blobPosition2_decorators = [serialize()];
            _blobNearSize2_decorators = [serialize()];
            _blobPulse2_decorators = [serialize()];
            _blobFade2_decorators = [serialize()];
            _blobTexture_decorators = [serialize()];
            _leftIndexPosition_decorators = [serialize()];
            _rightIndexPosition_decorators = [serialize()];
            _leftIndexMiddlePosition_decorators = [serialize()];
            _rightIndexMiddlePosition_decorators = [serialize()];
            _decalScaleXY_decorators = [serialize()];
            _decalFrontOnly_decorators = [serialize()];
            _rimIntensity_decorators = [serialize()];
            _rimHueShift_decorators = [serialize()];
            _rimSaturationShift_decorators = [serialize()];
            _rimValueShift_decorators = [serialize()];
            _iridescenceIntensity_decorators = [serialize()];
            __esDecorate(null, null, _radius_decorators, { kind: "field", name: "radius", static: false, private: false, access: { has: obj => "radius" in obj, get: obj => obj.radius, set: (obj, value) => { obj.radius = value; } }, metadata: _metadata }, _radius_initializers, _radius_extraInitializers);
            __esDecorate(null, null, _bevelFront_decorators, { kind: "field", name: "bevelFront", static: false, private: false, access: { has: obj => "bevelFront" in obj, get: obj => obj.bevelFront, set: (obj, value) => { obj.bevelFront = value; } }, metadata: _metadata }, _bevelFront_initializers, _bevelFront_extraInitializers);
            __esDecorate(null, null, _bevelFrontStretch_decorators, { kind: "field", name: "bevelFrontStretch", static: false, private: false, access: { has: obj => "bevelFrontStretch" in obj, get: obj => obj.bevelFrontStretch, set: (obj, value) => { obj.bevelFrontStretch = value; } }, metadata: _metadata }, _bevelFrontStretch_initializers, _bevelFrontStretch_extraInitializers);
            __esDecorate(null, null, _bevelBack_decorators, { kind: "field", name: "bevelBack", static: false, private: false, access: { has: obj => "bevelBack" in obj, get: obj => obj.bevelBack, set: (obj, value) => { obj.bevelBack = value; } }, metadata: _metadata }, _bevelBack_initializers, _bevelBack_extraInitializers);
            __esDecorate(null, null, _bevelBackStretch_decorators, { kind: "field", name: "bevelBackStretch", static: false, private: false, access: { has: obj => "bevelBackStretch" in obj, get: obj => obj.bevelBackStretch, set: (obj, value) => { obj.bevelBackStretch = value; } }, metadata: _metadata }, _bevelBackStretch_initializers, _bevelBackStretch_extraInitializers);
            __esDecorate(null, null, _radiusTopLeft_decorators, { kind: "field", name: "radiusTopLeft", static: false, private: false, access: { has: obj => "radiusTopLeft" in obj, get: obj => obj.radiusTopLeft, set: (obj, value) => { obj.radiusTopLeft = value; } }, metadata: _metadata }, _radiusTopLeft_initializers, _radiusTopLeft_extraInitializers);
            __esDecorate(null, null, _radiusTopRight_decorators, { kind: "field", name: "radiusTopRight", static: false, private: false, access: { has: obj => "radiusTopRight" in obj, get: obj => obj.radiusTopRight, set: (obj, value) => { obj.radiusTopRight = value; } }, metadata: _metadata }, _radiusTopRight_initializers, _radiusTopRight_extraInitializers);
            __esDecorate(null, null, _radiusBottomLeft_decorators, { kind: "field", name: "radiusBottomLeft", static: false, private: false, access: { has: obj => "radiusBottomLeft" in obj, get: obj => obj.radiusBottomLeft, set: (obj, value) => { obj.radiusBottomLeft = value; } }, metadata: _metadata }, _radiusBottomLeft_initializers, _radiusBottomLeft_extraInitializers);
            __esDecorate(null, null, _radiusBottomRight_decorators, { kind: "field", name: "radiusBottomRight", static: false, private: false, access: { has: obj => "radiusBottomRight" in obj, get: obj => obj.radiusBottomRight, set: (obj, value) => { obj.radiusBottomRight = value; } }, metadata: _metadata }, _radiusBottomRight_initializers, _radiusBottomRight_extraInitializers);
            __esDecorate(null, null, _bulgeEnabled_decorators, { kind: "field", name: "bulgeEnabled", static: false, private: false, access: { has: obj => "bulgeEnabled" in obj, get: obj => obj.bulgeEnabled, set: (obj, value) => { obj.bulgeEnabled = value; } }, metadata: _metadata }, _bulgeEnabled_initializers, _bulgeEnabled_extraInitializers);
            __esDecorate(null, null, _bulgeHeight_decorators, { kind: "field", name: "bulgeHeight", static: false, private: false, access: { has: obj => "bulgeHeight" in obj, get: obj => obj.bulgeHeight, set: (obj, value) => { obj.bulgeHeight = value; } }, metadata: _metadata }, _bulgeHeight_initializers, _bulgeHeight_extraInitializers);
            __esDecorate(null, null, _bulgeRadius_decorators, { kind: "field", name: "bulgeRadius", static: false, private: false, access: { has: obj => "bulgeRadius" in obj, get: obj => obj.bulgeRadius, set: (obj, value) => { obj.bulgeRadius = value; } }, metadata: _metadata }, _bulgeRadius_initializers, _bulgeRadius_extraInitializers);
            __esDecorate(null, null, _sunIntensity_decorators, { kind: "field", name: "sunIntensity", static: false, private: false, access: { has: obj => "sunIntensity" in obj, get: obj => obj.sunIntensity, set: (obj, value) => { obj.sunIntensity = value; } }, metadata: _metadata }, _sunIntensity_initializers, _sunIntensity_extraInitializers);
            __esDecorate(null, null, _sunTheta_decorators, { kind: "field", name: "sunTheta", static: false, private: false, access: { has: obj => "sunTheta" in obj, get: obj => obj.sunTheta, set: (obj, value) => { obj.sunTheta = value; } }, metadata: _metadata }, _sunTheta_initializers, _sunTheta_extraInitializers);
            __esDecorate(null, null, _sunPhi_decorators, { kind: "field", name: "sunPhi", static: false, private: false, access: { has: obj => "sunPhi" in obj, get: obj => obj.sunPhi, set: (obj, value) => { obj.sunPhi = value; } }, metadata: _metadata }, _sunPhi_initializers, _sunPhi_extraInitializers);
            __esDecorate(null, null, _indirectDiffuse_decorators, { kind: "field", name: "indirectDiffuse", static: false, private: false, access: { has: obj => "indirectDiffuse" in obj, get: obj => obj.indirectDiffuse, set: (obj, value) => { obj.indirectDiffuse = value; } }, metadata: _metadata }, _indirectDiffuse_initializers, _indirectDiffuse_extraInitializers);
            __esDecorate(null, null, _albedo_decorators, { kind: "field", name: "albedo", static: false, private: false, access: { has: obj => "albedo" in obj, get: obj => obj.albedo, set: (obj, value) => { obj.albedo = value; } }, metadata: _metadata }, _albedo_initializers, _albedo_extraInitializers);
            __esDecorate(null, null, _specular_decorators, { kind: "field", name: "specular", static: false, private: false, access: { has: obj => "specular" in obj, get: obj => obj.specular, set: (obj, value) => { obj.specular = value; } }, metadata: _metadata }, _specular_initializers, _specular_extraInitializers);
            __esDecorate(null, null, _shininess_decorators, { kind: "field", name: "shininess", static: false, private: false, access: { has: obj => "shininess" in obj, get: obj => obj.shininess, set: (obj, value) => { obj.shininess = value; } }, metadata: _metadata }, _shininess_initializers, _shininess_extraInitializers);
            __esDecorate(null, null, _sharpness_decorators, { kind: "field", name: "sharpness", static: false, private: false, access: { has: obj => "sharpness" in obj, get: obj => obj.sharpness, set: (obj, value) => { obj.sharpness = value; } }, metadata: _metadata }, _sharpness_initializers, _sharpness_extraInitializers);
            __esDecorate(null, null, _subsurface_decorators, { kind: "field", name: "subsurface", static: false, private: false, access: { has: obj => "subsurface" in obj, get: obj => obj.subsurface, set: (obj, value) => { obj.subsurface = value; } }, metadata: _metadata }, _subsurface_initializers, _subsurface_extraInitializers);
            __esDecorate(null, null, _leftGradientColor_decorators, { kind: "field", name: "leftGradientColor", static: false, private: false, access: { has: obj => "leftGradientColor" in obj, get: obj => obj.leftGradientColor, set: (obj, value) => { obj.leftGradientColor = value; } }, metadata: _metadata }, _leftGradientColor_initializers, _leftGradientColor_extraInitializers);
            __esDecorate(null, null, _rightGradientColor_decorators, { kind: "field", name: "rightGradientColor", static: false, private: false, access: { has: obj => "rightGradientColor" in obj, get: obj => obj.rightGradientColor, set: (obj, value) => { obj.rightGradientColor = value; } }, metadata: _metadata }, _rightGradientColor_initializers, _rightGradientColor_extraInitializers);
            __esDecorate(null, null, _reflection_decorators, { kind: "field", name: "reflection", static: false, private: false, access: { has: obj => "reflection" in obj, get: obj => obj.reflection, set: (obj, value) => { obj.reflection = value; } }, metadata: _metadata }, _reflection_initializers, _reflection_extraInitializers);
            __esDecorate(null, null, _frontReflect_decorators, { kind: "field", name: "frontReflect", static: false, private: false, access: { has: obj => "frontReflect" in obj, get: obj => obj.frontReflect, set: (obj, value) => { obj.frontReflect = value; } }, metadata: _metadata }, _frontReflect_initializers, _frontReflect_extraInitializers);
            __esDecorate(null, null, _edgeReflect_decorators, { kind: "field", name: "edgeReflect", static: false, private: false, access: { has: obj => "edgeReflect" in obj, get: obj => obj.edgeReflect, set: (obj, value) => { obj.edgeReflect = value; } }, metadata: _metadata }, _edgeReflect_initializers, _edgeReflect_extraInitializers);
            __esDecorate(null, null, _power_decorators, { kind: "field", name: "power", static: false, private: false, access: { has: obj => "power" in obj, get: obj => obj.power, set: (obj, value) => { obj.power = value; } }, metadata: _metadata }, _power_initializers, _power_extraInitializers);
            __esDecorate(null, null, _skyColor_decorators, { kind: "field", name: "skyColor", static: false, private: false, access: { has: obj => "skyColor" in obj, get: obj => obj.skyColor, set: (obj, value) => { obj.skyColor = value; } }, metadata: _metadata }, _skyColor_initializers, _skyColor_extraInitializers);
            __esDecorate(null, null, _horizonColor_decorators, { kind: "field", name: "horizonColor", static: false, private: false, access: { has: obj => "horizonColor" in obj, get: obj => obj.horizonColor, set: (obj, value) => { obj.horizonColor = value; } }, metadata: _metadata }, _horizonColor_initializers, _horizonColor_extraInitializers);
            __esDecorate(null, null, _groundColor_decorators, { kind: "field", name: "groundColor", static: false, private: false, access: { has: obj => "groundColor" in obj, get: obj => obj.groundColor, set: (obj, value) => { obj.groundColor = value; } }, metadata: _metadata }, _groundColor_initializers, _groundColor_extraInitializers);
            __esDecorate(null, null, _horizonPower_decorators, { kind: "field", name: "horizonPower", static: false, private: false, access: { has: obj => "horizonPower" in obj, get: obj => obj.horizonPower, set: (obj, value) => { obj.horizonPower = value; } }, metadata: _metadata }, _horizonPower_initializers, _horizonPower_extraInitializers);
            __esDecorate(null, null, _width_decorators, { kind: "field", name: "width", static: false, private: false, access: { has: obj => "width" in obj, get: obj => obj.width, set: (obj, value) => { obj.width = value; } }, metadata: _metadata }, _width_initializers, _width_extraInitializers);
            __esDecorate(null, null, _fuzz_decorators, { kind: "field", name: "fuzz", static: false, private: false, access: { has: obj => "fuzz" in obj, get: obj => obj.fuzz, set: (obj, value) => { obj.fuzz = value; } }, metadata: _metadata }, _fuzz_initializers, _fuzz_extraInitializers);
            __esDecorate(null, null, _minFuzz_decorators, { kind: "field", name: "minFuzz", static: false, private: false, access: { has: obj => "minFuzz" in obj, get: obj => obj.minFuzz, set: (obj, value) => { obj.minFuzz = value; } }, metadata: _metadata }, _minFuzz_initializers, _minFuzz_extraInitializers);
            __esDecorate(null, null, _clipFade_decorators, { kind: "field", name: "clipFade", static: false, private: false, access: { has: obj => "clipFade" in obj, get: obj => obj.clipFade, set: (obj, value) => { obj.clipFade = value; } }, metadata: _metadata }, _clipFade_initializers, _clipFade_extraInitializers);
            __esDecorate(null, null, _hueShift_decorators, { kind: "field", name: "hueShift", static: false, private: false, access: { has: obj => "hueShift" in obj, get: obj => obj.hueShift, set: (obj, value) => { obj.hueShift = value; } }, metadata: _metadata }, _hueShift_initializers, _hueShift_extraInitializers);
            __esDecorate(null, null, _saturationShift_decorators, { kind: "field", name: "saturationShift", static: false, private: false, access: { has: obj => "saturationShift" in obj, get: obj => obj.saturationShift, set: (obj, value) => { obj.saturationShift = value; } }, metadata: _metadata }, _saturationShift_initializers, _saturationShift_extraInitializers);
            __esDecorate(null, null, _valueShift_decorators, { kind: "field", name: "valueShift", static: false, private: false, access: { has: obj => "valueShift" in obj, get: obj => obj.valueShift, set: (obj, value) => { obj.valueShift = value; } }, metadata: _metadata }, _valueShift_initializers, _valueShift_extraInitializers);
            __esDecorate(null, null, _blobPosition_decorators, { kind: "field", name: "blobPosition", static: false, private: false, access: { has: obj => "blobPosition" in obj, get: obj => obj.blobPosition, set: (obj, value) => { obj.blobPosition = value; } }, metadata: _metadata }, _blobPosition_initializers, _blobPosition_extraInitializers);
            __esDecorate(null, null, _blobIntensity_decorators, { kind: "field", name: "blobIntensity", static: false, private: false, access: { has: obj => "blobIntensity" in obj, get: obj => obj.blobIntensity, set: (obj, value) => { obj.blobIntensity = value; } }, metadata: _metadata }, _blobIntensity_initializers, _blobIntensity_extraInitializers);
            __esDecorate(null, null, _blobNearSize_decorators, { kind: "field", name: "blobNearSize", static: false, private: false, access: { has: obj => "blobNearSize" in obj, get: obj => obj.blobNearSize, set: (obj, value) => { obj.blobNearSize = value; } }, metadata: _metadata }, _blobNearSize_initializers, _blobNearSize_extraInitializers);
            __esDecorate(null, null, _blobFarSize_decorators, { kind: "field", name: "blobFarSize", static: false, private: false, access: { has: obj => "blobFarSize" in obj, get: obj => obj.blobFarSize, set: (obj, value) => { obj.blobFarSize = value; } }, metadata: _metadata }, _blobFarSize_initializers, _blobFarSize_extraInitializers);
            __esDecorate(null, null, _blobNearDistance_decorators, { kind: "field", name: "blobNearDistance", static: false, private: false, access: { has: obj => "blobNearDistance" in obj, get: obj => obj.blobNearDistance, set: (obj, value) => { obj.blobNearDistance = value; } }, metadata: _metadata }, _blobNearDistance_initializers, _blobNearDistance_extraInitializers);
            __esDecorate(null, null, _blobFarDistance_decorators, { kind: "field", name: "blobFarDistance", static: false, private: false, access: { has: obj => "blobFarDistance" in obj, get: obj => obj.blobFarDistance, set: (obj, value) => { obj.blobFarDistance = value; } }, metadata: _metadata }, _blobFarDistance_initializers, _blobFarDistance_extraInitializers);
            __esDecorate(null, null, _blobFadeLength_decorators, { kind: "field", name: "blobFadeLength", static: false, private: false, access: { has: obj => "blobFadeLength" in obj, get: obj => obj.blobFadeLength, set: (obj, value) => { obj.blobFadeLength = value; } }, metadata: _metadata }, _blobFadeLength_initializers, _blobFadeLength_extraInitializers);
            __esDecorate(null, null, _blobPulse_decorators, { kind: "field", name: "blobPulse", static: false, private: false, access: { has: obj => "blobPulse" in obj, get: obj => obj.blobPulse, set: (obj, value) => { obj.blobPulse = value; } }, metadata: _metadata }, _blobPulse_initializers, _blobPulse_extraInitializers);
            __esDecorate(null, null, _blobFade_decorators, { kind: "field", name: "blobFade", static: false, private: false, access: { has: obj => "blobFade" in obj, get: obj => obj.blobFade, set: (obj, value) => { obj.blobFade = value; } }, metadata: _metadata }, _blobFade_initializers, _blobFade_extraInitializers);
            __esDecorate(null, null, _blobPosition2_decorators, { kind: "field", name: "blobPosition2", static: false, private: false, access: { has: obj => "blobPosition2" in obj, get: obj => obj.blobPosition2, set: (obj, value) => { obj.blobPosition2 = value; } }, metadata: _metadata }, _blobPosition2_initializers, _blobPosition2_extraInitializers);
            __esDecorate(null, null, _blobNearSize2_decorators, { kind: "field", name: "blobNearSize2", static: false, private: false, access: { has: obj => "blobNearSize2" in obj, get: obj => obj.blobNearSize2, set: (obj, value) => { obj.blobNearSize2 = value; } }, metadata: _metadata }, _blobNearSize2_initializers, _blobNearSize2_extraInitializers);
            __esDecorate(null, null, _blobPulse2_decorators, { kind: "field", name: "blobPulse2", static: false, private: false, access: { has: obj => "blobPulse2" in obj, get: obj => obj.blobPulse2, set: (obj, value) => { obj.blobPulse2 = value; } }, metadata: _metadata }, _blobPulse2_initializers, _blobPulse2_extraInitializers);
            __esDecorate(null, null, _blobFade2_decorators, { kind: "field", name: "blobFade2", static: false, private: false, access: { has: obj => "blobFade2" in obj, get: obj => obj.blobFade2, set: (obj, value) => { obj.blobFade2 = value; } }, metadata: _metadata }, _blobFade2_initializers, _blobFade2_extraInitializers);
            __esDecorate(null, null, _blobTexture_decorators, { kind: "field", name: "blobTexture", static: false, private: false, access: { has: obj => "blobTexture" in obj, get: obj => obj.blobTexture, set: (obj, value) => { obj.blobTexture = value; } }, metadata: _metadata }, _blobTexture_initializers, _blobTexture_extraInitializers);
            __esDecorate(null, null, _leftIndexPosition_decorators, { kind: "field", name: "leftIndexPosition", static: false, private: false, access: { has: obj => "leftIndexPosition" in obj, get: obj => obj.leftIndexPosition, set: (obj, value) => { obj.leftIndexPosition = value; } }, metadata: _metadata }, _leftIndexPosition_initializers, _leftIndexPosition_extraInitializers);
            __esDecorate(null, null, _rightIndexPosition_decorators, { kind: "field", name: "rightIndexPosition", static: false, private: false, access: { has: obj => "rightIndexPosition" in obj, get: obj => obj.rightIndexPosition, set: (obj, value) => { obj.rightIndexPosition = value; } }, metadata: _metadata }, _rightIndexPosition_initializers, _rightIndexPosition_extraInitializers);
            __esDecorate(null, null, _leftIndexMiddlePosition_decorators, { kind: "field", name: "leftIndexMiddlePosition", static: false, private: false, access: { has: obj => "leftIndexMiddlePosition" in obj, get: obj => obj.leftIndexMiddlePosition, set: (obj, value) => { obj.leftIndexMiddlePosition = value; } }, metadata: _metadata }, _leftIndexMiddlePosition_initializers, _leftIndexMiddlePosition_extraInitializers);
            __esDecorate(null, null, _rightIndexMiddlePosition_decorators, { kind: "field", name: "rightIndexMiddlePosition", static: false, private: false, access: { has: obj => "rightIndexMiddlePosition" in obj, get: obj => obj.rightIndexMiddlePosition, set: (obj, value) => { obj.rightIndexMiddlePosition = value; } }, metadata: _metadata }, _rightIndexMiddlePosition_initializers, _rightIndexMiddlePosition_extraInitializers);
            __esDecorate(null, null, _decalScaleXY_decorators, { kind: "field", name: "decalScaleXY", static: false, private: false, access: { has: obj => "decalScaleXY" in obj, get: obj => obj.decalScaleXY, set: (obj, value) => { obj.decalScaleXY = value; } }, metadata: _metadata }, _decalScaleXY_initializers, _decalScaleXY_extraInitializers);
            __esDecorate(null, null, _decalFrontOnly_decorators, { kind: "field", name: "decalFrontOnly", static: false, private: false, access: { has: obj => "decalFrontOnly" in obj, get: obj => obj.decalFrontOnly, set: (obj, value) => { obj.decalFrontOnly = value; } }, metadata: _metadata }, _decalFrontOnly_initializers, _decalFrontOnly_extraInitializers);
            __esDecorate(null, null, _rimIntensity_decorators, { kind: "field", name: "rimIntensity", static: false, private: false, access: { has: obj => "rimIntensity" in obj, get: obj => obj.rimIntensity, set: (obj, value) => { obj.rimIntensity = value; } }, metadata: _metadata }, _rimIntensity_initializers, _rimIntensity_extraInitializers);
            __esDecorate(null, null, _rimHueShift_decorators, { kind: "field", name: "rimHueShift", static: false, private: false, access: { has: obj => "rimHueShift" in obj, get: obj => obj.rimHueShift, set: (obj, value) => { obj.rimHueShift = value; } }, metadata: _metadata }, _rimHueShift_initializers, _rimHueShift_extraInitializers);
            __esDecorate(null, null, _rimSaturationShift_decorators, { kind: "field", name: "rimSaturationShift", static: false, private: false, access: { has: obj => "rimSaturationShift" in obj, get: obj => obj.rimSaturationShift, set: (obj, value) => { obj.rimSaturationShift = value; } }, metadata: _metadata }, _rimSaturationShift_initializers, _rimSaturationShift_extraInitializers);
            __esDecorate(null, null, _rimValueShift_decorators, { kind: "field", name: "rimValueShift", static: false, private: false, access: { has: obj => "rimValueShift" in obj, get: obj => obj.rimValueShift, set: (obj, value) => { obj.rimValueShift = value; } }, metadata: _metadata }, _rimValueShift_initializers, _rimValueShift_extraInitializers);
            __esDecorate(null, null, _iridescenceIntensity_decorators, { kind: "field", name: "iridescenceIntensity", static: false, private: false, access: { has: obj => "iridescenceIntensity" in obj, get: obj => obj.iridescenceIntensity, set: (obj, value) => { obj.iridescenceIntensity = value; } }, metadata: _metadata }, _iridescenceIntensity_initializers, _iridescenceIntensity_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        /**
         * URL pointing to the texture used to define the coloring for the Iridescent Map effect.
         */
        _a.BLUE_GRADIENT_TEXTURE_URL = "https://assets.babylonjs.com/core/MRTK/MRDL/mrtk-mrdl-blue-gradient.png",
        _a;
})();
export { MRDLSliderThumbMaterial };
let _Registered = false;
/**
 * Registers the MRDLSliderThumbMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterMRDLSliderThumbMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.MRDLSliderThumbMaterial", MRDLSliderThumbMaterial);
}
