// @babylonjs/gui@9.16.1/2D/controls/displayGrid.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { Control } from "./control.pure.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
/** Class used to render a grid  */
let DisplayGrid = (() => {
    var _a;
    let _classSuper = Control;
    let _instanceExtraInitializers = [];
    let _get_displayMinorLines_decorators;
    let _get_displayMajorLines_decorators;
    let _get_background_decorators;
    let _get_cellWidth_decorators;
    let _get_cellHeight_decorators;
    let _get_minorLineTickness_decorators;
    let _get_minorLineColor_decorators;
    let _get_majorLineTickness_decorators;
    let _get_majorLineColor_decorators;
    let _get_majorLineFrequency_decorators;
    return _a = class DisplayGrid extends _classSuper {
            /** Gets or sets a boolean indicating if minor lines must be rendered (true by default)) */
            get displayMinorLines() {
                return this._displayMinorLines;
            }
            set displayMinorLines(value) {
                if (this._displayMinorLines === value) {
                    return;
                }
                this._displayMinorLines = value;
                this._markAsDirty();
            }
            /** Gets or sets a boolean indicating if major lines must be rendered (true by default)) */
            get displayMajorLines() {
                return this._displayMajorLines;
            }
            set displayMajorLines(value) {
                if (this._displayMajorLines === value) {
                    return;
                }
                this._displayMajorLines = value;
                this._markAsDirty();
            }
            /** Gets or sets background color (Black by default) */
            get background() {
                return this._background;
            }
            set background(value) {
                if (this._background === value) {
                    return;
                }
                this._background = value;
                this._markAsDirty();
            }
            /** Gets or sets the width of each cell (20 by default) */
            get cellWidth() {
                return this._cellWidth;
            }
            set cellWidth(value) {
                this._cellWidth = value;
                this._markAsDirty();
            }
            /** Gets or sets the height of each cell (20 by default) */
            get cellHeight() {
                return this._cellHeight;
            }
            set cellHeight(value) {
                this._cellHeight = value;
                this._markAsDirty();
            }
            /** Gets or sets the tickness of minor lines (1 by default) */
            get minorLineTickness() {
                return this._minorLineTickness;
            }
            set minorLineTickness(value) {
                this._minorLineTickness = value;
                this._markAsDirty();
            }
            /** Gets or sets the color of minor lines (DarkGray by default) */
            get minorLineColor() {
                return this._minorLineColor;
            }
            set minorLineColor(value) {
                this._minorLineColor = value;
                this._markAsDirty();
            }
            /** Gets or sets the tickness of major lines (2 by default) */
            get majorLineTickness() {
                return this._majorLineTickness;
            }
            set majorLineTickness(value) {
                this._majorLineTickness = value;
                this._markAsDirty();
            }
            /** Gets or sets the color of major lines (White by default) */
            get majorLineColor() {
                return this._majorLineColor;
            }
            set majorLineColor(value) {
                this._majorLineColor = value;
                this._markAsDirty();
            }
            /** Gets or sets the frequency of major lines (default is 1 every 5 minor lines)*/
            get majorLineFrequency() {
                return this._majorLineFrequency;
            }
            set majorLineFrequency(value) {
                this._majorLineFrequency = value;
                this._markAsDirty();
            }
            /**
             * Creates a new GridDisplayRectangle
             * @param name defines the control name
             */
            constructor(name) {
                super(name);
                this.name = (__runInitializers(this, _instanceExtraInitializers), name);
                this._cellWidth = 20;
                this._cellHeight = 20;
                this._minorLineTickness = 1;
                this._minorLineColor = "DarkGray";
                this._majorLineTickness = 2;
                this._majorLineColor = "White";
                this._majorLineFrequency = 5;
                this._background = "Black";
                this._displayMajorLines = true;
                this._displayMinorLines = true;
            }
            _draw(context) {
                context.save();
                this._applyStates(context);
                if (this._isEnabled) {
                    if (this._background) {
                        context.fillStyle = this._background;
                        context.fillRect(this._currentMeasure.left, this._currentMeasure.top, this._currentMeasure.width, this._currentMeasure.height);
                    }
                    const cellCountX = this._currentMeasure.width / this._cellWidth;
                    const cellCountY = this._currentMeasure.height / this._cellHeight;
                    // Minor lines
                    const left = this._currentMeasure.left + this._currentMeasure.width / 2;
                    const top = this._currentMeasure.top + this._currentMeasure.height / 2;
                    if (this._displayMinorLines) {
                        context.strokeStyle = this._minorLineColor;
                        context.lineWidth = this._minorLineTickness;
                        for (let x = -cellCountX / 2 + 1; x < cellCountX / 2; x++) {
                            const cellX = left + x * this.cellWidth;
                            context.beginPath();
                            context.moveTo(cellX, this._currentMeasure.top);
                            context.lineTo(cellX, this._currentMeasure.top + this._currentMeasure.height);
                            context.stroke();
                        }
                        for (let y = -cellCountY / 2 + 1; y < cellCountY / 2; y++) {
                            const cellY = top + y * this.cellHeight;
                            context.beginPath();
                            context.moveTo(this._currentMeasure.left, cellY);
                            context.lineTo(this._currentMeasure.left + this._currentMeasure.width, cellY);
                            context.stroke();
                        }
                    }
                    // Major lines
                    if (this._displayMajorLines) {
                        context.strokeStyle = this._majorLineColor;
                        context.lineWidth = this._majorLineTickness;
                        for (let x = -cellCountX / 2 + this._majorLineFrequency; x < cellCountX / 2; x += this._majorLineFrequency) {
                            const cellX = left + x * this.cellWidth;
                            context.beginPath();
                            context.moveTo(cellX, this._currentMeasure.top);
                            context.lineTo(cellX, this._currentMeasure.top + this._currentMeasure.height);
                            context.stroke();
                        }
                        for (let y = -cellCountY / 2 + this._majorLineFrequency; y < cellCountY / 2; y += this._majorLineFrequency) {
                            const cellY = top + y * this.cellHeight;
                            context.moveTo(this._currentMeasure.left, cellY);
                            context.lineTo(this._currentMeasure.left + this._currentMeasure.width, cellY);
                            context.closePath();
                            context.stroke();
                        }
                    }
                }
                context.restore();
            }
            _getTypeName() {
                return "DisplayGrid";
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_displayMinorLines_decorators = [serialize()];
            _get_displayMajorLines_decorators = [serialize()];
            _get_background_decorators = [serialize()];
            _get_cellWidth_decorators = [serialize()];
            _get_cellHeight_decorators = [serialize()];
            _get_minorLineTickness_decorators = [serialize()];
            _get_minorLineColor_decorators = [serialize()];
            _get_majorLineTickness_decorators = [serialize()];
            _get_majorLineColor_decorators = [serialize()];
            _get_majorLineFrequency_decorators = [serialize()];
            __esDecorate(_a, null, _get_displayMinorLines_decorators, { kind: "getter", name: "displayMinorLines", static: false, private: false, access: { has: obj => "displayMinorLines" in obj, get: obj => obj.displayMinorLines }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_displayMajorLines_decorators, { kind: "getter", name: "displayMajorLines", static: false, private: false, access: { has: obj => "displayMajorLines" in obj, get: obj => obj.displayMajorLines }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_background_decorators, { kind: "getter", name: "background", static: false, private: false, access: { has: obj => "background" in obj, get: obj => obj.background }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_cellWidth_decorators, { kind: "getter", name: "cellWidth", static: false, private: false, access: { has: obj => "cellWidth" in obj, get: obj => obj.cellWidth }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_cellHeight_decorators, { kind: "getter", name: "cellHeight", static: false, private: false, access: { has: obj => "cellHeight" in obj, get: obj => obj.cellHeight }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_minorLineTickness_decorators, { kind: "getter", name: "minorLineTickness", static: false, private: false, access: { has: obj => "minorLineTickness" in obj, get: obj => obj.minorLineTickness }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_minorLineColor_decorators, { kind: "getter", name: "minorLineColor", static: false, private: false, access: { has: obj => "minorLineColor" in obj, get: obj => obj.minorLineColor }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_majorLineTickness_decorators, { kind: "getter", name: "majorLineTickness", static: false, private: false, access: { has: obj => "majorLineTickness" in obj, get: obj => obj.majorLineTickness }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_majorLineColor_decorators, { kind: "getter", name: "majorLineColor", static: false, private: false, access: { has: obj => "majorLineColor" in obj, get: obj => obj.majorLineColor }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_majorLineFrequency_decorators, { kind: "getter", name: "majorLineFrequency", static: false, private: false, access: { has: obj => "majorLineFrequency" in obj, get: obj => obj.majorLineFrequency }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
export { DisplayGrid };
let _Registered = false;
/**
 * Registers the DisplayGrid class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterDisplayGrid() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.DisplayGrid", DisplayGrid);
}
