// @babylonjs/gui@9.16.1/2D/controls/ellipse.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { Container } from "./container.pure.js";
import { Control } from "./control.pure.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { serialize } from "@babylonjs/core/Misc/decorators.js";
/** Class used to create 2D ellipse containers */
let Ellipse = (() => {
    var _a;
    let _classSuper = Container;
    let _instanceExtraInitializers = [];
    let _get_thickness_decorators;
    let _get_arc_decorators;
    return _a = class Ellipse extends _classSuper {
            /** Gets or sets border thickness */
            get thickness() {
                return this._thickness;
            }
            set thickness(value) {
                if (this._thickness === value) {
                    return;
                }
                this._thickness = value;
                this._markAsDirty();
            }
            /** Gets or sets arcing of the ellipse (ratio of the circumference between 0 and 1) */
            get arc() {
                return this._arc;
            }
            set arc(value) {
                if (this._arc === value) {
                    return;
                }
                this._arc = value;
                this._markAsDirty();
            }
            /**
             * Creates a new Ellipse
             * @param name defines the control name
             */
            constructor(name) {
                super(name);
                this.name = (__runInitializers(this, _instanceExtraInitializers), name);
                this._thickness = 1;
                this._arc = 1;
            }
            _getTypeName() {
                return "Ellipse";
            }
            _localDraw(context) {
                context.save();
                if (this.shadowBlur || this.shadowOffsetX || this.shadowOffsetY) {
                    context.shadowColor = this.shadowColor;
                    context.shadowBlur = this.shadowBlur;
                    context.shadowOffsetX = this.shadowOffsetX * this._host.idealRatio;
                    context.shadowOffsetY = this.shadowOffsetY * this._host.idealRatio;
                }
                Control.drawEllipse(this._currentMeasure.left + this._currentMeasure.width / 2, this._currentMeasure.top + this._currentMeasure.height / 2, this._currentMeasure.width / 2 - this._thickness / 2, this._currentMeasure.height / 2 - this._thickness / 2, this._arc, context);
                if (this._backgroundGradient || this._background) {
                    context.fillStyle = this._getBackgroundColor(context);
                    context.fill();
                }
                if (this.shadowBlur || this.shadowOffsetX || this.shadowOffsetY) {
                    context.shadowBlur = 0;
                    context.shadowOffsetX = 0;
                    context.shadowOffsetY = 0;
                }
                if (this._thickness) {
                    if (this.color) {
                        context.strokeStyle = this.color;
                    }
                    context.lineWidth = this._thickness;
                    context.stroke();
                }
                context.restore();
            }
            _additionalProcessing(parentMeasure, context) {
                super._additionalProcessing(parentMeasure, context);
                this._measureForChildren.width -= 2 * this._thickness;
                this._measureForChildren.height -= 2 * this._thickness;
                this._measureForChildren.left += this._thickness;
                this._measureForChildren.top += this._thickness;
            }
            _clipForChildren(context) {
                Control.drawEllipse(this._currentMeasure.left + this._currentMeasure.width / 2, this._currentMeasure.top + this._currentMeasure.height / 2, this._currentMeasure.width / 2, this._currentMeasure.height / 2, this._arc, context);
                context.clip();
            }
            _renderHighlightSpecific(context) {
                Control.drawEllipse(this._currentMeasure.left + this._currentMeasure.width / 2, this._currentMeasure.top + this._currentMeasure.height / 2, this._currentMeasure.width / 2 - this._highlightLineWidth / 2, this._currentMeasure.height / 2 - this._highlightLineWidth / 2, this._arc, context);
                context.stroke();
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _get_thickness_decorators = [serialize()];
            _get_arc_decorators = [serialize()];
            __esDecorate(_a, null, _get_thickness_decorators, { kind: "getter", name: "thickness", static: false, private: false, access: { has: obj => "thickness" in obj, get: obj => obj.thickness }, metadata: _metadata }, null, _instanceExtraInitializers);
            __esDecorate(_a, null, _get_arc_decorators, { kind: "getter", name: "arc", static: false, private: false, access: { has: obj => "arc" in obj, get: obj => obj.arc }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
export { Ellipse };
let _Registered = false;
/**
 * Registers the Ellipse class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterEllipse() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.Ellipse", Ellipse);
}
