// @babylonjs/gui@9.16.1/3D/materials/fluentButton/fluentButtonMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { serializeAsColor4, serializeAsVector3, serialize } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { Vector3, Vector4 } from "@babylonjs/core/Maths/math.vector.pure.js";
import { Texture } from "@babylonjs/core/Materials/Textures/texture.pure.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { Color3, Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
import { Tools } from "@babylonjs/core/Misc/tools.pure.js";
/** @internal */
class FluentButtonMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        this.RELATIVE_WIDTH = true;
        this.ENABLE_FADE = true;
        this._needNormals = true;
        this._needUVs = true;
        this.rebuild();
    }
}
/**
 * Class used to render square buttons with fluent design
 * @since 5.0.0
 */
let FluentButtonMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _edgeWidth_decorators;
    let _edgeWidth_initializers = [];
    let _edgeWidth_extraInitializers = [];
    let _edgeColor_decorators;
    let _edgeColor_initializers = [];
    let _edgeColor_extraInitializers = [];
    let _proximityMaxIntensity_decorators;
    let _proximityMaxIntensity_initializers = [];
    let _proximityMaxIntensity_extraInitializers = [];
    let _proximityFarDistance_decorators;
    let _proximityFarDistance_initializers = [];
    let _proximityFarDistance_extraInitializers = [];
    let _proximityNearRadius_decorators;
    let _proximityNearRadius_initializers = [];
    let _proximityNearRadius_extraInitializers = [];
    let _proximityAnisotropy_decorators;
    let _proximityAnisotropy_initializers = [];
    let _proximityAnisotropy_extraInitializers = [];
    let _selectionFuzz_decorators;
    let _selectionFuzz_initializers = [];
    let _selectionFuzz_extraInitializers = [];
    let _selected_decorators;
    let _selected_initializers = [];
    let _selected_extraInitializers = [];
    let _selectionFade_decorators;
    let _selectionFade_initializers = [];
    let _selectionFade_extraInitializers = [];
    let _selectionFadeSize_decorators;
    let _selectionFadeSize_initializers = [];
    let _selectionFadeSize_extraInitializers = [];
    let _selectedDistance_decorators;
    let _selectedDistance_initializers = [];
    let _selectedDistance_extraInitializers = [];
    let _selectedFadeLength_decorators;
    let _selectedFadeLength_initializers = [];
    let _selectedFadeLength_extraInitializers = [];
    let _blobIntensity_decorators;
    let _blobIntensity_initializers = [];
    let _blobIntensity_extraInitializers = [];
    let _blobFarSize_decorators;
    let _blobFarSize_initializers = [];
    let _blobFarSize_extraInitializers = [];
    let _blobNearDistance_decorators;
    let _blobNearDistance_initializers = [];
    let _blobNearDistance_extraInitializers = [];
    let _blobFarDistance_decorators;
    let _blobFarDistance_initializers = [];
    let _blobFarDistance_extraInitializers = [];
    let _blobFadeLength_decorators;
    let _blobFadeLength_initializers = [];
    let _blobFadeLength_extraInitializers = [];
    let _leftBlobEnable_decorators;
    let _leftBlobEnable_initializers = [];
    let _leftBlobEnable_extraInitializers = [];
    let _leftBlobNearSize_decorators;
    let _leftBlobNearSize_initializers = [];
    let _leftBlobNearSize_extraInitializers = [];
    let _leftBlobPulse_decorators;
    let _leftBlobPulse_initializers = [];
    let _leftBlobPulse_extraInitializers = [];
    let _leftBlobFade_decorators;
    let _leftBlobFade_initializers = [];
    let _leftBlobFade_extraInitializers = [];
    let _leftBlobInnerFade_decorators;
    let _leftBlobInnerFade_initializers = [];
    let _leftBlobInnerFade_extraInitializers = [];
    let _rightBlobEnable_decorators;
    let _rightBlobEnable_initializers = [];
    let _rightBlobEnable_extraInitializers = [];
    let _rightBlobNearSize_decorators;
    let _rightBlobNearSize_initializers = [];
    let _rightBlobNearSize_extraInitializers = [];
    let _rightBlobPulse_decorators;
    let _rightBlobPulse_initializers = [];
    let _rightBlobPulse_extraInitializers = [];
    let _rightBlobFade_decorators;
    let _rightBlobFade_initializers = [];
    let _rightBlobFade_extraInitializers = [];
    let _rightBlobInnerFade_decorators;
    let _rightBlobInnerFade_initializers = [];
    let _rightBlobInnerFade_extraInitializers = [];
    let _activeFaceDir_decorators;
    let _activeFaceDir_initializers = [];
    let _activeFaceDir_extraInitializers = [];
    let _activeFaceUp_decorators;
    let _activeFaceUp_initializers = [];
    let _activeFaceUp_extraInitializers = [];
    let _enableFade_decorators;
    let _enableFade_initializers = [];
    let _enableFade_extraInitializers = [];
    let _fadeWidth_decorators;
    let _fadeWidth_initializers = [];
    let _fadeWidth_extraInitializers = [];
    let _smoothActiveFace_decorators;
    let _smoothActiveFace_initializers = [];
    let _smoothActiveFace_extraInitializers = [];
    let _showFrame_decorators;
    let _showFrame_initializers = [];
    let _showFrame_extraInitializers = [];
    let _useBlobTexture_decorators;
    let _useBlobTexture_initializers = [];
    let _useBlobTexture_extraInitializers = [];
    let _globalLeftIndexTipPosition_decorators;
    let _globalLeftIndexTipPosition_initializers = [];
    let _globalLeftIndexTipPosition_extraInitializers = [];
    let _globalRightIndexTipPosition_decorators;
    let _globalRightIndexTipPosition_initializers = [];
    let _globalRightIndexTipPosition_extraInitializers = [];
    return _a = class FluentButtonMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                this._shadersLoaded = false;
                /**
                 * Gets or sets the width of the glowing edge, relative to the scale of the button.
                 * (Default is 4% of the height).
                 */
                this.edgeWidth = __runInitializers(this, _edgeWidth_initializers, 0.04);
                /**
                 * Gets or sets the color of the glowing edge.
                 */
                this.edgeColor = (__runInitializers(this, _edgeWidth_extraInitializers), __runInitializers(this, _edgeColor_initializers, new Color4(0.592157, 0.592157, 0.592157, 1.0)));
                /**
                 * Gets or sets the maximum intensity of the proximity light.
                 */
                this.proximityMaxIntensity = (__runInitializers(this, _edgeColor_extraInitializers), __runInitializers(this, _proximityMaxIntensity_initializers, 0.45));
                /**
                 * Gets or sets the maximum distance for the proximity light (Default is 16mm).
                 */
                this.proximityFarDistance = (__runInitializers(this, _proximityMaxIntensity_extraInitializers), __runInitializers(this, _proximityFarDistance_initializers, 0.16));
                /**
                 * Gets or sets the radius of the proximity light when near to the surface.
                 */
                this.proximityNearRadius = (__runInitializers(this, _proximityFarDistance_extraInitializers), __runInitializers(this, _proximityNearRadius_initializers, 1.5));
                /**
                 * Gets or sets the anisotropy of the proximity light.
                 */
                this.proximityAnisotropy = (__runInitializers(this, _proximityNearRadius_extraInitializers), __runInitializers(this, _proximityAnisotropy_initializers, 1));
                /**
                 * Gets or sets the amount of fuzzing in the selection focus.
                 */
                this.selectionFuzz = (__runInitializers(this, _proximityAnisotropy_extraInitializers), __runInitializers(this, _selectionFuzz_initializers, 0.5));
                /**
                 * Gets or sets an override value to display the button as selected.
                 */
                this.selected = (__runInitializers(this, _selectionFuzz_extraInitializers), __runInitializers(this, _selected_initializers, 0));
                /**
                 * Gets or sets a value to manually fade the blob size.
                 */
                this.selectionFade = (__runInitializers(this, _selected_extraInitializers), __runInitializers(this, _selectionFade_initializers, 0));
                /**
                 * Gets or sets a value to manually shrink the blob size as it fades (see selectionFade).
                 */
                this.selectionFadeSize = (__runInitializers(this, _selectionFade_extraInitializers), __runInitializers(this, _selectionFadeSize_initializers, 0.3));
                /**
                 * Gets or sets the distance from the button the cursor should be for the button
                 * to appear selected (Default is 8cm).
                 */
                this.selectedDistance = (__runInitializers(this, _selectionFadeSize_extraInitializers), __runInitializers(this, _selectedDistance_initializers, 0.08));
                /**
                 * Gets or sets the fall-off distance for the selection fade (Default is 8cm).
                 */
                this.selectedFadeLength = (__runInitializers(this, _selectedDistance_extraInitializers), __runInitializers(this, _selectedFadeLength_initializers, 0.08));
                /**
                 * Gets or sets the intensity of the luminous blob (Ranges 0-1, default is 0.5).
                 */
                this.blobIntensity = (__runInitializers(this, _selectedFadeLength_extraInitializers), __runInitializers(this, _blobIntensity_initializers, 0.5));
                /**
                 * The size of the blob when the pointer is at the blobFarDistance (Default is 5cm).
                 */
                this.blobFarSize = (__runInitializers(this, _blobIntensity_extraInitializers), __runInitializers(this, _blobFarSize_initializers, 0.05));
                /**
                 * The distance at which the pointer is considered near. See [left|right]BlobNearSize. (Default is 0cm).
                 */
                this.blobNearDistance = (__runInitializers(this, _blobFarSize_extraInitializers), __runInitializers(this, _blobNearDistance_initializers, 0));
                /**
                 * The distance at which the pointer is considered far. See [left|right]BlobFarSize. (Default is 8cm).
                 */
                this.blobFarDistance = (__runInitializers(this, _blobNearDistance_extraInitializers), __runInitializers(this, _blobFarDistance_initializers, 0.08));
                /**
                 * The distance over which the blob intensity fades from full to none (Default is 8cm).
                 */
                this.blobFadeLength = (__runInitializers(this, _blobFarDistance_extraInitializers), __runInitializers(this, _blobFadeLength_initializers, 0.08));
                /**
                 * Gets or sets whether the blob corresponding to the left index finger is enabled.
                 */
                this.leftBlobEnable = (__runInitializers(this, _blobFadeLength_extraInitializers), __runInitializers(this, _leftBlobEnable_initializers, true));
                /**
                 * Gets or sets the size of the left blob when the left pointer is considered near. See blobNearDistance. (Default is 2.5cm).
                 */
                this.leftBlobNearSize = (__runInitializers(this, _leftBlobEnable_extraInitializers), __runInitializers(this, _leftBlobNearSize_initializers, 0.025));
                /**
                 * Gets or sets the progress of the pulse animation on the left blob (Ranges 0-1).
                 */
                this.leftBlobPulse = (__runInitializers(this, _leftBlobNearSize_extraInitializers), __runInitializers(this, _leftBlobPulse_initializers, 0));
                /**
                 * Gets or sets the fade factor on the left blob.
                 */
                this.leftBlobFade = (__runInitializers(this, _leftBlobPulse_extraInitializers), __runInitializers(this, _leftBlobFade_initializers, 1));
                /**
                 * Gets or sets the inner fade on the left blob;
                 */
                this.leftBlobInnerFade = (__runInitializers(this, _leftBlobFade_extraInitializers), __runInitializers(this, _leftBlobInnerFade_initializers, 0.01));
                /**
                 * Gets or sets whether the blob corresponding to the right index finger is enabled.
                 */
                this.rightBlobEnable = (__runInitializers(this, _leftBlobInnerFade_extraInitializers), __runInitializers(this, _rightBlobEnable_initializers, true));
                /**
                 * Gets or sets the size of the right blob when the right pointer is considered near. See blobNearDistance. (Default is 2.5cm).
                 */
                this.rightBlobNearSize = (__runInitializers(this, _rightBlobEnable_extraInitializers), __runInitializers(this, _rightBlobNearSize_initializers, 0.025));
                /**
                 * Gets or sets the progress of the pulse animation on the right blob (Ranges 0-1).
                 */
                this.rightBlobPulse = (__runInitializers(this, _rightBlobNearSize_extraInitializers), __runInitializers(this, _rightBlobPulse_initializers, 0));
                /**
                 * Gets or sets the fade factor on the right blob.
                 */
                this.rightBlobFade = (__runInitializers(this, _rightBlobPulse_extraInitializers), __runInitializers(this, _rightBlobFade_initializers, 1));
                /**
                 * Gets or sets the inner fade on the right blob;
                 */
                this.rightBlobInnerFade = (__runInitializers(this, _rightBlobFade_extraInitializers), __runInitializers(this, _rightBlobInnerFade_initializers, 0.01));
                /**
                 * Gets or sets the direction of the active face before the world transform is applied.
                 * This should almost always be set to -z.
                 */
                this.activeFaceDir = (__runInitializers(this, _rightBlobInnerFade_extraInitializers), __runInitializers(this, _activeFaceDir_initializers, new Vector3(0, 0, -1)));
                /**
                 * Gets or sets the button's up direction before the world transform is applied.
                 * This should almost always be set to +y.
                 */
                this.activeFaceUp = (__runInitializers(this, _activeFaceDir_extraInitializers), __runInitializers(this, _activeFaceUp_initializers, new Vector3(0, 1, 0)));
                /**
                 * Gets or sets whether the edge fade effect is enabled.
                 */
                this.enableFade = (__runInitializers(this, _activeFaceUp_extraInitializers), __runInitializers(this, _enableFade_initializers, true));
                /**
                 * Gets or sets a value corresponding to the width of the edge fade effect (Default 1.5).
                 */
                this.fadeWidth = (__runInitializers(this, _enableFade_extraInitializers), __runInitializers(this, _fadeWidth_initializers, 1.5));
                /**
                 * Gets or sets whether the active face is smoothly interpolated.
                 */
                this.smoothActiveFace = (__runInitializers(this, _fadeWidth_extraInitializers), __runInitializers(this, _smoothActiveFace_initializers, true));
                /**
                 * Gets or sets whether the frame of the fluent button model is visible.
                 * This is usually only enabled for debugging purposes.
                 */
                this.showFrame = (__runInitializers(this, _smoothActiveFace_extraInitializers), __runInitializers(this, _showFrame_initializers, false));
                /**
                 * Gets or sets whether the blob color texture is used for the proximity
                 * light effect. This is usually only disabled for debugging purposes.
                 */
                this.useBlobTexture = (__runInitializers(this, _showFrame_extraInitializers), __runInitializers(this, _useBlobTexture_initializers, true));
                /**
                 * Gets or sets the world-space position of the tip of the left index finger.
                 */
                this.globalLeftIndexTipPosition = (__runInitializers(this, _useBlobTexture_extraInitializers), __runInitializers(this, _globalLeftIndexTipPosition_initializers, Vector3.Zero()));
                /**
                 * Gets or sets the world-space position of the tip of the right index finger.
                 */
                this.globalRightIndexTipPosition = (__runInitializers(this, _globalLeftIndexTipPosition_extraInitializers), __runInitializers(this, _globalRightIndexTipPosition_initializers, Vector3.Zero()));
                this._blobTexture = __runInitializers(this, _globalRightIndexTipPosition_extraInitializers);
                this.alphaMode = Constants.ALPHA_ADD;
                this.disableDepthWrite = true;
                this.backFaceCulling = false;
                this._blobTextureUrl = Tools.GetAssetUrl(_a.BLOB_TEXTURE_URL);
                this._blobTexture = new Texture(this._blobTextureUrl, this.getScene(), true, false, Texture.NEAREST_SAMPLINGMODE);
            }
            needAlphaBlending() {
                return true;
            }
            needAlphaTesting() {
                return true;
            }
            getAlphaTestTexture() {
                return null;
            }
            // Methods
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new FluentButtonMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                const blobTextureError = this._blobTexture.errorObject;
                if (this._blobTexture.loadingError || blobTextureError) {
                    const textureErrorMessage = blobTextureError?.message || (blobTextureError?.exception instanceof Error ? blobTextureError.exception.message : undefined);
                    throw new Error(`FluentButtonMaterial "${this.name}" failed to load blob texture "${this._blobTextureUrl}"${textureErrorMessage ? `: ${textureErrorMessage}` : ""}. Check FluentButtonMaterial.BLOB_TEXTURE_URL and asset availability.`, { cause: blobTextureError?.exception });
                }
                if (!this._blobTexture.isReady()) {
                    return false;
                }
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, true, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "fluentButton";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "viewProjection",
                        "cameraPosition",
                        "_Edge_Width_",
                        "_Edge_Color_",
                        "_Relative_Width_",
                        "_Proximity_Max_Intensity_",
                        "_Proximity_Far_Distance_",
                        "_Proximity_Near_Radius_",
                        "_Proximity_Anisotropy_",
                        "_Selection_Fuzz_",
                        "_Selected_",
                        "_Selection_Fade_",
                        "_Selection_Fade_Size_",
                        "_Selected_Distance_",
                        "_Selected_Fade_Length_",
                        "_Blob_Enable_",
                        "_Blob_Position_",
                        "_Blob_Intensity_",
                        "_Blob_Near_Size_",
                        "_Blob_Far_Size_",
                        "_Blob_Near_Distance_",
                        "_Blob_Far_Distance_",
                        "_Blob_Fade_Length_",
                        "_Blob_Inner_Fade_",
                        "_Blob_Pulse_",
                        "_Blob_Fade_",
                        "_Blob_Texture_",
                        "_Blob_Enable_2_",
                        "_Blob_Position_2_",
                        "_Blob_Near_Size_2_",
                        "_Blob_Inner_Fade_2_",
                        "_Blob_Pulse_2_",
                        "_Blob_Fade_2_",
                        "_Active_Face_Dir_",
                        "_Active_Face_Up_",
                        "_Enable_Fade_",
                        "_Fade_Width_",
                        "_Smooth_Active_Face_",
                        "_Show_Frame_",
                        "_Use_Blob_Texture_",
                        "Use_Global_Left_Index",
                        "Use_Global_Right_Index",
                        "Global_Left_Index_Tip_Position",
                        "Global_Right_Index_Tip_Position",
                        "Global_Left_Thumb_Tip_Position",
                        "Global_Right_Thumb_Tip_Position",
                        "Global_Left_Index_Tip_Proximity",
                        "Global_Right_Index_Tip_Proximity",
                    ];
                    const samplers = ["_Blob_Texture_"];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                        shaderLanguage: this._shaderLanguage,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        shaderLanguage: this._shaderLanguage,
                        extraInitializationsAsync: this._shadersLoaded
                            ? undefined
                            : async () => {
                                if (this.shaderLanguage === 1 /* ShaderLanguage.WGSL */) {
                                    await Promise.all([import("./wgsl/fluentButton.vertex.js"), import("./wgsl/fluentButton.fragment.js")]);
                                }
                                else {
                                    await Promise.all([import("./shaders/fluentButton.vertex.js"), import("./shaders/fluentButton.fragment.js")]);
                                }
                                this._shadersLoaded = true;
                            },
                    }, engine), defines, this._materialContext);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const scene = this.getScene();
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", scene.getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", scene.activeCamera.position);
                // "Blob Texture"
                this._activeEffect.setTexture("_Blob_Texture_", this._blobTexture);
                // "Wireframe"
                this._activeEffect.setFloat("_Edge_Width_", this.edgeWidth);
                this._activeEffect.setColor4("_Edge_Color_", new Color3(this.edgeColor.r, this.edgeColor.g, this.edgeColor.b), this.edgeColor.a);
                //define _Relative_Width_ true;
                // "Proximity"
                this._activeEffect.setFloat("_Proximity_Max_Intensity_", this.proximityMaxIntensity);
                this._activeEffect.setFloat("_Proximity_Far_Distance_", this.proximityFarDistance);
                this._activeEffect.setFloat("_Proximity_Near_Radius_", this.proximityNearRadius);
                this._activeEffect.setFloat("_Proximity_Anisotropy_", this.proximityAnisotropy);
                // "Selection"
                this._activeEffect.setFloat("_Selection_Fuzz_", this.selectionFuzz);
                this._activeEffect.setFloat("_Selected_", this.selected);
                this._activeEffect.setFloat("_Selection_Fade_", this.selectionFade);
                this._activeEffect.setFloat("_Selection_Fade_Size_", this.selectionFadeSize);
                this._activeEffect.setFloat("_Selected_Distance_", this.selectedDistance);
                this._activeEffect.setFloat("_Selected_Fade_Length_", this.selectedFadeLength);
                // "Blob"
                this._activeEffect.setFloat("_Blob_Enable_", this.leftBlobEnable ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Blob_Intensity_", this.blobIntensity);
                this._activeEffect.setFloat("_Blob_Near_Size_", this.leftBlobNearSize);
                this._activeEffect.setFloat("_Blob_Far_Size_", this.blobFarSize);
                this._activeEffect.setFloat("_Blob_Near_Distance_", this.blobNearDistance);
                this._activeEffect.setFloat("_Blob_Far_Distance_", this.blobFarDistance);
                this._activeEffect.setFloat("_Blob_Fade_Length_", this.blobFadeLength);
                this._activeEffect.setFloat("_Blob_Inner_Fade_", this.leftBlobInnerFade);
                this._activeEffect.setFloat("_Blob_Pulse_", this.leftBlobPulse);
                this._activeEffect.setFloat("_Blob_Fade_", this.leftBlobFade);
                // "Blob 2"
                this._activeEffect.setFloat("_Blob_Enable_2_", this.rightBlobEnable ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Blob_Near_Size_2_", this.rightBlobNearSize);
                this._activeEffect.setFloat("_Blob_Inner_Fade_2_", this.rightBlobInnerFade);
                this._activeEffect.setFloat("_Blob_Pulse_2_", this.rightBlobPulse);
                this._activeEffect.setFloat("_Blob_Fade_2_", this.rightBlobFade);
                // "Active Face"
                this._activeEffect.setVector3("_Active_Face_Dir_", this.activeFaceDir);
                this._activeEffect.setVector3("_Active_Face_Up_", this.activeFaceUp);
                // "Hololens Edge Fade"
                //define _Enable_Fade_ true;
                this._activeEffect.setFloat("_Fade_Width_", this.fadeWidth);
                this._activeEffect.setFloat("_Smooth_Active_Face_", this.smoothActiveFace ? 1.0 : 0.0);
                // "Debug"
                this._activeEffect.setFloat("_Show_Frame_", this.showFrame ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Use_Blob_Texture_", this.useBlobTexture ? 1.0 : 0.0);
                // Global inputs
                this._activeEffect.setFloat("Use_Global_Left_Index", 1.0);
                this._activeEffect.setFloat("Use_Global_Right_Index", 1.0);
                this._activeEffect.setVector4("Global_Left_Index_Tip_Position", new Vector4(this.globalLeftIndexTipPosition.x, this.globalLeftIndexTipPosition.y, this.globalLeftIndexTipPosition.z, 1.0));
                this._activeEffect.setVector4("Global_Right_Index_Tip_Position", new Vector4(this.globalRightIndexTipPosition.x, this.globalRightIndexTipPosition.y, this.globalRightIndexTipPosition.z, 1.0));
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            /**
             * Disposes the material.
             * @param forceDisposeEffect specifies if effects should be forcefully disposed
             */
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = super.serialize();
                serializationObject.customType = "BABYLON.FluentButtonMaterial";
                return serializationObject;
            }
            getClassName() {
                return "FluentButtonMaterial";
            }
            // Statics
            /**
             * Creates a fluent button material from parsed material data.
             * @param source defines the JSON representation of the material
             * @param scene defines the hosting scene
             * @param rootUrl defines the root URL to use to load textures and relative dependencies
             * @returns a new fluent button material
             */
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _edgeWidth_decorators = [serialize()];
            _edgeColor_decorators = [serializeAsColor4()];
            _proximityMaxIntensity_decorators = [serialize()];
            _proximityFarDistance_decorators = [serialize()];
            _proximityNearRadius_decorators = [serialize()];
            _proximityAnisotropy_decorators = [serialize()];
            _selectionFuzz_decorators = [serialize()];
            _selected_decorators = [serialize()];
            _selectionFade_decorators = [serialize()];
            _selectionFadeSize_decorators = [serialize()];
            _selectedDistance_decorators = [serialize()];
            _selectedFadeLength_decorators = [serialize()];
            _blobIntensity_decorators = [serialize()];
            _blobFarSize_decorators = [serialize()];
            _blobNearDistance_decorators = [serialize()];
            _blobFarDistance_decorators = [serialize()];
            _blobFadeLength_decorators = [serialize()];
            _leftBlobEnable_decorators = [serialize()];
            _leftBlobNearSize_decorators = [serialize()];
            _leftBlobPulse_decorators = [serialize()];
            _leftBlobFade_decorators = [serialize()];
            _leftBlobInnerFade_decorators = [serialize()];
            _rightBlobEnable_decorators = [serialize()];
            _rightBlobNearSize_decorators = [serialize()];
            _rightBlobPulse_decorators = [serialize()];
            _rightBlobFade_decorators = [serialize()];
            _rightBlobInnerFade_decorators = [serialize()];
            _activeFaceDir_decorators = [serializeAsVector3()];
            _activeFaceUp_decorators = [serializeAsVector3()];
            _enableFade_decorators = [serialize()];
            _fadeWidth_decorators = [serialize()];
            _smoothActiveFace_decorators = [serialize()];
            _showFrame_decorators = [serialize()];
            _useBlobTexture_decorators = [serialize()];
            _globalLeftIndexTipPosition_decorators = [serializeAsVector3()];
            _globalRightIndexTipPosition_decorators = [serializeAsVector3()];
            __esDecorate(null, null, _edgeWidth_decorators, { kind: "field", name: "edgeWidth", static: false, private: false, access: { has: obj => "edgeWidth" in obj, get: obj => obj.edgeWidth, set: (obj, value) => { obj.edgeWidth = value; } }, metadata: _metadata }, _edgeWidth_initializers, _edgeWidth_extraInitializers);
            __esDecorate(null, null, _edgeColor_decorators, { kind: "field", name: "edgeColor", static: false, private: false, access: { has: obj => "edgeColor" in obj, get: obj => obj.edgeColor, set: (obj, value) => { obj.edgeColor = value; } }, metadata: _metadata }, _edgeColor_initializers, _edgeColor_extraInitializers);
            __esDecorate(null, null, _proximityMaxIntensity_decorators, { kind: "field", name: "proximityMaxIntensity", static: false, private: false, access: { has: obj => "proximityMaxIntensity" in obj, get: obj => obj.proximityMaxIntensity, set: (obj, value) => { obj.proximityMaxIntensity = value; } }, metadata: _metadata }, _proximityMaxIntensity_initializers, _proximityMaxIntensity_extraInitializers);
            __esDecorate(null, null, _proximityFarDistance_decorators, { kind: "field", name: "proximityFarDistance", static: false, private: false, access: { has: obj => "proximityFarDistance" in obj, get: obj => obj.proximityFarDistance, set: (obj, value) => { obj.proximityFarDistance = value; } }, metadata: _metadata }, _proximityFarDistance_initializers, _proximityFarDistance_extraInitializers);
            __esDecorate(null, null, _proximityNearRadius_decorators, { kind: "field", name: "proximityNearRadius", static: false, private: false, access: { has: obj => "proximityNearRadius" in obj, get: obj => obj.proximityNearRadius, set: (obj, value) => { obj.proximityNearRadius = value; } }, metadata: _metadata }, _proximityNearRadius_initializers, _proximityNearRadius_extraInitializers);
            __esDecorate(null, null, _proximityAnisotropy_decorators, { kind: "field", name: "proximityAnisotropy", static: false, private: false, access: { has: obj => "proximityAnisotropy" in obj, get: obj => obj.proximityAnisotropy, set: (obj, value) => { obj.proximityAnisotropy = value; } }, metadata: _metadata }, _proximityAnisotropy_initializers, _proximityAnisotropy_extraInitializers);
            __esDecorate(null, null, _selectionFuzz_decorators, { kind: "field", name: "selectionFuzz", static: false, private: false, access: { has: obj => "selectionFuzz" in obj, get: obj => obj.selectionFuzz, set: (obj, value) => { obj.selectionFuzz = value; } }, metadata: _metadata }, _selectionFuzz_initializers, _selectionFuzz_extraInitializers);
            __esDecorate(null, null, _selected_decorators, { kind: "field", name: "selected", static: false, private: false, access: { has: obj => "selected" in obj, get: obj => obj.selected, set: (obj, value) => { obj.selected = value; } }, metadata: _metadata }, _selected_initializers, _selected_extraInitializers);
            __esDecorate(null, null, _selectionFade_decorators, { kind: "field", name: "selectionFade", static: false, private: false, access: { has: obj => "selectionFade" in obj, get: obj => obj.selectionFade, set: (obj, value) => { obj.selectionFade = value; } }, metadata: _metadata }, _selectionFade_initializers, _selectionFade_extraInitializers);
            __esDecorate(null, null, _selectionFadeSize_decorators, { kind: "field", name: "selectionFadeSize", static: false, private: false, access: { has: obj => "selectionFadeSize" in obj, get: obj => obj.selectionFadeSize, set: (obj, value) => { obj.selectionFadeSize = value; } }, metadata: _metadata }, _selectionFadeSize_initializers, _selectionFadeSize_extraInitializers);
            __esDecorate(null, null, _selectedDistance_decorators, { kind: "field", name: "selectedDistance", static: false, private: false, access: { has: obj => "selectedDistance" in obj, get: obj => obj.selectedDistance, set: (obj, value) => { obj.selectedDistance = value; } }, metadata: _metadata }, _selectedDistance_initializers, _selectedDistance_extraInitializers);
            __esDecorate(null, null, _selectedFadeLength_decorators, { kind: "field", name: "selectedFadeLength", static: false, private: false, access: { has: obj => "selectedFadeLength" in obj, get: obj => obj.selectedFadeLength, set: (obj, value) => { obj.selectedFadeLength = value; } }, metadata: _metadata }, _selectedFadeLength_initializers, _selectedFadeLength_extraInitializers);
            __esDecorate(null, null, _blobIntensity_decorators, { kind: "field", name: "blobIntensity", static: false, private: false, access: { has: obj => "blobIntensity" in obj, get: obj => obj.blobIntensity, set: (obj, value) => { obj.blobIntensity = value; } }, metadata: _metadata }, _blobIntensity_initializers, _blobIntensity_extraInitializers);
            __esDecorate(null, null, _blobFarSize_decorators, { kind: "field", name: "blobFarSize", static: false, private: false, access: { has: obj => "blobFarSize" in obj, get: obj => obj.blobFarSize, set: (obj, value) => { obj.blobFarSize = value; } }, metadata: _metadata }, _blobFarSize_initializers, _blobFarSize_extraInitializers);
            __esDecorate(null, null, _blobNearDistance_decorators, { kind: "field", name: "blobNearDistance", static: false, private: false, access: { has: obj => "blobNearDistance" in obj, get: obj => obj.blobNearDistance, set: (obj, value) => { obj.blobNearDistance = value; } }, metadata: _metadata }, _blobNearDistance_initializers, _blobNearDistance_extraInitializers);
            __esDecorate(null, null, _blobFarDistance_decorators, { kind: "field", name: "blobFarDistance", static: false, private: false, access: { has: obj => "blobFarDistance" in obj, get: obj => obj.blobFarDistance, set: (obj, value) => { obj.blobFarDistance = value; } }, metadata: _metadata }, _blobFarDistance_initializers, _blobFarDistance_extraInitializers);
            __esDecorate(null, null, _blobFadeLength_decorators, { kind: "field", name: "blobFadeLength", static: false, private: false, access: { has: obj => "blobFadeLength" in obj, get: obj => obj.blobFadeLength, set: (obj, value) => { obj.blobFadeLength = value; } }, metadata: _metadata }, _blobFadeLength_initializers, _blobFadeLength_extraInitializers);
            __esDecorate(null, null, _leftBlobEnable_decorators, { kind: "field", name: "leftBlobEnable", static: false, private: false, access: { has: obj => "leftBlobEnable" in obj, get: obj => obj.leftBlobEnable, set: (obj, value) => { obj.leftBlobEnable = value; } }, metadata: _metadata }, _leftBlobEnable_initializers, _leftBlobEnable_extraInitializers);
            __esDecorate(null, null, _leftBlobNearSize_decorators, { kind: "field", name: "leftBlobNearSize", static: false, private: false, access: { has: obj => "leftBlobNearSize" in obj, get: obj => obj.leftBlobNearSize, set: (obj, value) => { obj.leftBlobNearSize = value; } }, metadata: _metadata }, _leftBlobNearSize_initializers, _leftBlobNearSize_extraInitializers);
            __esDecorate(null, null, _leftBlobPulse_decorators, { kind: "field", name: "leftBlobPulse", static: false, private: false, access: { has: obj => "leftBlobPulse" in obj, get: obj => obj.leftBlobPulse, set: (obj, value) => { obj.leftBlobPulse = value; } }, metadata: _metadata }, _leftBlobPulse_initializers, _leftBlobPulse_extraInitializers);
            __esDecorate(null, null, _leftBlobFade_decorators, { kind: "field", name: "leftBlobFade", static: false, private: false, access: { has: obj => "leftBlobFade" in obj, get: obj => obj.leftBlobFade, set: (obj, value) => { obj.leftBlobFade = value; } }, metadata: _metadata }, _leftBlobFade_initializers, _leftBlobFade_extraInitializers);
            __esDecorate(null, null, _leftBlobInnerFade_decorators, { kind: "field", name: "leftBlobInnerFade", static: false, private: false, access: { has: obj => "leftBlobInnerFade" in obj, get: obj => obj.leftBlobInnerFade, set: (obj, value) => { obj.leftBlobInnerFade = value; } }, metadata: _metadata }, _leftBlobInnerFade_initializers, _leftBlobInnerFade_extraInitializers);
            __esDecorate(null, null, _rightBlobEnable_decorators, { kind: "field", name: "rightBlobEnable", static: false, private: false, access: { has: obj => "rightBlobEnable" in obj, get: obj => obj.rightBlobEnable, set: (obj, value) => { obj.rightBlobEnable = value; } }, metadata: _metadata }, _rightBlobEnable_initializers, _rightBlobEnable_extraInitializers);
            __esDecorate(null, null, _rightBlobNearSize_decorators, { kind: "field", name: "rightBlobNearSize", static: false, private: false, access: { has: obj => "rightBlobNearSize" in obj, get: obj => obj.rightBlobNearSize, set: (obj, value) => { obj.rightBlobNearSize = value; } }, metadata: _metadata }, _rightBlobNearSize_initializers, _rightBlobNearSize_extraInitializers);
            __esDecorate(null, null, _rightBlobPulse_decorators, { kind: "field", name: "rightBlobPulse", static: false, private: false, access: { has: obj => "rightBlobPulse" in obj, get: obj => obj.rightBlobPulse, set: (obj, value) => { obj.rightBlobPulse = value; } }, metadata: _metadata }, _rightBlobPulse_initializers, _rightBlobPulse_extraInitializers);
            __esDecorate(null, null, _rightBlobFade_decorators, { kind: "field", name: "rightBlobFade", static: false, private: false, access: { has: obj => "rightBlobFade" in obj, get: obj => obj.rightBlobFade, set: (obj, value) => { obj.rightBlobFade = value; } }, metadata: _metadata }, _rightBlobFade_initializers, _rightBlobFade_extraInitializers);
            __esDecorate(null, null, _rightBlobInnerFade_decorators, { kind: "field", name: "rightBlobInnerFade", static: false, private: false, access: { has: obj => "rightBlobInnerFade" in obj, get: obj => obj.rightBlobInnerFade, set: (obj, value) => { obj.rightBlobInnerFade = value; } }, metadata: _metadata }, _rightBlobInnerFade_initializers, _rightBlobInnerFade_extraInitializers);
            __esDecorate(null, null, _activeFaceDir_decorators, { kind: "field", name: "activeFaceDir", static: false, private: false, access: { has: obj => "activeFaceDir" in obj, get: obj => obj.activeFaceDir, set: (obj, value) => { obj.activeFaceDir = value; } }, metadata: _metadata }, _activeFaceDir_initializers, _activeFaceDir_extraInitializers);
            __esDecorate(null, null, _activeFaceUp_decorators, { kind: "field", name: "activeFaceUp", static: false, private: false, access: { has: obj => "activeFaceUp" in obj, get: obj => obj.activeFaceUp, set: (obj, value) => { obj.activeFaceUp = value; } }, metadata: _metadata }, _activeFaceUp_initializers, _activeFaceUp_extraInitializers);
            __esDecorate(null, null, _enableFade_decorators, { kind: "field", name: "enableFade", static: false, private: false, access: { has: obj => "enableFade" in obj, get: obj => obj.enableFade, set: (obj, value) => { obj.enableFade = value; } }, metadata: _metadata }, _enableFade_initializers, _enableFade_extraInitializers);
            __esDecorate(null, null, _fadeWidth_decorators, { kind: "field", name: "fadeWidth", static: false, private: false, access: { has: obj => "fadeWidth" in obj, get: obj => obj.fadeWidth, set: (obj, value) => { obj.fadeWidth = value; } }, metadata: _metadata }, _fadeWidth_initializers, _fadeWidth_extraInitializers);
            __esDecorate(null, null, _smoothActiveFace_decorators, { kind: "field", name: "smoothActiveFace", static: false, private: false, access: { has: obj => "smoothActiveFace" in obj, get: obj => obj.smoothActiveFace, set: (obj, value) => { obj.smoothActiveFace = value; } }, metadata: _metadata }, _smoothActiveFace_initializers, _smoothActiveFace_extraInitializers);
            __esDecorate(null, null, _showFrame_decorators, { kind: "field", name: "showFrame", static: false, private: false, access: { has: obj => "showFrame" in obj, get: obj => obj.showFrame, set: (obj, value) => { obj.showFrame = value; } }, metadata: _metadata }, _showFrame_initializers, _showFrame_extraInitializers);
            __esDecorate(null, null, _useBlobTexture_decorators, { kind: "field", name: "useBlobTexture", static: false, private: false, access: { has: obj => "useBlobTexture" in obj, get: obj => obj.useBlobTexture, set: (obj, value) => { obj.useBlobTexture = value; } }, metadata: _metadata }, _useBlobTexture_initializers, _useBlobTexture_extraInitializers);
            __esDecorate(null, null, _globalLeftIndexTipPosition_decorators, { kind: "field", name: "globalLeftIndexTipPosition", static: false, private: false, access: { has: obj => "globalLeftIndexTipPosition" in obj, get: obj => obj.globalLeftIndexTipPosition, set: (obj, value) => { obj.globalLeftIndexTipPosition = value; } }, metadata: _metadata }, _globalLeftIndexTipPosition_initializers, _globalLeftIndexTipPosition_extraInitializers);
            __esDecorate(null, null, _globalRightIndexTipPosition_decorators, { kind: "field", name: "globalRightIndexTipPosition", static: false, private: false, access: { has: obj => "globalRightIndexTipPosition" in obj, get: obj => obj.globalRightIndexTipPosition, set: (obj, value) => { obj.globalRightIndexTipPosition = value; } }, metadata: _metadata }, _globalRightIndexTipPosition_initializers, _globalRightIndexTipPosition_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        /**
         * URL pointing to the texture used to define the coloring for the fluent blob effect.
         */
        _a.BLOB_TEXTURE_URL = "https://assets.babylonjs.com/core/MRTK/mrtk-fluent-button-blob.png",
        _a;
})();
export { FluentButtonMaterial };
let _Registered = false;
/**
 * Registers the FluentButtonMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterFluentButtonMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.FluentButtonMaterial", FluentButtonMaterial);
}
