// @babylonjs/gui@9.16.1/3D/materials/fluentBackplate/fluentBackplateMaterial.pure.js
import { __esDecorate, __runInitializers } from "@babylonjs/core/tslib.es6.js";
import { serialize, serializeAsVector3 } from "@babylonjs/core/Misc/decorators.js";
import { SerializationHelper } from "@babylonjs/core/Misc/decorators.serialization.js";
import { Vector3, Vector4 } from "@babylonjs/core/Maths/math.vector.pure.js";
import { Texture } from "@babylonjs/core/Materials/Textures/texture.pure.js";
import { MaterialDefines } from "@babylonjs/core/Materials/materialDefines.js";
import { PushMaterial } from "@babylonjs/core/Materials/pushMaterial.js";
import { VertexBuffer } from "@babylonjs/core/Buffers/buffer.pure.js";
import { RegisterClass } from "@babylonjs/core/Misc/typeStore.js";
import { Color4 } from "@babylonjs/core/Maths/math.color.pure.js";
import { EffectFallbacks } from "@babylonjs/core/Materials/effectFallbacks.js";
import { Constants } from "@babylonjs/core/Engines/constants.js";
import { HandleFallbacksForShadows, PrepareAttributesForInstances, PrepareDefinesForAttributes, PrepareUniformsAndSamplersList } from "@babylonjs/core/Materials/materialHelper.functions.js";
import { Tools } from "@babylonjs/core/Misc/tools.pure.js";
/** @internal */
class FluentBackplateMaterialDefines extends MaterialDefines {
    constructor() {
        super();
        this.BLOB_ENABLE = true;
        this.BLOB_ENABLE_2 = true;
        this.SMOOTH_EDGES = true;
        this.IRIDESCENT_MAP_ENABLE = true;
        this._needNormals = true;
        this.rebuild();
    }
}
/**
 * Class used to render square buttons with fluent design
 */
let FluentBackplateMaterial = (() => {
    var _a;
    let _classSuper = PushMaterial;
    let _radius_decorators;
    let _radius_initializers = [];
    let _radius_extraInitializers = [];
    let _lineWidth_decorators;
    let _lineWidth_initializers = [];
    let _lineWidth_extraInitializers = [];
    let _absoluteSizes_decorators;
    let _absoluteSizes_initializers = [];
    let _absoluteSizes_extraInitializers = [];
    let _baseColor_decorators;
    let _baseColor_initializers = [];
    let _baseColor_extraInitializers = [];
    let _lineColor_decorators;
    let _lineColor_initializers = [];
    let _lineColor_extraInitializers = [];
    let _blobIntensity_decorators;
    let _blobIntensity_initializers = [];
    let _blobIntensity_extraInitializers = [];
    let _blobFarSize_decorators;
    let _blobFarSize_initializers = [];
    let _blobFarSize_extraInitializers = [];
    let _blobNearDistance_decorators;
    let _blobNearDistance_initializers = [];
    let _blobNearDistance_extraInitializers = [];
    let _blobFarDistance_decorators;
    let _blobFarDistance_initializers = [];
    let _blobFarDistance_extraInitializers = [];
    let _blobFadeLength_decorators;
    let _blobFadeLength_initializers = [];
    let _blobFadeLength_extraInitializers = [];
    let _blobNearSize_decorators;
    let _blobNearSize_initializers = [];
    let _blobNearSize_extraInitializers = [];
    let _blobPulse_decorators;
    let _blobPulse_initializers = [];
    let _blobPulse_extraInitializers = [];
    let _blobFade_decorators;
    let _blobFade_initializers = [];
    let _blobFade_extraInitializers = [];
    let _blobNearSize2_decorators;
    let _blobNearSize2_initializers = [];
    let _blobNearSize2_extraInitializers = [];
    let _blobPulse2_decorators;
    let _blobPulse2_initializers = [];
    let _blobPulse2_extraInitializers = [];
    let _blobFade2_decorators;
    let _blobFade2_initializers = [];
    let _blobFade2_extraInitializers = [];
    let _highlightColor_decorators;
    let _highlightColor_initializers = [];
    let _highlightColor_extraInitializers = [];
    let _highlightWidth_decorators;
    let _highlightWidth_initializers = [];
    let _highlightWidth_extraInitializers = [];
    let _iridescenceIntensity_decorators;
    let _iridescenceIntensity_initializers = [];
    let _iridescenceIntensity_extraInitializers = [];
    let _iridescenceEdgeIntensity_decorators;
    let _iridescenceEdgeIntensity_initializers = [];
    let _iridescenceEdgeIntensity_extraInitializers = [];
    let _fadeOut_decorators;
    let _fadeOut_initializers = [];
    let _fadeOut_extraInitializers = [];
    let _globalLeftIndexTipPosition_decorators;
    let _globalLeftIndexTipPosition_initializers = [];
    let _globalLeftIndexTipPosition_extraInitializers = [];
    let _globalRightIndexTipPosition_decorators;
    let _globalRightIndexTipPosition_initializers = [];
    let _globalRightIndexTipPosition_extraInitializers = [];
    return _a = class FluentBackplateMaterial extends _classSuper {
            constructor(name, scene) {
                super(name, scene);
                this._shadersLoaded = false;
                /**
                 * Gets or sets the corner radius on the backplate. Best to keep this value between 0.01 and 0.5. Default is 0.03.
                 */
                this.radius = __runInitializers(this, _radius_initializers, 0.03);
                /**
                 * Gets or sets the line width of the backplate.
                 */
                this.lineWidth = (__runInitializers(this, _radius_extraInitializers), __runInitializers(this, _lineWidth_initializers, 0.01));
                /**
                 * Gets or sets whether to use absolute sizes when calculating effects on the backplate.
                 * Since desktop and VR/AR have different relative sizes, it's usually best to keep this false.
                 */
                this.absoluteSizes = (__runInitializers(this, _lineWidth_extraInitializers), __runInitializers(this, _absoluteSizes_initializers, false));
                /** @internal */
                this._filterWidth = (__runInitializers(this, _absoluteSizes_extraInitializers), 1);
                /**
                 * Gets or sets the base color of the backplate.
                 */
                this.baseColor = __runInitializers(this, _baseColor_initializers, new Color4(0.0392157, 0.0666667, 0.207843, 1));
                /**
                 * Gets or sets the line color of the backplate.
                 */
                this.lineColor = (__runInitializers(this, _baseColor_extraInitializers), __runInitializers(this, _lineColor_initializers, new Color4(0.14902, 0.133333, 0.384314, 1)));
                /**
                 * Gets or sets the intensity of the fluent hover glow effect.
                 */
                this.blobIntensity = (__runInitializers(this, _lineColor_extraInitializers), __runInitializers(this, _blobIntensity_initializers, 0.98));
                /**
                 * Gets or sets the far size of the fluent hover glow effect.
                 */
                this.blobFarSize = (__runInitializers(this, _blobIntensity_extraInitializers), __runInitializers(this, _blobFarSize_initializers, 0.04));
                /**
                 * Gets or sets the distance considered "near" to the backplate, which controls the size of the fluent hover glow effect (see blobNearSize).
                 */
                this.blobNearDistance = (__runInitializers(this, _blobFarSize_extraInitializers), __runInitializers(this, _blobNearDistance_initializers, 0));
                /**
                 * Gets or sets the distance considered "far" from the backplate, which controls the size of the fluent hover glow effect (see blobFarSize).
                 */
                this.blobFarDistance = (__runInitializers(this, _blobNearDistance_extraInitializers), __runInitializers(this, _blobFarDistance_initializers, 0.08));
                /**
                 * Gets or sets the length of the fluent hover glow effect fade.
                 */
                this.blobFadeLength = (__runInitializers(this, _blobFarDistance_extraInitializers), __runInitializers(this, _blobFadeLength_initializers, 0.08));
                /**
                 * Gets or sets the size of the fluent hover glow effect when the left pointer is considered "near" to the backplate (see blobNearDistance).
                 */
                this.blobNearSize = (__runInitializers(this, _blobFadeLength_extraInitializers), __runInitializers(this, _blobNearSize_initializers, 0.22));
                /**
                 * Gets or sets the progress of the fluent hover glow effect selection animation corresponding to the left pointer (0.0 - 1.0).
                 */
                this.blobPulse = (__runInitializers(this, _blobNearSize_extraInitializers), __runInitializers(this, _blobPulse_initializers, 0));
                /**
                 * Gets or sets the opacity of the fluent hover glow effect corresponding to the left pointer (0.0 - 1.0). Default is 0.
                 */
                this.blobFade = (__runInitializers(this, _blobPulse_extraInitializers), __runInitializers(this, _blobFade_initializers, 0));
                /**
                 * Gets or sets the size of the fluent hover glow effect when the right pointer is considered "near" to the backplate (see blobNearDistance).
                 */
                this.blobNearSize2 = (__runInitializers(this, _blobFade_extraInitializers), __runInitializers(this, _blobNearSize2_initializers, 0.22));
                /**
                 * Gets or sets the progress of the fluent hover glow effect selection animation corresponding to the right pointer (0.0 - 1.0).
                 */
                this.blobPulse2 = (__runInitializers(this, _blobNearSize2_extraInitializers), __runInitializers(this, _blobPulse2_initializers, 0));
                /**
                 * Gets or sets the opacity of the fluent hover glow effect corresponding to the right pointer (0.0 - 1.0). Default is 0.
                 */
                this.blobFade2 = (__runInitializers(this, _blobPulse2_extraInitializers), __runInitializers(this, _blobFade2_initializers, 0));
                /** @internal */
                this._rate = (__runInitializers(this, _blobFade2_extraInitializers), 0.135);
                /**
                 * Gets or sets the color of the highlights on the backplate line.
                 */
                this.highlightColor = __runInitializers(this, _highlightColor_initializers, new Color4(0.98, 0.98, 0.98, 1));
                /**
                 * Gets or sets the width of the highlights on the backplate line.
                 */
                this.highlightWidth = (__runInitializers(this, _highlightColor_extraInitializers), __runInitializers(this, _highlightWidth_initializers, 0.25));
                /** @internal */
                this._highlightTransform = (__runInitializers(this, _highlightWidth_extraInitializers), new Vector4(1, 1, 0, 0));
                /** @internal */
                this._highlight = 1;
                /**
                 * Gets or sets the intensity of the iridescence effect.
                 */
                this.iridescenceIntensity = __runInitializers(this, _iridescenceIntensity_initializers, 0);
                /**
                 * Gets or sets the intensity of the iridescence effect on the backplate edges.
                 */
                this.iridescenceEdgeIntensity = (__runInitializers(this, _iridescenceIntensity_extraInitializers), __runInitializers(this, _iridescenceEdgeIntensity_initializers, 1));
                /** @internal */
                this._angle = (__runInitializers(this, _iridescenceEdgeIntensity_extraInitializers), -45);
                /**
                 * Gets or sets the opacity of the backplate (0.0 - 1.0).
                 */
                this.fadeOut = __runInitializers(this, _fadeOut_initializers, 1);
                /** @internal */
                this._reflected = (__runInitializers(this, _fadeOut_extraInitializers), true);
                /** @internal */
                this._frequency = 1;
                /** @internal */
                this._verticalOffset = 0;
                /**
                 * Gets or sets the world-space position of the tip of the left index finger.
                 */
                this.globalLeftIndexTipPosition = __runInitializers(this, _globalLeftIndexTipPosition_initializers, Vector3.Zero());
                this._globalLeftIndexTipPosition4 = (__runInitializers(this, _globalLeftIndexTipPosition_extraInitializers), Vector4.Zero());
                /**
                 * Gets or sets the world-space position of the tip of the right index finger.
                 */
                this.globalRightIndexTipPosition = __runInitializers(this, _globalRightIndexTipPosition_initializers, Vector3.Zero());
                this._globalRightIndexTipPosition4 = (__runInitializers(this, _globalRightIndexTipPosition_extraInitializers), Vector4.Zero());
                this.alphaMode = Constants.ALPHA_DISABLE;
                this.backFaceCulling = false;
                const blobTextureUrl = Tools.GetAssetUrl(_a.BLOB_TEXTURE_URL);
                const iridescentMapUrl = Tools.GetAssetUrl(_a.IM_TEXTURE_URL);
                this._blobTexture = new Texture(blobTextureUrl, this.getScene(), true, false, Texture.NEAREST_SAMPLINGMODE);
                this._iridescentMap = new Texture(iridescentMapUrl, this.getScene(), true, false, Texture.NEAREST_SAMPLINGMODE);
            }
            needAlphaBlending() {
                return false;
            }
            needAlphaTesting() {
                return false;
            }
            getAlphaTestTexture() {
                return null;
            }
            isReadyForSubMesh(mesh, subMesh) {
                const drawWrapper = subMesh._drawWrapper;
                if (this.isFrozen) {
                    if (drawWrapper.effect && drawWrapper._wasPreviouslyReady) {
                        return true;
                    }
                }
                if (!this._blobTexture.isReady() || !this._iridescentMap.isReady()) {
                    return false;
                }
                if (!subMesh.materialDefines) {
                    subMesh.materialDefines = new FluentBackplateMaterialDefines();
                }
                const defines = subMesh.materialDefines;
                const scene = this.getScene();
                if (this._isReadyForSubMesh(subMesh)) {
                    return true;
                }
                const engine = scene.getEngine();
                // Attribs
                PrepareDefinesForAttributes(mesh, defines, false, false);
                // Get correct effect
                if (defines.isDirty) {
                    defines.markAsProcessed();
                    scene.resetCachedMaterial();
                    // Fallbacks
                    const fallbacks = new EffectFallbacks();
                    if (defines.FOG) {
                        fallbacks.addFallback(1, "FOG");
                    }
                    HandleFallbacksForShadows(defines, fallbacks);
                    defines.IMAGEPROCESSINGPOSTPROCESS = scene.imageProcessingConfiguration.applyByPostProcess;
                    //Attributes
                    const attribs = [VertexBuffer.PositionKind];
                    if (defines.NORMAL) {
                        attribs.push(VertexBuffer.NormalKind);
                    }
                    if (defines.UV1) {
                        attribs.push(VertexBuffer.UVKind);
                    }
                    if (defines.UV2) {
                        attribs.push(VertexBuffer.UV2Kind);
                    }
                    if (defines.VERTEXCOLOR) {
                        attribs.push(VertexBuffer.ColorKind);
                    }
                    if (defines.TANGENT) {
                        attribs.push(VertexBuffer.TangentKind);
                    }
                    PrepareAttributesForInstances(attribs, defines);
                    // Legacy browser patch
                    const shaderName = "fluentBackplate";
                    const join = defines.toString();
                    const uniforms = [
                        "world",
                        "viewProjection",
                        "cameraPosition",
                        "_Radius_",
                        "_Line_Width_",
                        "_Absolute_Sizes_",
                        "_Filter_Width_",
                        "_Base_Color_",
                        "_Line_Color_",
                        "_Radius_Top_Left_",
                        "_Radius_Top_Right_",
                        "_Radius_Bottom_Left_",
                        "_Radius_Bottom_Right_",
                        "_Blob_Position_",
                        "_Blob_Intensity_",
                        "_Blob_Near_Size_",
                        "_Blob_Far_Size_",
                        "_Blob_Near_Distance_",
                        "_Blob_Far_Distance_",
                        "_Blob_Fade_Length_",
                        "_Blob_Pulse_",
                        "_Blob_Fade_",
                        "_Blob_Texture_",
                        "_Blob_Position_2_",
                        "_Blob_Near_Size_2_",
                        "_Blob_Pulse_2_",
                        "_Blob_Fade_2_",
                        "_Rate_",
                        "_Highlight_Color_",
                        "_Highlight_Width_",
                        "_Highlight_Transform_",
                        "_Highlight_",
                        "_Iridescence_Intensity_",
                        "_Iridescence_Edge_Intensity_",
                        "_Angle_",
                        "_Fade_Out_",
                        "_Reflected_",
                        "_Frequency_",
                        "_Vertical_Offset_",
                        "_Iridescent_Map_",
                        "_Use_Global_Left_Index_",
                        "_Use_Global_Right_Index_",
                        "Global_Left_Index_Tip_Position",
                        "Global_Right_Index_Tip_Position",
                    ];
                    const samplers = ["_Blob_Texture_", "_Iridescent_Map_"];
                    const uniformBuffers = [];
                    PrepareUniformsAndSamplersList({
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: defines,
                        maxSimultaneousLights: 4,
                        shaderLanguage: this._shaderLanguage,
                    });
                    subMesh.setEffect(scene.getEngine().createEffect(shaderName, {
                        attributes: attribs,
                        uniformsNames: uniforms,
                        uniformBuffersNames: uniformBuffers,
                        samplers: samplers,
                        defines: join,
                        fallbacks: fallbacks,
                        onCompiled: this.onCompiled,
                        onError: this.onError,
                        indexParameters: { maxSimultaneousLights: 4 },
                        shaderLanguage: this._shaderLanguage,
                        extraInitializationsAsync: this._shadersLoaded
                            ? undefined
                            : async () => {
                                if (this.shaderLanguage === 1 /* ShaderLanguage.WGSL */) {
                                    await Promise.all([import("./wgsl/fluentBackplate.vertex.js"), import("./wgsl/fluentBackplate.fragment.js")]);
                                }
                                else {
                                    await Promise.all([import("./shaders/fluentBackplate.vertex.js"), import("./shaders/fluentBackplate.fragment.js")]);
                                }
                                this._shadersLoaded = true;
                            },
                    }, engine), defines, this._materialContext);
                }
                if (!subMesh.effect || !subMesh.effect.isReady()) {
                    return false;
                }
                defines._renderId = scene.getRenderId();
                drawWrapper._wasPreviouslyReady = true;
                return true;
            }
            bindForSubMesh(world, mesh, subMesh) {
                const defines = subMesh.materialDefines;
                if (!defines) {
                    return;
                }
                const effect = subMesh.effect;
                if (!effect) {
                    return;
                }
                this._activeEffect = effect;
                // Matrices
                this.bindOnlyWorldMatrix(world);
                this._activeEffect.setMatrix("viewProjection", this.getScene().getTransformMatrix());
                this._activeEffect.setVector3("cameraPosition", this.getScene().activeCamera?.position ?? Vector3.ZeroReadOnly);
                // "Round Rect"
                this._activeEffect.setFloat("_Radius_", this.radius);
                this._activeEffect.setFloat("_Line_Width_", this.lineWidth);
                this._activeEffect.setFloat("_Absolute_Sizes_", this.absoluteSizes ? 1.0 : 0.0);
                this._activeEffect.setFloat("_Filter_Width_", this._filterWidth);
                this._activeEffect.setDirectColor4("_Base_Color_", this.baseColor);
                this._activeEffect.setDirectColor4("_Line_Color_", this.lineColor);
                // "Radii Multipliers"
                this._activeEffect.setFloat("_Radius_Top_Left_", 1);
                this._activeEffect.setFloat("_Radius_Top_Right_", 1.0);
                this._activeEffect.setFloat("_Radius_Bottom_Left_", 1.0);
                this._activeEffect.setFloat("_Radius_Bottom_Right_", 1.0);
                // "Blob"
                //define BLOB_ENABLE true;
                this._activeEffect.setFloat("_Blob_Intensity_", this.blobIntensity);
                this._activeEffect.setFloat("_Blob_Near_Size_", this.blobNearSize);
                this._activeEffect.setFloat("_Blob_Far_Size_", this.blobFarSize);
                this._activeEffect.setFloat("_Blob_Near_Distance_", this.blobNearDistance);
                this._activeEffect.setFloat("_Blob_Far_Distance_", this.blobFarDistance);
                this._activeEffect.setFloat("_Blob_Fade_Length_", this.blobFadeLength);
                this._activeEffect.setFloat("_Blob_Pulse_", this.blobPulse);
                this._activeEffect.setFloat("_Blob_Fade_", this.blobFade);
                // "Blob Texture"
                this._activeEffect.setTexture("_Blob_Texture_", this._blobTexture);
                // "Blob 2"
                //define BLOB_ENABLE_2 true;
                this._activeEffect.setFloat("_Blob_Near_Size_2_", this.blobNearSize2);
                this._activeEffect.setFloat("_Blob_Pulse_2_", this.blobPulse2);
                this._activeEffect.setFloat("_Blob_Fade_2_", this.blobFade2);
                // "Line Highlight"
                this._activeEffect.setFloat("_Rate_", this._rate);
                this._activeEffect.setDirectColor4("_Highlight_Color_", this.highlightColor);
                this._activeEffect.setFloat("_Highlight_Width_", this.highlightWidth);
                this._activeEffect.setVector4("_Highlight_Transform_", this._highlightTransform);
                this._activeEffect.setFloat("_Highlight_", this._highlight);
                // "Iridescence"
                this._activeEffect.setFloat("_Iridescence_Intensity_", this.iridescenceIntensity);
                this._activeEffect.setFloat("_Iridescence_Edge_Intensity_", this.iridescenceEdgeIntensity);
                this._activeEffect.setFloat("_Angle_", this._angle);
                // "Fade"
                this._activeEffect.setFloat("_Fade_Out_", this.fadeOut);
                // "Antialiasing"
                //define SMOOTH_EDGES true;
                // "ChooseAngle"
                this._activeEffect.setFloat("_Reflected_", this._reflected ? 1.0 : 0.0);
                // "Multiply"
                this._activeEffect.setFloat("_Frequency_", this._frequency);
                this._activeEffect.setFloat("_Vertical_Offset_", this._verticalOffset);
                // "Color Texture"
                //define IRIDESCENT_MAP_ENABLE true;
                this._activeEffect.setTexture("_Iridescent_Map_", this._iridescentMap);
                // "Global"
                this._activeEffect.setFloat("_Use_Global_Left_Index_", 1.0);
                this._activeEffect.setFloat("_Use_Global_Right_Index_", 1.0);
                this._globalLeftIndexTipPosition4.set(this.globalLeftIndexTipPosition.x, this.globalLeftIndexTipPosition.y, this.globalLeftIndexTipPosition.z, 1.0);
                this._activeEffect.setVector4("Global_Left_Index_Tip_Position", this._globalLeftIndexTipPosition4);
                this._globalRightIndexTipPosition4.set(this.globalRightIndexTipPosition.x, this.globalRightIndexTipPosition.y, this.globalRightIndexTipPosition.z, 1.0);
                this._activeEffect.setVector4("Global_Right_Index_Tip_Position", this._globalRightIndexTipPosition4);
                this._afterBind(mesh, this._activeEffect, subMesh);
            }
            /**
             * Get the list of animatables in the material.
             * @returns the list of animatables object used in the material
             */
            getAnimatables() {
                return [];
            }
            /**
             * Disposes the material.
             * @param forceDisposeEffect specifies if effects should be forcefully disposed
             */
            dispose(forceDisposeEffect) {
                super.dispose(forceDisposeEffect);
                this._blobTexture.dispose();
                this._iridescentMap.dispose();
            }
            clone(name) {
                return SerializationHelper.Clone(() => new _a(name, this.getScene()), this);
            }
            serialize() {
                const serializationObject = super.serialize();
                serializationObject.customType = "BABYLON.FluentBackplateMaterial";
                return serializationObject;
            }
            getClassName() {
                return "FluentBackplateMaterial";
            }
            // Statics
            /**
             * Creates a fluent backplate material from parsed material data.
             * @param source defines the JSON representation of the material
             * @param scene defines the hosting scene
             * @param rootUrl defines the root URL to use to load textures and relative dependencies
             * @returns a new fluent backplate material
             */
            static Parse(source, scene, rootUrl) {
                return SerializationHelper.Parse(() => new _a(source.name, scene), source, scene, rootUrl);
            }
        },
        (() => {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            _radius_decorators = [serialize()];
            _lineWidth_decorators = [serialize()];
            _absoluteSizes_decorators = [serialize()];
            _baseColor_decorators = [serialize()];
            _lineColor_decorators = [serialize()];
            _blobIntensity_decorators = [serialize()];
            _blobFarSize_decorators = [serialize()];
            _blobNearDistance_decorators = [serialize()];
            _blobFarDistance_decorators = [serialize()];
            _blobFadeLength_decorators = [serialize()];
            _blobNearSize_decorators = [serialize()];
            _blobPulse_decorators = [serialize()];
            _blobFade_decorators = [serialize()];
            _blobNearSize2_decorators = [serialize()];
            _blobPulse2_decorators = [serialize()];
            _blobFade2_decorators = [serialize()];
            _highlightColor_decorators = [serialize()];
            _highlightWidth_decorators = [serialize()];
            _iridescenceIntensity_decorators = [serialize()];
            _iridescenceEdgeIntensity_decorators = [serialize()];
            _fadeOut_decorators = [serialize()];
            _globalLeftIndexTipPosition_decorators = [serializeAsVector3()];
            _globalRightIndexTipPosition_decorators = [serializeAsVector3()];
            __esDecorate(null, null, _radius_decorators, { kind: "field", name: "radius", static: false, private: false, access: { has: obj => "radius" in obj, get: obj => obj.radius, set: (obj, value) => { obj.radius = value; } }, metadata: _metadata }, _radius_initializers, _radius_extraInitializers);
            __esDecorate(null, null, _lineWidth_decorators, { kind: "field", name: "lineWidth", static: false, private: false, access: { has: obj => "lineWidth" in obj, get: obj => obj.lineWidth, set: (obj, value) => { obj.lineWidth = value; } }, metadata: _metadata }, _lineWidth_initializers, _lineWidth_extraInitializers);
            __esDecorate(null, null, _absoluteSizes_decorators, { kind: "field", name: "absoluteSizes", static: false, private: false, access: { has: obj => "absoluteSizes" in obj, get: obj => obj.absoluteSizes, set: (obj, value) => { obj.absoluteSizes = value; } }, metadata: _metadata }, _absoluteSizes_initializers, _absoluteSizes_extraInitializers);
            __esDecorate(null, null, _baseColor_decorators, { kind: "field", name: "baseColor", static: false, private: false, access: { has: obj => "baseColor" in obj, get: obj => obj.baseColor, set: (obj, value) => { obj.baseColor = value; } }, metadata: _metadata }, _baseColor_initializers, _baseColor_extraInitializers);
            __esDecorate(null, null, _lineColor_decorators, { kind: "field", name: "lineColor", static: false, private: false, access: { has: obj => "lineColor" in obj, get: obj => obj.lineColor, set: (obj, value) => { obj.lineColor = value; } }, metadata: _metadata }, _lineColor_initializers, _lineColor_extraInitializers);
            __esDecorate(null, null, _blobIntensity_decorators, { kind: "field", name: "blobIntensity", static: false, private: false, access: { has: obj => "blobIntensity" in obj, get: obj => obj.blobIntensity, set: (obj, value) => { obj.blobIntensity = value; } }, metadata: _metadata }, _blobIntensity_initializers, _blobIntensity_extraInitializers);
            __esDecorate(null, null, _blobFarSize_decorators, { kind: "field", name: "blobFarSize", static: false, private: false, access: { has: obj => "blobFarSize" in obj, get: obj => obj.blobFarSize, set: (obj, value) => { obj.blobFarSize = value; } }, metadata: _metadata }, _blobFarSize_initializers, _blobFarSize_extraInitializers);
            __esDecorate(null, null, _blobNearDistance_decorators, { kind: "field", name: "blobNearDistance", static: false, private: false, access: { has: obj => "blobNearDistance" in obj, get: obj => obj.blobNearDistance, set: (obj, value) => { obj.blobNearDistance = value; } }, metadata: _metadata }, _blobNearDistance_initializers, _blobNearDistance_extraInitializers);
            __esDecorate(null, null, _blobFarDistance_decorators, { kind: "field", name: "blobFarDistance", static: false, private: false, access: { has: obj => "blobFarDistance" in obj, get: obj => obj.blobFarDistance, set: (obj, value) => { obj.blobFarDistance = value; } }, metadata: _metadata }, _blobFarDistance_initializers, _blobFarDistance_extraInitializers);
            __esDecorate(null, null, _blobFadeLength_decorators, { kind: "field", name: "blobFadeLength", static: false, private: false, access: { has: obj => "blobFadeLength" in obj, get: obj => obj.blobFadeLength, set: (obj, value) => { obj.blobFadeLength = value; } }, metadata: _metadata }, _blobFadeLength_initializers, _blobFadeLength_extraInitializers);
            __esDecorate(null, null, _blobNearSize_decorators, { kind: "field", name: "blobNearSize", static: false, private: false, access: { has: obj => "blobNearSize" in obj, get: obj => obj.blobNearSize, set: (obj, value) => { obj.blobNearSize = value; } }, metadata: _metadata }, _blobNearSize_initializers, _blobNearSize_extraInitializers);
            __esDecorate(null, null, _blobPulse_decorators, { kind: "field", name: "blobPulse", static: false, private: false, access: { has: obj => "blobPulse" in obj, get: obj => obj.blobPulse, set: (obj, value) => { obj.blobPulse = value; } }, metadata: _metadata }, _blobPulse_initializers, _blobPulse_extraInitializers);
            __esDecorate(null, null, _blobFade_decorators, { kind: "field", name: "blobFade", static: false, private: false, access: { has: obj => "blobFade" in obj, get: obj => obj.blobFade, set: (obj, value) => { obj.blobFade = value; } }, metadata: _metadata }, _blobFade_initializers, _blobFade_extraInitializers);
            __esDecorate(null, null, _blobNearSize2_decorators, { kind: "field", name: "blobNearSize2", static: false, private: false, access: { has: obj => "blobNearSize2" in obj, get: obj => obj.blobNearSize2, set: (obj, value) => { obj.blobNearSize2 = value; } }, metadata: _metadata }, _blobNearSize2_initializers, _blobNearSize2_extraInitializers);
            __esDecorate(null, null, _blobPulse2_decorators, { kind: "field", name: "blobPulse2", static: false, private: false, access: { has: obj => "blobPulse2" in obj, get: obj => obj.blobPulse2, set: (obj, value) => { obj.blobPulse2 = value; } }, metadata: _metadata }, _blobPulse2_initializers, _blobPulse2_extraInitializers);
            __esDecorate(null, null, _blobFade2_decorators, { kind: "field", name: "blobFade2", static: false, private: false, access: { has: obj => "blobFade2" in obj, get: obj => obj.blobFade2, set: (obj, value) => { obj.blobFade2 = value; } }, metadata: _metadata }, _blobFade2_initializers, _blobFade2_extraInitializers);
            __esDecorate(null, null, _highlightColor_decorators, { kind: "field", name: "highlightColor", static: false, private: false, access: { has: obj => "highlightColor" in obj, get: obj => obj.highlightColor, set: (obj, value) => { obj.highlightColor = value; } }, metadata: _metadata }, _highlightColor_initializers, _highlightColor_extraInitializers);
            __esDecorate(null, null, _highlightWidth_decorators, { kind: "field", name: "highlightWidth", static: false, private: false, access: { has: obj => "highlightWidth" in obj, get: obj => obj.highlightWidth, set: (obj, value) => { obj.highlightWidth = value; } }, metadata: _metadata }, _highlightWidth_initializers, _highlightWidth_extraInitializers);
            __esDecorate(null, null, _iridescenceIntensity_decorators, { kind: "field", name: "iridescenceIntensity", static: false, private: false, access: { has: obj => "iridescenceIntensity" in obj, get: obj => obj.iridescenceIntensity, set: (obj, value) => { obj.iridescenceIntensity = value; } }, metadata: _metadata }, _iridescenceIntensity_initializers, _iridescenceIntensity_extraInitializers);
            __esDecorate(null, null, _iridescenceEdgeIntensity_decorators, { kind: "field", name: "iridescenceEdgeIntensity", static: false, private: false, access: { has: obj => "iridescenceEdgeIntensity" in obj, get: obj => obj.iridescenceEdgeIntensity, set: (obj, value) => { obj.iridescenceEdgeIntensity = value; } }, metadata: _metadata }, _iridescenceEdgeIntensity_initializers, _iridescenceEdgeIntensity_extraInitializers);
            __esDecorate(null, null, _fadeOut_decorators, { kind: "field", name: "fadeOut", static: false, private: false, access: { has: obj => "fadeOut" in obj, get: obj => obj.fadeOut, set: (obj, value) => { obj.fadeOut = value; } }, metadata: _metadata }, _fadeOut_initializers, _fadeOut_extraInitializers);
            __esDecorate(null, null, _globalLeftIndexTipPosition_decorators, { kind: "field", name: "globalLeftIndexTipPosition", static: false, private: false, access: { has: obj => "globalLeftIndexTipPosition" in obj, get: obj => obj.globalLeftIndexTipPosition, set: (obj, value) => { obj.globalLeftIndexTipPosition = value; } }, metadata: _metadata }, _globalLeftIndexTipPosition_initializers, _globalLeftIndexTipPosition_extraInitializers);
            __esDecorate(null, null, _globalRightIndexTipPosition_decorators, { kind: "field", name: "globalRightIndexTipPosition", static: false, private: false, access: { has: obj => "globalRightIndexTipPosition" in obj, get: obj => obj.globalRightIndexTipPosition, set: (obj, value) => { obj.globalRightIndexTipPosition = value; } }, metadata: _metadata }, _globalRightIndexTipPosition_initializers, _globalRightIndexTipPosition_extraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        /**
         * URL pointing to the texture used to define the coloring for the fluent blob effect.
         */
        _a.BLOB_TEXTURE_URL = "https://assets.babylonjs.com/core/MRTK/mrtk-fluent-backplate-blob.png",
        /**
         * URL pointing to the texture used to define iridescent map.
         */
        _a.IM_TEXTURE_URL = "https://assets.babylonjs.com/core/MRTK/mrtk-fluent-backplate-iridescence.png",
        _a;
})();
export { FluentBackplateMaterial };
let _Registered = false;
/**
 * Registers the FluentBackplateMaterial class with the type store for serialization support.
 * Safe to call multiple times; only the first call has an effect.
 */
export function RegisterFluentBackplateMaterial() {
    if (_Registered) {
        return;
    }
    _Registered = true;
    RegisterClass("BABYLON.GUI.FluentBackplateMaterial", FluentBackplateMaterial);
}
